
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

    <title>Chaos Game Representation &#8212; Computational Discovery on Jupyter</title>
    
  <!-- Loaded before other Sphinx assets -->
  <link href="../_static/styles/theme.css?digest=1999514e3f237ded88cf" rel="stylesheet">
<link href="../_static/styles/pydata-sphinx-theme.css?digest=1999514e3f237ded88cf" rel="stylesheet">

    
  <link rel="stylesheet"
    href="../_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    <link rel="stylesheet" type="text/css" href="../_static/pygments.css" />
    <link rel="stylesheet" href="../_static/styles/sphinx-book-theme.css?digest=62ba249389abaaa9ffc34bf36a076bdc1d65ee18" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../_static/mystnb.css" />
    <link rel="stylesheet" type="text/css" href="../_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="../_static/proof.css" />
    <link rel="stylesheet" type="text/css" href="../_static/design-style.b7bb847fb20b106c3d81b95245e65545.min.css" />
    
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../_static/scripts/pydata-sphinx-theme.js?digest=1999514e3f237ded88cf">

    <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
    <script src="../_static/jquery.js"></script>
    <script src="../_static/underscore.js"></script>
    <script src="../_static/doctools.js"></script>
    <script src="../_static/clipboard.min.js"></script>
    <script src="../_static/copybutton.js"></script>
    <script src="../_static/scripts/sphinx-book-theme.js?digest=f31d14ad54b65d19161ba51d4ffff3a77ae00456"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="../_static/togglebutton.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown, .tag_hide_input div.cell_input, .tag_hide-input div.cell_input, .tag_hide_output div.cell_output, .tag_hide-output div.cell_output, .tag_hide_cell.cell, .tag_hide-cell.cell';</script>
    <script src="../_static/design-tabs.js"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"
const thebe_selector = ".thebe,.cell"
const thebe_selector_input = "pre"
const thebe_selector_output = ".output, .cell_output"
</script>
    <script async="async" src="../_static/sphinx-thebe.js"></script>
    <script>window.MathJax = {"options": {"processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
    <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <link rel="shortcut icon" href="../_static/logo.jpg"/>
    <link rel="author" title="About these documents" href="../about.html" />
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Zanshin" href="../Looking-back.html" />
    <link rel="prev" title="Mandelbrot Polynomials and Matrices" href="mandelbrot.html" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="None">
    

    <!-- Google Analytics -->
    
  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="60">
<!-- Checkboxes to toggle the left sidebar -->
<input type="checkbox" class="sidebar-toggle" name="__navigation" id="__navigation" aria-label="Toggle navigation sidebar">
<label class="overlay overlay-navbar" for="__navigation">
    <div class="visually-hidden">Toggle navigation sidebar</div>
</label>
<!-- Checkboxes to toggle the in-page toc -->
<input type="checkbox" class="sidebar-toggle" name="__page-toc" id="__page-toc" aria-label="Toggle in-page Table of Contents">
<label class="overlay overlay-pagetoc" for="__page-toc">
    <div class="visually-hidden">Toggle in-page Table of Contents</div>
</label>
<!-- Headers at the top -->
<div class="announcement header-item noprint"></div>
<div class="header header-item noprint"></div>

    
    <div class="container-fluid" id="banner"></div>

    

    <div class="container-xl">
      <div class="row">
          
<!-- Sidebar -->
<div class="bd-sidebar noprint" id="site-navigation">
    <div class="bd-sidebar__content">
        <div class="bd-sidebar__top"><div class="navbar-brand-box">
    <a class="navbar-brand text-wrap" href="../index.html">
      
        <!-- `logo` is deprecated in Sphinx 4.0, so remove this when we stop supporting 3 -->
        
      
      
      <img src="../_static/logo.jpg" class="logo" alt="logo">
      
      
      <h1 class="site-logo" id="site-title">Computational Discovery on Jupyter</h1>
      
    </a>
</div><form class="bd-search d-flex align-items-center" action="../search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search this book..." aria-label="Search this book..." autocomplete="off" >
</form><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item active">
        
        <ul class="nav bd-sidenav bd-sidenav__home-link">
            <li class="toctree-l1">
                <a class="reference internal" href="../index.html">
                    Welcome to Computational Discovery on Jupyter
                </a>
            </li>
        </ul>
        <p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Preface
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../preamble.html">
   Preamble
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../Installing-Python-and-Jupyter.html">
   Before we begin: installing Python and Jupyter
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Contents
 </span>
</p>
<ul class="current nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="Computing-with-Fibonacci.html">
   Computing with Fibonacci Numbers
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="continued-fractions.html">
   Continued Fractions
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="rootfinding.html">
   Rootfinding, Newton’s Method, and Dynamical Systems
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="fractals-and-julia-sets.html">
   Fractals and Julia Sets
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="bohemian-matrices.html">
   Bounded Height Matrices of Integers (Bohemian Matrices)
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="mandelbrot.html">
   Mandelbrot Polynomials and Matrices
  </a>
 </li>
 <li class="toctree-l1 current active">
  <a class="current reference internal" href="#">
   Chaos Game Representation
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Epilogue
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../Looking-back.html">
   Zanshin
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Appendix
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../Appendix/What-to-do-when-things-go-wrong.html">
   What do to when things go wrong
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../Appendix/floating-point.html">
   The Bare Minimum about Floating-Point
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../Appendix/binomials-factorials-combinatorics.html">
   Binomials, Factorials, and other Combinatorial Things
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../Appendix/complex-numbers.html">
   Complex Numbers
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../Appendix/symbolic-computation.html">
   Symbolic Computation: The Pitfalls
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Reports on Activities
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../Solutions/Reports%20on%20Fibonacci%20Activities.html">
   Reports on Fibonacci Activities
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../Solutions/Reports%20on%20Continued%20Fractions%20Activities.html">
   Reports on Continued Fractions Activities
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../Solutions/Reports%20on%20Rootfinding%20Activities.html">
   Reports on Rootfinding Activities
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../Solutions/Reports%20on%20Julia%20Set%20Activities.html">
   Reports on Julia Set Activities
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../Solutions/Reports%20on%20Bohemian%20Activities.html">
   Reports on Bohemian activities
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../Solutions/Reports%20on%20Mandelbrot%20Activities.html">
   Reports on Mandelbrot Activities
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../Solutions/Reports%20on%20Chaos%20Game%20Activities.html">
   Reports on Chaos Game Activities
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Back Matter
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../acknowledgements.html">
   Acknowledgements
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../about.html">
   About the authors
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../references.html">
   References
  </a>
 </li>
</ul>

    </div>
</nav></div>
        <div class="bd-sidebar__bottom">
             <!-- To handle the deprecated key -->
            
            <div class="navbar_extra_footer">
            Powered by <a href="https://jupyterbook.org">Jupyter Book</a>
            </div>
            
        </div>
    </div>
    <div id="rtd-footer-container"></div>
</div>


          


          
<!-- A tiny helper pixel to detect if we've scrolled -->
<div class="sbt-scroll-pixel-helper"></div>
<!-- Main content -->
<div class="col py-0 content-container">
    
    <div class="header-article row sticky-top noprint">
        



<div class="col py-1 d-flex header-article-main">
    <div class="header-article__left">
        
        <label for="__navigation"
  class="headerbtn"
  data-toggle="tooltip"
data-placement="right"
title="Toggle navigation"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-bars"></i>
  </span>

</label>

        
    </div>
    <div class="header-article__right">
<div class="menu-dropdown menu-dropdown-launch-buttons">
  <button class="headerbtn menu-dropdown__trigger"
      aria-label="Launch interactive content">
      <i class="fas fa-rocket"></i>
  </button>
  <div class="menu-dropdown__content">
    <ul>
    </ul>
  </div>
</div>

<button onclick="toggleFullScreen()"
  class="headerbtn"
  data-toggle="tooltip"
data-placement="bottom"
title="Fullscreen mode"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>

<div class="menu-dropdown menu-dropdown-repository-buttons">
  <button class="headerbtn menu-dropdown__trigger"
      aria-label="Source repositories">
      <i class="fab fa-github"></i>
  </button>
  <div class="menu-dropdown__content">
    <ul>
      <li>
        <a href="https://github.com/Computational-Discovery-on-Jupyter/Computational-Discovery-on-Jupyter"
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Source repository"
>
  

<span class="headerbtn__icon-container">
  <i class="fab fa-github"></i>
  </span>
<span class="headerbtn__text-container">repository</span>
</a>

      </li>
      
      <li>
        <a href="https://github.com/Computational-Discovery-on-Jupyter/Computational-Discovery-on-Jupyter/edit/master/book/Contents/chaos-game-representation.ipynb"
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Edit this page"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-pencil-alt"></i>
  </span>
<span class="headerbtn__text-container">suggest edit</span>
</a>

      </li>
      
    </ul>
  </div>
</div>

<div class="menu-dropdown menu-dropdown-download-buttons">
  <button class="headerbtn menu-dropdown__trigger"
      aria-label="Download this page">
      <i class="fas fa-download"></i>
  </button>
  <div class="menu-dropdown__content">
    <ul>
      <li>
        <a href="../_sources/Contents/chaos-game-representation.ipynb"
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Download source file"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="headerbtn__text-container">.ipynb</span>
</a>

      </li>
      
      <li>
        
<button onclick="printPdf(this)"
  class="headerbtn"
  data-toggle="tooltip"
data-placement="left"
title="Print to PDF"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="headerbtn__text-container">.pdf</span>
</button>

      </li>
      
    </ul>
  </div>
</div>
<label for="__page-toc"
  class="headerbtn headerbtn-page-toc"
  
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-list"></i>
  </span>

</label>

    </div>
</div>

<!-- Table of contents -->
<div class="col-md-3 bd-toc show noprint">
    <div class="tocsection onthispage pt-5 pb-3">
        <i class="fas fa-list"></i> Contents
    </div>
    <nav id="bd-toc-nav" aria-label="Page">
        <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#introduction">
   Introduction
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#a-note-to-the-student-reader">
   A Note to the Student/Reader
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#a-note-to-the-instructor">
   A Note to the Instructor
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#creating-fractals-using-iterated-function-systems-ifs">
   Creating Fractals using Iterated Function Systems (IFS)
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#affine-transformation-in-the-euclidean-plane">
     Affine Transformation in the Euclidean Plane
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#iterated-function-system">
     Iterated Function System
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#chaos-game">
   Chaos Game
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#generalization-of-cgr">
     Generalization of CGR
    </a>
    <ul class="nav section-nav flex-column">
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#different-polygons">
       Different Polygons
      </a>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#dividing-rate">
       Dividing Rate
      </a>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#sec-generalization">
       Generalization of CGR
      </a>
     </li>
    </ul>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#cgr-of-biological-sequences">
   CGR of Biological Sequences
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#nucleotide-sequences">
     Nucleotide Sequences
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#map-of-life">
     Map of Life
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#protein-visualization">
     Protein Visualization
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#cgr-of-mathematical-sequences">
   CGR of Mathematical Sequences
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#digits-of-pi">
     Digits of Pi
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#fibonacci-sequence">
     Fibonacci Sequence
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#prime-numbers">
     Prime Numbers
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#partial-quotients-of-continued-fractions">
     Partial Quotients of Continued Fractions
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#random-vs-pseudorandom">
   Random vs Pseudorandom
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#concluding-remarks">
   Concluding Remarks
  </a>
 </li>
</ul>

    </nav>
</div>
    </div>
    <div class="article row">
        <div class="col pl-md-3 pl-lg-5 content-container">
            <!-- Table of contents that is only displayed when printing the page -->
            <div id="jb-print-docs-body" class="onlyprint">
                <h1>Chaos Game Representation</h1>
                <!-- Table of contents -->
                <div id="print-main-content">
                    <div id="jb-print-toc">
                        
                        <div>
                            <h2> Contents </h2>
                        </div>
                        <nav aria-label="Page">
                            <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#introduction">
   Introduction
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#a-note-to-the-student-reader">
   A Note to the Student/Reader
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#a-note-to-the-instructor">
   A Note to the Instructor
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#creating-fractals-using-iterated-function-systems-ifs">
   Creating Fractals using Iterated Function Systems (IFS)
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#affine-transformation-in-the-euclidean-plane">
     Affine Transformation in the Euclidean Plane
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#iterated-function-system">
     Iterated Function System
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#chaos-game">
   Chaos Game
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#generalization-of-cgr">
     Generalization of CGR
    </a>
    <ul class="nav section-nav flex-column">
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#different-polygons">
       Different Polygons
      </a>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#dividing-rate">
       Dividing Rate
      </a>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#sec-generalization">
       Generalization of CGR
      </a>
     </li>
    </ul>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#cgr-of-biological-sequences">
   CGR of Biological Sequences
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#nucleotide-sequences">
     Nucleotide Sequences
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#map-of-life">
     Map of Life
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#protein-visualization">
     Protein Visualization
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#cgr-of-mathematical-sequences">
   CGR of Mathematical Sequences
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#digits-of-pi">
     Digits of Pi
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#fibonacci-sequence">
     Fibonacci Sequence
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#prime-numbers">
     Prime Numbers
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#partial-quotients-of-continued-fractions">
     Partial Quotients of Continued Fractions
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#random-vs-pseudorandom">
   Random vs Pseudorandom
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#concluding-remarks">
   Concluding Remarks
  </a>
 </li>
</ul>

                        </nav>
                    </div>
                </div>
            </div>
            <main id="main-content" role="main">
                
              <div>
                
  <section class="tex2jax_ignore mathjax_ignore" id="chaos-game-representation">
<h1>Chaos Game Representation<a class="headerlink" href="#chaos-game-representation" title="Permalink to this headline">#</a></h1>
<div class="cell tag_hide-cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># run this code block before executing the rest of the notebook</span>
<span class="o">%</span><span class="k">matplotlib</span> inline
<span class="kn">import</span> <span class="nn">math</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>
<span class="kn">from</span> <span class="nn">mpl_toolkits</span> <span class="kn">import</span> <span class="n">mplot3d</span>
<span class="kn">from</span> <span class="nn">mpl_toolkits.mplot3d</span> <span class="kn">import</span> <span class="n">Axes3D</span>
<span class="kn">from</span> <span class="nn">matplotlib.colors</span> <span class="kn">import</span> <span class="n">ListedColormap</span>
<span class="kn">import</span> <span class="nn">random</span>
<span class="kn">import</span> <span class="nn">csv</span>
<span class="kn">import</span> <span class="nn">sympy</span>
<span class="kn">from</span> <span class="nn">sympy.ntheory.continued_fraction</span> <span class="kn">import</span> <span class="n">continued_fraction_iterator</span>
<span class="kn">import</span> <span class="nn">skimage.io</span>
<span class="kn">from</span> <span class="nn">skimage.metrics</span> <span class="kn">import</span> <span class="n">structural_similarity</span> <span class="k">as</span> <span class="n">ssim</span>
<span class="kn">import</span> <span class="nn">pickle</span>
<span class="kn">import</span> <span class="nn">seaborn</span> <span class="k">as</span> <span class="nn">sns</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">from</span> <span class="nn">sklearn</span> <span class="kn">import</span> <span class="n">manifold</span><span class="p">,</span> <span class="n">decomposition</span>
<span class="kn">import</span> <span class="nn">multiprocessing</span> <span class="k">as</span> <span class="nn">mp</span>
</pre></div>
</div>
</div>
</div>
<section id="introduction">
<h2>Introduction<a class="headerlink" href="#introduction" title="Permalink to this headline">#</a></h2>
<blockquote class="epigraph">
<div><p>A heavy warning used to be given [by lecturers] that pictures are not rigorous; this has never had its bluff called and has permanently frightened its victims into playing for safety. Some pictures, of course, are not rigorous, but I should say most are (and I use them whenever possible myself).</p>
<p class="attribution">—<a class="reference external" href="https://mathshistory.st-andrews.ac.uk/Biographies/Littlewood/">J.E. Littlewood</a>,  <span id="id1">[<a class="reference internal" href="../references.html#id11" title="J. E. Littlewood. Littlewood's miscellany. Cambridge University Press, 1986.">Littlewood, 1986</a>]</span>(p. 53) </p>
</div></blockquote>
<p>Finding hidden patterns in long sequences can be both difficult and valuable. Representing these sequences in a visual way can often help. The human brain is very adept at detecting patterns visually. Indeed, sometimes <em>too</em> adept: when humans detect spurious patterns that <em>aren’t</em> really there, it’s called <em>pareidolia</em>. That human tendency is what originally provoked the “heavy warning” against pictures that Littlewood complained about above. Nonetheless, pictures can help a lot, to generate conjectures, and sometimes even indicate a path towards a proof, even if sometimes “only” an experimental proof using comparison with data.</p>
<p>The so-called chaos game representation (CGR) of a sequence of integers is a particularly useful technique for pattern detection, which visualizes a one-dimensional sequence in a two-dimensional space. The CGR is presented as a scatter plot, most frequently on a square in which each corner represents an element that appears in the sequence. The results of these CGRs can look like <a class="reference external" href="https://en.wikipedia.org/wiki/Fractal">fractals</a>, but even so can be visually recognizable and distinguishable from each other. Interestingly, the distinguishing characteristics can sometimes be made <em>quantitative</em>, with a good notion of “distance between images.” This is an indirect way of constructing an equivalent notion of “distance between sequences,” but it turns out to be simpler to think about.</p>
<p>There is an excellent <a class="reference external" href="https://youtu.be/5nuYD2M2AX8">video by Robin Truax on CGR and Sierpinksi’s triangle</a>; worth watching now, or you can save it for later.  Now back to our unit.</p>
<p>Many applications, such as analysis of DNA sequences <span id="id2">[<a class="reference internal" href="../references.html#id24" title="N. Goldman. Nucleotide, dinucleotide and trinucleotide frequencies explain patterns observed in chaos game representations of DNA sequences. Nucleic Acids Research, 21(10):2487–2491, 1993.">Goldman, 1993</a>, <a class="reference internal" href="../references.html#id27" title="H. J. Jeffrey. Chaos game representation of gene structure. Nucleic Acids Research, 18(8):2163–2170, 1990.">Jeffrey, 1990</a>, <a class="reference internal" href="../references.html#id28" title="H. J. Jeffrey. Chaos game visualization of sequences. Computers &amp; Graphics, 16(1):25–33, 1992.">Jeffrey, 1992</a>, <a class="reference internal" href="../references.html#id30" title="R. Karamichalis, L. Kari, S. Konstantinidis, and S. Kopecki. An investigation into inter-and intragenomic variations of graphic genomic signatures. BMC Bioinformatics, 16(1):246, 2015.">Karamichalis <em>et al.</em>, 2015</a>]</span> and protein structure <span id="id3">[<a class="reference internal" href="../references.html#id15" title="S. Basu, A. Pan, C. Dutta, and J. Das. Chaos game representation of proteins. Journal of Molecular Graphics and Modelling, 15(5):279–289, 1997.">Basu <em>et al.</em>, 1997</a>, <a class="reference internal" href="../references.html#id22" title="A. Fiser, G. E. Tusnady, and I. Simon. Chaos game representation of protein structures. Journal of molecular graphics, 12(4):302–304, 1994.">Fiser <em>et al.</em>, 1994</a>]</span>, have shown the usefulness of CGR and the notion of “distance between images”; we will discuss these applications briefly in <a class="reference internal" href="#sec-dna"><span class="std std-ref">CGR of Biological Sequences</span></a>. Before that, in <a class="reference internal" href="#sec-ifs"><span class="std std-ref">Creating Fractals using Iterated Function Systems (IFS)</span></a>, we will look at a random iteration algorithm for creating fractals. In <a class="reference internal" href="#sec-math"><span class="std std-ref">CGR of Mathematical Sequences</span></a>, we will apply CGR to simple abstract mathematical sequences, including the digits of <span class="math notranslate nohighlight">\(\pi\)</span> and the partial quotients of continued fractions. Additionally, we will look at certain “distances between images” that are produced. Finally, we will explain how some of the patterns arise, and what these depictions can indicate to us, or sometimes even tell us rigorously, about the sequences.</p>
<p>One important pedagogical purpose of this module is to provoke a discussion about randomness, or what it means to be random. In this paper, we at first use the word “random” very loosely, on purpose.  Indeed, we will use several words loosely, such as “fractal” and “<a class="reference external" href="https://en.wikipedia.org/wiki/Attractor">attractor</a>” or “attracting set”. We promise to provide references at the end that will provide more careful definitions, and tighten up the discussion.</p>
<blockquote class="epigraph">
<div><p>What is probability? I asked myself this question many years ago, and found that various authors gave different answers. I found that there were several main schools of thought with many variations.</p>
<p class="attribution">—<a class="reference external" href="http://mathshistory.st-andrews.ac.uk/Biographies/Hamming.html">Richard W. Hamming</a>, <span id="id4">[<a class="reference internal" href="../references.html#id25" title="R. W. Hamming. Art of Probability. Addison Wesley Publishing Company, 1991.">Hamming, 1991</a>]</span></p>
</div></blockquote>
<figure class="align-default" id="cgr-globin">
<a class="reference internal image-reference" href="../_images/HUMHBB_crop.png"><img alt="../_images/HUMHBB_crop.png" src="../_images/HUMHBB_crop.png" style="height: 300px;" /></a>
<figcaption>
<p><span class="caption-number">Fig. 8 </span><span class="caption-text">DNA of human beta globin region on chromosome 11 (HUMHBB)—73308 bps</span><a class="headerlink" href="#cgr-globin" title="Permalink to this image">#</a></p>
</figcaption>
</figure>
</section>
<section id="a-note-to-the-student-reader">
<h2>A Note to the Student/Reader<a class="headerlink" href="#a-note-to-the-student-reader" title="Permalink to this headline">#</a></h2>
<p>This is the most complicated unit in this OER, both for concepts and for code.  After having been exposed to the prior material, we think that you can handle this now, even though it is more complicated.  Another complication is that there are genuine applications discussed here, with real data (unlike in any of the previous chapters).  The newest parts of the unit include the necessity for handling data.  In some sense that code is not very mathematical, being data transfer rather than data transformation; but it is necessary to learn, if one is going to be working with real data.  Indeed, that may be motivating for some of you, with the buzzword “Data Science” floating in the back of your mind.  Yes, this unit can be considered as an introduction to Data Science.</p>
<p>This entails yet <em>another</em> difficulty, which is that all of the data-handling packages listed above have to be installed correctly, and this is not always straightforward.</p>
<p>As far as the math goes, we also spend quite a bit of time on the notion of <em>randomness</em>.  Again, this is largely new to this unit.</p>
</section>
<section id="a-note-to-the-instructor">
<h2>A Note to the Instructor<a class="headerlink" href="#a-note-to-the-instructor" title="Permalink to this headline">#</a></h2>
<p>A version of this chapter will be published in SIAM Review in its Education section.  We are grateful that the SIAM publishers (of the Journal and of SIAM Books) have agreed to this.  There are some differences to the paper version, of course.  The paper version was submitted two years ago, and accepted eight months ago, and this unit has been edited a few times since then.  The main difference is that we have tried to make this unit address the student/reader more directly, while the paper version was addressed more to the instructor.</p>
<p>One of the main uses of this unit will be to provoke discussion on what it means to be <em>random</em>.  How you choose to handle this with your class is of course up to you, but we see several opportunities, chief among them to get the students to ask their <em>own</em> questions about randomness.</p>
<p>This module can be used to teach students about a useful visualization of sequences, in which case the emphasis from the instructor could be on interpreting the results, perhaps by using structural similarity or distance maps. The module can also be used to teach elementary programming techniques, in which case the instructor can emphasize programming tools, correctness, efficiency, style, or analysis.</p>
<p>Using elementary sequences of integers makes the module accessible even to first-year students, avoiding difficult biological details: but even so, students are exposed to the deep concepts of pattern and randomness more or less straight away.</p>
<p>Another thing is that much of this material was <em>new to us</em> when Lila Kari introduced the class to the notions, at a guest lecture.  We learned it at much the same time as the students did.  This worked well, and so we recommend that you let the students lead some of the discussions.  If you are in a similar situation to the one we were in, we think that you will enjoy this.  The students might be a little anxious about leading, but it can be made to work.  Of course you will want to read ahead and read some enrichment material; there is a lot.</p>
</section>
<section id="creating-fractals-using-iterated-function-systems-ifs">
<span id="sec-ifs"></span><h2>Creating Fractals using Iterated Function Systems (IFS)<a class="headerlink" href="#creating-fractals-using-iterated-function-systems-ifs" title="Permalink to this headline">#</a></h2>
<p>One random iteration algorithm for creating pictures of fractals uses the fixed attracting set of an <em>iterated function system</em> (IFS). Before defining IFS, we will first define some necessary preliminaries.</p>
<section id="affine-transformation-in-the-euclidean-plane">
<h3>Affine Transformation in the Euclidean Plane<a class="headerlink" href="#affine-transformation-in-the-euclidean-plane" title="Permalink to this headline">#</a></h3>
<p>We write a two dimensional <em>affine</em> transformation in the Euclidean plane <span class="math notranslate nohighlight">\(w: \mathbb{R}^{2} \to \mathbb{R}^{2}\)</span> in the form</p>
<div class="amsmath math notranslate nohighlight">
\[\begin{equation*}
	w(x_{1}, x_{2}) = \left(ax_{1} + bx_{2} + e, cx_{1} + dx_{2} + f \right)
\end{equation*}\]</div>
<p>where <span class="math notranslate nohighlight">\(a\)</span>, <span class="math notranslate nohighlight">\(b\)</span>, <span class="math notranslate nohighlight">\(c\)</span>, <span class="math notranslate nohighlight">\(d\)</span>, <span class="math notranslate nohighlight">\(e\)</span>, and <span class="math notranslate nohighlight">\(f\)</span> are real numbers <span id="id5">[<a class="reference internal" href="../references.html#id14" title="M. F. Barnsley. Fractals everywhere. Academic press, 2014.">Barnsley, 2014</a>]</span>. We will use the following more compact notation:</p>
<div class="math notranslate nohighlight" id="equation-eqn-ifs">
<span class="eqno">(71)<a class="headerlink" href="#equation-eqn-ifs" title="Permalink to this equation">#</a></span>\[\begin{split}
\begin{align}
	w(x) &amp;= 
	w
	\begin{pmatrix}
		x_{1} \\
		x_{2}
	\end{pmatrix}
	=
	\begin{bmatrix}
		a &amp; b \\
		c &amp; d
	\end{bmatrix}
	\begin{bmatrix}
		x_{1} \\
		x_{2}
	\end{bmatrix}
	+
	\begin{bmatrix}
		e \\
		f
	\end{bmatrix} \nonumber \\ 
	&amp;= Ax + t .
	\label{eqn:ifs} 
\end{align}
\end{split}\]</div>
<p>The matrix <span class="math notranslate nohighlight">\(A\)</span> is a <span class="math notranslate nohighlight">\(2 \times 2\)</span> real matrix and can always be written in the form</p>
<div class="math notranslate nohighlight">
\[\begin{split}
	\begin{bmatrix}
		a &amp; b \\
		c &amp; d
	\end{bmatrix}
	=
	\begin{bmatrix}
		r_{1}\cos\theta_{1} &amp; -r_{2}\sin\theta_{2} \\
		r_{1}\sin\theta_{1} &amp; \phantom{-}r_{2}\cos\theta_{2}
	\end{bmatrix} ,
	\label{eqn:tmatrix}
\end{split}\]</div>
<p>where <span class="math notranslate nohighlight">\(\left(r_{1}, \theta_{1}\right)\)</span> are the polar coordinates of the point <span class="math notranslate nohighlight">\(\left(a, c\right)\)</span> and <span class="math notranslate nohighlight">\(\left(r_{2}, \left(\theta_{2} + \frac{\pi}{2}\right)\right)\)</span> are the polar coordinates of the point <span class="math notranslate nohighlight">\(\left(b, d\right)\)</span>. An example of a <em>linear</em> transformation</p>
<div class="amsmath math notranslate nohighlight">
\[\begin{equation*}
	\begin{bmatrix}
		x_{1} \\
		x_{2}
	\end{bmatrix}
	\to
	A
	\begin{bmatrix}
		x_{1} \\
		x_{2}
	\end{bmatrix}
\end{equation*}\]</div>
<p>in <span class="math notranslate nohighlight">\(\mathbb{R}^{2}\)</span> is shown in in the following figure. We can see from this figure that the original shape remains as a parallelogram, but now has changed in size and in rotation. This is what we mean by <em>linear</em> transformation.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">p</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">]])</span>
<span class="n">t</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([[</span><span class="mi">2</span><span class="o">*</span><span class="n">math</span><span class="o">.</span><span class="n">cos</span><span class="p">(</span><span class="n">math</span><span class="o">.</span><span class="n">pi</span><span class="o">/</span><span class="mi">3</span><span class="p">),</span> <span class="o">-</span><span class="mi">3</span><span class="o">*</span><span class="n">math</span><span class="o">.</span><span class="n">sin</span><span class="p">(</span><span class="n">math</span><span class="o">.</span><span class="n">pi</span><span class="o">/</span><span class="mi">6</span><span class="p">)],</span> <span class="p">[</span><span class="mi">2</span><span class="o">*</span><span class="n">math</span><span class="o">.</span><span class="n">sin</span><span class="p">(</span><span class="n">math</span><span class="o">.</span><span class="n">pi</span><span class="o">/</span><span class="mi">3</span><span class="p">),</span> <span class="mi">3</span><span class="o">*</span><span class="n">math</span><span class="o">.</span><span class="n">cos</span><span class="p">(</span><span class="n">math</span><span class="o">.</span><span class="n">pi</span><span class="o">/</span><span class="mi">6</span><span class="p">)]])</span>
<span class="n">A</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">matmul</span><span class="p">(</span><span class="n">t</span><span class="p">,</span> <span class="n">p</span><span class="p">)</span>
<span class="n">fig</span><span class="p">,</span> <span class="p">(</span><span class="n">ax1</span><span class="p">,</span> <span class="n">ax2</span><span class="p">)</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">)</span>
<span class="n">fig</span><span class="o">.</span><span class="n">suptitle</span><span class="p">(</span><span class="s1">&#39;Affine Transformation Example&#39;</span><span class="p">)</span>
<span class="n">ax1</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">p</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="p">:],</span> <span class="n">p</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="p">:])</span>
<span class="n">ax1</span><span class="o">.</span><span class="n">set_aspect</span><span class="p">(</span><span class="s1">&#39;equal&#39;</span><span class="p">,</span> <span class="s1">&#39;box&#39;</span><span class="p">)</span>
<span class="n">ax2</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">A</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="p">:],</span> <span class="n">A</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="p">:])</span>
<span class="n">ax2</span><span class="o">.</span><span class="n">set_aspect</span><span class="p">(</span><span class="s1">&#39;equal&#39;</span><span class="p">,</span> <span class="s1">&#39;box&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/chaos-game-representation_8_0.png" src="../_images/chaos-game-representation_8_0.png" />
</div>
</div>
</section>
<section id="iterated-function-system">
<h3>Iterated Function System<a class="headerlink" href="#iterated-function-system" title="Permalink to this headline">#</a></h3>
<p>An <em>iterated function system</em> (IFS)<a class="footnote-reference brackets" href="#id65" id="id6">1</a> is a finite set of <a class="reference external" href="https://en.wikipedia.org/wiki/Contraction_mapping">contraction mappings on a metric space</a>.  That is, each mapping shrinks distances between points, so that the distance between <span class="math notranslate nohighlight">\(f(x)\)</span> and <span class="math notranslate nohighlight">\(f(y)\)</span> is smaller than the distance between <span class="math notranslate nohighlight">\(x\)</span> and <span class="math notranslate nohighlight">\(y\)</span>.  These can be used to create pictures of fractals. These pictures essentially show what is called the attractor of the IFS. We will be defining what an attractor is in <a class="reference internal" href="#sec-chaos"><span class="std std-ref">Chaos Game</span></a>. In this paper, we will only be looking at the form where each contraction mapping is an affine transformation on the metric space <span class="math notranslate nohighlight">\(\mathbb{R}^2\)</span>.</p>
<p>We will illustrate the algorithms for an IFS as an example. Consider the following maps:</p>
<div class="math notranslate nohighlight" id="equation-eqn-sierpinski">
<span class="eqno">(72)<a class="headerlink" href="#equation-eqn-sierpinski" title="Permalink to this equation">#</a></span>\[\begin{split}
\begin{align}
	w_1(x, y) &amp;=
	\left[
		\begin{array}{cc}
			\frac{1}{2} &amp; 0 \\
			0 &amp; \frac{1}{2}
		\end{array}
	\right]
	\left[
		\begin{array}{c}
			x \\
			y
		\end{array}
	\right] 
	+
	\left[
		\begin{array}{c}
			0 \\
			0
		\end{array}
	\right]  = \begin{bmatrix} \frac{1}{2}\,x \\ \frac{1}{2}\,y \end{bmatrix}, \nonumber \\
	w_2(x, y) &amp;=
	\left[
		\begin{array}{cc}
			\frac{1}{2} &amp; 0 \\
			0 &amp; \frac{1}{2}
		\end{array}
	\right]
	\left[
		\begin{array}{c}
			x \\
			y
		\end{array}
	\right]
	+
	\left[
		\begin{array}{c}
			0 \\
			\frac{1}{2}
		\end{array}
	\right] = \begin{bmatrix} \frac{1}{2}\,x \\ \frac{1}{2}\,y + \frac12 \end{bmatrix}, \nonumber \\
	w_3(x, y) &amp;=
	\left[
		\begin{array}{cc}
			\frac{1}{2} &amp; 0 \\
			0 &amp; \frac{1}{2}
		\end{array}
	\right]
	\left[
		\begin{array}{c}
			x \\
			y
		\end{array}
	\right]
	+
	\left[
		\begin{array}{c}
			\frac{1}{2} \\
			\frac{1}{2}
		\end{array}
	\right] = \begin{bmatrix} \frac{1}{2}\,x + \frac12 \\ \frac{1}{2}\,y + \frac12 \end{bmatrix},
	\label{eqn:sierpinski}
\end{align}
\end{split}\]</div>
<p>each with the <em>probability factor</em> of <span class="math notranslate nohighlight">\(\frac{1}{3}\)</span>, representing that each of these three options is to be chosen in an equally-likely way. These values can also be displayed as a table, where the order of the coefficients <span class="math notranslate nohighlight">\(a\)</span> through <span class="math notranslate nohighlight">\(f\)</span> corresponds to the order presented in Equation <a class="reference internal" href="#equation-eqn-ifs">(71)</a> and <span class="math notranslate nohighlight">\(p\)</span> represents the probability factor. The corresponding table for Equation <a class="reference internal" href="#equation-eqn-sierpinski">(72)</a> can be found in <a class="reference internal" href="#ifs-sierpinski"><span class="std std-numref">Table 1</span></a>.</p>
<table class="colwidths-auto table" id="ifs-sierpinski">
<caption><span class="caption-number">Table 1 </span><span class="caption-text">IFS to produce a Sierpinski triangle</span><a class="headerlink" href="#ifs-sierpinski" title="Permalink to this table">#</a></caption>
<thead>
<tr class="row-odd"><th class="head"><p>w</p></th>
<th class="head"><p>a</p></th>
<th class="head"><p>b</p></th>
<th class="head"><p>c</p></th>
<th class="head"><p>d</p></th>
<th class="head"><p>e</p></th>
<th class="head"><p>f</p></th>
<th class="head"><p>p</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>1</p></td>
<td><p><span class="math notranslate nohighlight">\(\frac{1}{2}\)</span></p></td>
<td><p><span class="math notranslate nohighlight">\(0\)</span></p></td>
<td><p><span class="math notranslate nohighlight">\(0\)</span></p></td>
<td><p><span class="math notranslate nohighlight">\(\frac{1}{2}\)</span></p></td>
<td><p><span class="math notranslate nohighlight">\(0\)</span></p></td>
<td><p><span class="math notranslate nohighlight">\(0\)</span></p></td>
<td><p><span class="math notranslate nohighlight">\(\frac{1}{3}\)</span></p></td>
</tr>
<tr class="row-odd"><td><p>2</p></td>
<td><p><span class="math notranslate nohighlight">\(\frac{1}{2}\)</span></p></td>
<td><p><span class="math notranslate nohighlight">\(0\)</span></p></td>
<td><p><span class="math notranslate nohighlight">\(0\)</span></p></td>
<td><p><span class="math notranslate nohighlight">\(\frac{1}{2}\)</span></p></td>
<td><p><span class="math notranslate nohighlight">\(0\)</span></p></td>
<td><p><span class="math notranslate nohighlight">\(\frac{1}{2}\)</span></p></td>
<td><p><span class="math notranslate nohighlight">\(\frac{1}{3}\)</span></p></td>
</tr>
<tr class="row-even"><td><p>3</p></td>
<td><p><span class="math notranslate nohighlight">\(\frac{1}{2}\)</span></p></td>
<td><p><span class="math notranslate nohighlight">\(0\)</span></p></td>
<td><p><span class="math notranslate nohighlight">\(0\)</span></p></td>
<td><p><span class="math notranslate nohighlight">\(\frac{1}{2}\)</span></p></td>
<td><p><span class="math notranslate nohighlight">\(\frac{1}{2}\)</span></p></td>
<td><p><span class="math notranslate nohighlight">\(\frac{1}{2}\)</span></p></td>
<td><p><span class="math notranslate nohighlight">\(\frac{1}{3}\)</span></p></td>
</tr>
</tbody>
</table>
<p>To create a fractal picture using IFS, we first choose a starting point <span class="math notranslate nohighlight">\((x_0, y_0)\)</span>. We then <em>randomly</em> choose a map from our IFS and evaluate it at our starting point <span class="math notranslate nohighlight">\((x_0, y_0)\)</span> to get our next point <span class="math notranslate nohighlight">\((x_1, y_1)\)</span>. We do this again many times (determined by the user) until we see a pattern. The Python code shown below computes and plots 50000 points corresponding to the IFS from <a class="reference internal" href="#ifs-sierpinski"><span class="std std-numref">Table 1</span></a>. To choose which map to use for each iteration, we used a uniform random number generator <code class="docutils literal notranslate"><span class="pre">random.randint</span></code> to choose a number between (and including) 0 and 2. Each value corresponds to a map in our IFS. It is clear that the resulting figure shows a fractal, which appears to be<a class="footnote-reference brackets" href="#id67" id="id7">2</a> the Sierpinski triangle <span id="id8">[<a class="reference internal" href="../references.html#id14" title="M. F. Barnsley. Fractals everywhere. Academic press, 2014.">Barnsley, 2014</a>, <a class="reference internal" href="../references.html#id21" title="D. P. Feldman. Chaos and fractals: an elementary introduction. Oxford University Press, 2012.">Feldman, 2012</a>]</span>, where its three vertices are located at <span class="math notranslate nohighlight">\((0, 0)\)</span>, <span class="math notranslate nohighlight">\((0, 1)\)</span> and <span class="math notranslate nohighlight">\((1, 1)\)</span>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">n</span> <span class="o">=</span> <span class="mi">50000</span>
<span class="n">pts</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="mi">2</span><span class="p">,</span> <span class="n">n</span><span class="p">))</span>
<span class="n">rseq</span> <span class="o">=</span> <span class="p">[</span><span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">2</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">n</span><span class="o">-</span><span class="mi">1</span><span class="p">)]</span>
<span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">n</span><span class="p">):</span>
    <span class="k">if</span> <span class="n">rseq</span><span class="p">[</span><span class="n">i</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
        <span class="n">pts</span><span class="p">[:,</span> <span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([[</span><span class="mf">0.5</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mf">0.5</span><span class="p">]])</span><span class="o">.</span><span class="n">dot</span><span class="p">(</span><span class="n">pts</span><span class="p">[:,</span> <span class="n">i</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span> <span class="o">+</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">])</span>
    <span class="k">elif</span> <span class="n">rseq</span><span class="p">[</span><span class="n">i</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
        <span class="n">pts</span><span class="p">[:,</span> <span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([[</span><span class="mf">0.5</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mf">0.5</span><span class="p">]])</span><span class="o">.</span><span class="n">dot</span><span class="p">(</span><span class="n">pts</span><span class="p">[:,</span> <span class="n">i</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span> <span class="o">+</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span> <span class="mf">0.5</span><span class="p">])</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">pts</span><span class="p">[:,</span> <span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([[</span><span class="mf">0.5</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mf">0.5</span><span class="p">]])</span><span class="o">.</span><span class="n">dot</span><span class="p">(</span><span class="n">pts</span><span class="p">[:,</span> <span class="n">i</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span> <span class="o">+</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="mf">0.5</span><span class="p">,</span> <span class="mf">0.5</span><span class="p">])</span>
<span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">pts</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="p">:],</span> <span class="n">pts</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="p">:],</span> <span class="s1">&#39;k.&#39;</span><span class="p">,</span> <span class="n">markersize</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">gca</span><span class="p">()</span><span class="o">.</span><span class="n">set_aspect</span><span class="p">(</span><span class="s1">&#39;equal&#39;</span><span class="p">,</span> <span class="n">adjustable</span><span class="o">=</span><span class="s1">&#39;box&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/chaos-game-representation_11_0.png" src="../_images/chaos-game-representation_11_0.png" />
</div>
</div>
<p>Iterated function systems can be quite versatile; a set of several attractors can be used to create different shapes and even can resemble real-life objects. An example of this is Barnsley’s fern, which resembles the Black Spleenwort <span id="id9">[<a class="reference internal" href="../references.html#id14" title="M. F. Barnsley. Fractals everywhere. Academic press, 2014.">Barnsley, 2014</a>]</span>. The map is displayed in the following table:</p>
<table class="colwidths-auto table">
<thead>
<tr class="row-odd"><th class="head"><p>w</p></th>
<th class="head"><p>a</p></th>
<th class="head"><p>b</p></th>
<th class="head"><p>c</p></th>
<th class="head"><p>d</p></th>
<th class="head"><p>e</p></th>
<th class="head"><p>f</p></th>
<th class="head"><p>p</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>1</p></td>
<td><p>0.00</p></td>
<td><p>0.00</p></td>
<td><p>0.00</p></td>
<td><p>0.16</p></td>
<td><p>0.00</p></td>
<td><p>0.00</p></td>
<td><p>0.01</p></td>
</tr>
<tr class="row-odd"><td><p>2</p></td>
<td><p>0.85</p></td>
<td><p>0.04</p></td>
<td><p>-0.04</p></td>
<td><p>0.85</p></td>
<td><p>0.00</p></td>
<td><p>1.60</p></td>
<td><p>0.85</p></td>
</tr>
<tr class="row-even"><td><p>3</p></td>
<td><p>0.20</p></td>
<td><p>-0.26</p></td>
<td><p>0.23</p></td>
<td><p>0.22</p></td>
<td><p>0.00</p></td>
<td><p>1.60</p></td>
<td><p>0.07</p></td>
</tr>
<tr class="row-odd"><td><p>4</p></td>
<td><p>-0.15</p></td>
<td><p>0.28</p></td>
<td><p>0.26</p></td>
<td><p>0.24</p></td>
<td><p>0.00</p></td>
<td><p>0.44</p></td>
<td><p>0.07</p></td>
</tr>
</tbody>
</table>
<p>We can use Python to produce the plot of the IFS given above.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">n</span> <span class="o">=</span> <span class="mi">50000</span>

<span class="n">fern</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="mi">2</span><span class="p">,</span> <span class="n">n</span><span class="p">))</span>

<span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">n</span><span class="p">):</span>
    <span class="n">r</span> <span class="o">=</span> <span class="n">random</span><span class="o">.</span><span class="n">random</span><span class="p">()</span>
    <span class="k">if</span> <span class="n">r</span> <span class="o">&lt;=</span> <span class="mf">0.01</span><span class="p">:</span>
        <span class="n">fern</span><span class="p">[:,</span> <span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mf">0.16</span><span class="p">]])</span><span class="o">.</span><span class="n">dot</span><span class="p">(</span><span class="n">fern</span><span class="p">[:,</span> <span class="n">i</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
    <span class="k">elif</span> <span class="n">r</span> <span class="o">&lt;=</span> <span class="mf">0.86</span><span class="p">:</span>
        <span class="n">fern</span><span class="p">[:,</span> <span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([[</span><span class="mf">0.85</span><span class="p">,</span> <span class="mf">0.04</span><span class="p">],</span> <span class="p">[</span><span class="o">-</span><span class="mf">0.04</span><span class="p">,</span> <span class="mf">0.85</span><span class="p">]])</span><span class="o">.</span><span class="n">dot</span><span class="p">(</span><span class="n">fern</span><span class="p">[:,</span> <span class="n">i</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span> <span class="o">+</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span> <span class="mf">1.6</span><span class="p">])</span>
    <span class="k">elif</span> <span class="n">r</span> <span class="o">&lt;=</span> <span class="mf">0.93</span><span class="p">:</span>
        <span class="n">fern</span><span class="p">[:,</span> <span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([[</span><span class="mf">0.2</span><span class="p">,</span> <span class="o">-</span><span class="mf">0.26</span><span class="p">],</span> <span class="p">[</span><span class="mf">0.23</span><span class="p">,</span> <span class="mf">0.22</span><span class="p">]])</span><span class="o">.</span><span class="n">dot</span><span class="p">(</span><span class="n">fern</span><span class="p">[:,</span> <span class="n">i</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span> <span class="o">+</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span> <span class="mf">1.6</span><span class="p">])</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">fern</span><span class="p">[:,</span> <span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([[</span><span class="o">-</span><span class="mf">0.15</span><span class="p">,</span> <span class="mf">0.28</span><span class="p">],</span> <span class="p">[</span><span class="mf">0.26</span><span class="p">,</span> <span class="mf">0.24</span><span class="p">]])</span><span class="o">.</span><span class="n">dot</span><span class="p">(</span><span class="n">fern</span><span class="p">[:,</span> <span class="n">i</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span> <span class="o">+</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span> <span class="mf">0.44</span><span class="p">])</span>

<span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">fern</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="p">:],</span> <span class="n">fern</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="p">:],</span> <span class="s1">&#39;g.&#39;</span><span class="p">,</span> <span class="n">markersize</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/chaos-game-representation_14_0.png" src="../_images/chaos-game-representation_14_0.png" />
</div>
</div>
<p>This image resembles a “black spleenwort” (genus Asplenium), which you might want to look up in order to compare the mathematical image above to a real plant.</p>
<div class="tip admonition" id="cgractivity-1">
<p class="admonition-title">CGR Activity 1</p>
<p>Playing with the coefficients of the transformation functions will create mutant fern varieties. One example of a mutant has the following IFS</p>
<table class="colwidths-auto table">
<thead>
<tr class="row-odd"><th class="head"><p><strong>w</strong></p></th>
<th class="head"><p><strong>a</strong></p></th>
<th class="head"><p><strong>b</strong></p></th>
<th class="head"><p><strong>c</strong></p></th>
<th class="head"><p><strong>d</strong></p></th>
<th class="head"><p><strong>e</strong></p></th>
<th class="head"><p><strong>f</strong></p></th>
<th class="head"><p><strong>p</strong></p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>1</p></td>
<td><p>0.00</p></td>
<td><p>0.00</p></td>
<td><p>0.00</p></td>
<td><p>0.25</p></td>
<td><p>0.00</p></td>
<td><p>-0.4</p></td>
<td><p>0.02</p></td>
</tr>
<tr class="row-odd"><td><p>2</p></td>
<td><p>0.95</p></td>
<td><p>0.005</p></td>
<td><p>-0.005</p></td>
<td><p>0.93</p></td>
<td><p>-0.002</p></td>
<td><p>0.5</p></td>
<td><p>0.84</p></td>
</tr>
<tr class="row-even"><td><p>3</p></td>
<td><p>0.035</p></td>
<td><p>-0.2</p></td>
<td><p>0.16</p></td>
<td><p>0.04</p></td>
<td><p>-0.09</p></td>
<td><p>0.02</p></td>
<td><p>0.07</p></td>
</tr>
<tr class="row-odd"><td><p>4</p></td>
<td><p>-0.04</p></td>
<td><p>0.2</p></td>
<td><p>0.16</p></td>
<td><p>0.04</p></td>
<td><p>0.083</p></td>
<td><p>0.12</p></td>
<td><p>0.07</p></td>
</tr>
</tbody>
</table>
<p>a) Modify the given Barnsley fern code to create the mutant given by the IFS above. What are the differences between the original Barnsley fern and the mutant? <a class="reference internal" href="../Solutions/Reports%20on%20Chaos%20Game%20Activities.html#cgractivityreport-1a"><span class="std std-ref">[What happened when we did this]</span></a></p>
<p>b) Create your own Barnsley fern mutant by changing the coefficients. <a class="reference internal" href="../Solutions/Reports%20on%20Chaos%20Game%20Activities.html#cgractivityreport-1b"><span class="std std-ref">[What happened when we did this]</span></a></p>
</div>
<div class="tip admonition" id="cgractivity-2">
<p class="admonition-title">CGR Activity 2</p>
<p><strong>Activity 2:</strong> <a class="reference external" href="http://paulbourke.net/fractals/ifs/">Paul Bourke’s website</a> provides several images produced by IFS. Replicate some of these examples using Python. <a class="reference internal" href="../Solutions/Reports%20on%20Chaos%20Game%20Activities.html#cgractivityreport-2"><span class="std std-ref">[What happened when we did this]</span></a></p>
</div>
</section>
</section>
<section id="chaos-game">
<span id="sec-chaos"></span><h2>Chaos Game<a class="headerlink" href="#chaos-game" title="Permalink to this headline">#</a></h2>
<p>CGR is based on a technique from chaotic dynamics called the “chaos game”<a class="footnote-reference brackets" href="#id69" id="id10">3</a>, popularized by Barnsley <span id="id11">[<a class="reference internal" href="../references.html#id14" title="M. F. Barnsley. Fractals everywhere. Academic press, 2014.">Barnsley, 2014</a>]</span> in 1988. The chaos game is an algorithm which allows one to produce pictures of fractal structures <span id="id12">[<a class="reference internal" href="../references.html#id27" title="H. J. Jeffrey. Chaos game representation of gene structure. Nucleic Acids Research, 18(8):2163–2170, 1990.">Jeffrey, 1990</a>]</span>. In its simplest form, this can be demonstrated with a piece of paper and pencil using the following steps (adapted from <span id="id13">[<a class="reference internal" href="../references.html#id27" title="H. J. Jeffrey. Chaos game representation of gene structure. Nucleic Acids Research, 18(8):2163–2170, 1990.">Jeffrey, 1990</a>]</span>):</p>
<ol class="simple">
<li><p>On a piece of paper, draw three points spaced out across the page; they do not necessarily need to be spaced out evenly to form an equilateral triangle though they might be.  These will determine the “<em>gameboard</em>”, where the picture will ultimately lie.</p></li>
<li><p>Label one of the points with “1, 2”, the other point with “3, 4”, and the last point with “5, 6”.</p></li>
<li><p>Draw a point anywhere on the page: this will be the starting point.</p></li>
<li><p>Roll a six-sided die. Draw an imaginary line from the current dot towards to the point with the corresponding number, and put a dot half way on the line. This becomes the new “current dot”.</p></li>
<li><p>Repeat step 4 (until you get bored).</p></li>
</ol>
<p>Obviously, one would not want to plot a large number of points by hand<a class="footnote-reference brackets" href="#id70" id="id14">4</a>; instead, this can be done by computer; the code is shown below. For this algorithm, we assigned <span class="math notranslate nohighlight">\((0, 0)\)</span>, <span class="math notranslate nohighlight">\((0, 1)\)</span> and <span class="math notranslate nohighlight">\((1, 1)\)</span> as the vertices of the triangle; they each are labelled with the value <span class="math notranslate nohighlight">\(0\)</span>, <span class="math notranslate nohighlight">\(1\)</span>, and <span class="math notranslate nohighlight">\(2\)</span>, respectively. Instead of using a die to decide which vertex to choose, similarly to the above code, we use the built-in function <code class="docutils literal notranslate"><span class="pre">random.randint</span></code>, which picks from the set <span class="math notranslate nohighlight">\(\{0, 1, 2\}\)</span>.</p>
<p>To calculate the new point, we take the average between the previous point and the corresponding vertex. For example, if the first random number generated was <span class="math notranslate nohighlight">\(2\)</span>, then we would take the average between our initial point (0, 0) and the vertex that corresponds to <span class="math notranslate nohighlight">\(2\)</span>, which is in our case is <span class="math notranslate nohighlight">\((1, 1)\)</span>. Our new point would be <span class="math notranslate nohighlight">\((0.5, 0.5)\)</span>. We then would repeat this process until one decides to stop.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>
<span class="kn">import</span> <span class="nn">random</span>

<span class="n">n</span> <span class="o">=</span> <span class="mi">50000</span>
<span class="n">pts</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="mi">2</span><span class="p">,</span> <span class="n">n</span><span class="p">))</span>
<span class="n">rseq</span> <span class="o">=</span> <span class="p">[</span><span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">2</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">n</span><span class="o">-</span><span class="mi">1</span><span class="p">)]</span>
<span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">n</span><span class="p">):</span>
    <span class="k">if</span> <span class="n">rseq</span><span class="p">[</span><span class="n">i</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
        <span class="n">pts</span><span class="p">[:,</span> <span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="mf">0.5</span><span class="o">*</span><span class="p">(</span><span class="n">pts</span><span class="p">[:,</span> <span class="n">i</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">+</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">]))</span> 
    <span class="k">elif</span> <span class="n">rseq</span><span class="p">[</span><span class="n">i</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
        <span class="n">pts</span><span class="p">[:,</span> <span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="mf">0.5</span><span class="o">*</span><span class="p">(</span><span class="n">pts</span><span class="p">[:,</span> <span class="n">i</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">+</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">]))</span> 
    <span class="k">else</span><span class="p">:</span>
        <span class="n">pts</span><span class="p">[:,</span> <span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="mf">0.5</span><span class="o">*</span><span class="p">(</span><span class="n">pts</span><span class="p">[:,</span> <span class="n">i</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">+</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">]))</span>
        
<span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">pts</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="p">:],</span> <span class="n">pts</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="p">:],</span> <span class="s1">&#39;k.&#39;</span><span class="p">,</span> <span class="n">markersize</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">gca</span><span class="p">()</span><span class="o">.</span><span class="n">set_aspect</span><span class="p">(</span><span class="s1">&#39;equal&#39;</span><span class="p">,</span> <span class="n">adjustable</span><span class="o">=</span><span class="s1">&#39;box&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/chaos-game-representation_19_0.png" src="../_images/chaos-game-representation_19_0.png" />
</div>
</div>
<p>One would expect that the outcome above would show that the dots would appear randomly across the gameboard; however, this is not the case: the figure shows the result of the chaos game if the game was played 50000 times. We can see that this figure shows a fractal-like pattern (such as one of a Sierpinski triangle), and looks almost (if not) identical to the figure which resulted from the IFS.</p>
<p>It is quite interesting that a sequence of random numbers could produce such a distinct pattern. Indeed, some bad <em>pseudorandom number generators</em> were identified as being bad precisely because they failed a similar-in-spirit visualization test <span id="id15">[<a class="reference internal" href="../references.html#id35" title="D. E. Knuth. Seminumerical Algorithms. Volume 2 of The Art of Computer Programming. Addison Wesley, Reading, MA, 3 edition, 1997.">Knuth, 1997</a>]</span>. We will discuss this issue further in <a class="reference internal" href="#sec-random"><span class="std std-ref">Random vs Pseudorandom</span></a>. However, what has happened here is that the <em>randomness</em> of the sequence has translated into a <em>uniform</em> distribution <em>on the fractal</em>. So, really, here the gameboard is the fractal! We will return to this point in the next section.</p>
<p><a class="reference internal" href="#sierpinski-steps"><span class="std std-numref">Fig. 9</span></a> shows an example of the first five steps of the chaos game (overlaid on top of a CGR with 10000 points), which corresponds to the following values that were randomly generated: 2, 1, 0, 0, 2. Displaying the CGR in this manner can give us some insight as to why this particular chaos game with three vertices gives us such a striking fractal: we can see from the figure that each of the points from the first five steps is located at a vertex (in this case the right-angled one) of one of the many triangles within the fractal. A careful observer would notice that the triangle at which the point is located becomes smaller each iteration. This implies that within the next few iterations, the triangle associated with the computed point would be too small to be seen in this illustration. [Note: if we zoom into that microscopic level, the dots would look random. Indeed, the random dots are ultimately dense on the Sierpinski triangle.] This sequence of points generated by the chaos game is called the <em>orbit</em> of the seed and it is attracted to the Sierpinski triangle. For another description of why this chaos game creates the Sierpinski triangle, see <span id="id16">[<a class="reference internal" href="../references.html#id23" title="Y. l. Fisher, M. McGuire, R. F. Voss, M. F. Barnsley, R. L. Devaney, and B. B. Mandelbrot. The science of fractal images. Springer Science &amp; Business Media, 2012.">Fisher <em>et al.</em>, 2012</a>]</span>.</p>
<figure class="align-default" id="sierpinski-steps">
<a class="reference internal image-reference" href="../_images/sierpinski_steps.png"><img alt="../_images/sierpinski_steps.png" src="../_images/sierpinski_steps.png" style="height: 300px;" /></a>
<figcaption>
<p><span class="caption-number">Fig. 9 </span><span class="caption-text">First 5 steps of chaos game with three vertices. In this example, the first 5 random integers generated are 2, 1, 0, 0, 2.</span><a class="headerlink" href="#sierpinski-steps" title="Permalink to this image">#</a></p>
</figcaption>
</figure>
<section id="generalization-of-cgr">
<h3>Generalization of CGR<a class="headerlink" href="#generalization-of-cgr" title="Permalink to this headline">#</a></h3>
<p>We have already seen that a fractal pattern can appear when we follow the chaos game described above for a polygon with three vertices (triangle) using a random number generator, but we are not limited to only this. There are many variations for CGR, and some are more useful than others. In this section, we will look into what patterns arise when we perform CGR on different polygons, changing the placement of the <span class="math notranslate nohighlight">\(i\)</span>-th point relative to the <span class="math notranslate nohighlight">\((i-1)\)</span>-th and its corresponding vertex, and explore which of these could potentially be useful for some applications presented in <a class="reference internal" href="#sec-dna"><span class="std std-ref">CGR of Biological Sequences</span></a>.</p>
<section id="different-polygons">
<span id="sec-polygons"></span><h4>Different Polygons<a class="headerlink" href="#different-polygons" title="Permalink to this headline">#</a></h4>
<p>The figure created from the following code shows the chaos game representation, where the <span class="math notranslate nohighlight">\(i\)</span>-th point is placed halfway between the <span class="math notranslate nohighlight">\((i-1)\)</span>-th point and the vertex corresponding to the <span class="math notranslate nohighlight">\(i\)</span>-th base, for various polygons, using a sequence of 50,000 uniformly generated (pseudo)random integers. As we can see below, when the CGR is a square, there is no apparent pattern: the chaos game produced a square uniformly (to the eye) filled with points. This particularly conforms to our intuition: a random sequence should <em>look</em> random.  The other polygons appear, in contrast, to exhibit some patterns. These patterns are due to several parts of their attractors overlapping<a class="footnote-reference brackets" href="#id71" id="id17">5</a>, or failing to overlap at all, indicated by the the uneven distribution of points (where points do appear); i.e. we can see some darker areas and some lighter areas and even gaps in the figures. Therefore, for this particular chaos game (in which we say that the <em>dividing rate</em> <span class="math notranslate nohighlight">\(r\)</span> is <span class="math notranslate nohighlight">\(0.5\)</span>—we will talk about this in <a class="reference internal" href="#sec-dividingrate"><span class="std std-ref">Dividing Rate</span></a>), the square is the best choice to visualize one-dimensional sequences. We will see in <a class="reference internal" href="#sec-generalization"><span class="std std-ref">Generalization of CGR</span></a> that we can use other polygons (we will be referring to them as <span class="math notranslate nohighlight">\(m\)</span>-gons) for CGR to uncover patterns in sequences not with 4 elements by using the appropriate dividing rate <span class="math notranslate nohighlight">\(r\)</span>: that is, instead of placing a point half-way to the target vertex, we will place it at some other position along the line, with <span class="math notranslate nohighlight">\(0 &lt; r &lt; 1\)</span>.  A proper choice of <span class="math notranslate nohighlight">\(r\)</span> will ensure that a random sequence will generate a uniform density of points, possibly on a fractal.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">n_gon</span><span class="p">(</span><span class="n">n</span><span class="p">,</span> <span class="o">*</span><span class="n">start</span><span class="p">):</span>
    <span class="k">if</span> <span class="n">start</span><span class="p">:</span>
        <span class="n">deg</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">linspace</span><span class="p">(</span><span class="n">start</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">start</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="mi">360</span><span class="p">,</span> <span class="n">n</span><span class="o">+</span><span class="mi">1</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">deg</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">linspace</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">360</span><span class="p">,</span> <span class="n">n</span><span class="o">+</span><span class="mi">1</span><span class="p">)</span>
    <span class="n">deg</span> <span class="o">=</span> <span class="n">deg</span><span class="p">[:</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
    <span class="n">rad</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">d</span> <span class="ow">in</span> <span class="n">deg</span><span class="p">:</span>
        <span class="n">rad</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">d</span><span class="o">/</span><span class="mi">360</span><span class="p">)</span><span class="o">*</span><span class="mi">2</span><span class="o">*</span><span class="n">math</span><span class="o">.</span><span class="n">pi</span><span class="p">)</span>
    <span class="n">cor</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="mi">2</span><span class="p">,</span> <span class="n">n</span><span class="o">+</span><span class="mi">1</span><span class="p">))</span>
    <span class="k">for</span> <span class="n">r</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">rad</span><span class="p">)):</span>
        <span class="n">x</span> <span class="o">=</span> <span class="n">math</span><span class="o">.</span><span class="n">cos</span><span class="p">(</span><span class="n">rad</span><span class="p">[</span><span class="n">r</span><span class="p">])</span>
        <span class="n">y</span> <span class="o">=</span> <span class="n">math</span><span class="o">.</span><span class="n">sin</span><span class="p">(</span><span class="n">rad</span><span class="p">[</span><span class="n">r</span><span class="p">])</span>
        <span class="n">cor</span><span class="p">[:,</span> <span class="n">r</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="n">x</span><span class="p">,</span> <span class="n">y</span><span class="p">])</span>
    <span class="n">cor</span><span class="p">[:,</span> <span class="n">n</span><span class="p">]</span> <span class="o">=</span> <span class="n">cor</span><span class="p">[:,</span> <span class="mi">0</span><span class="p">]</span>
    <span class="k">return</span> <span class="n">cor</span>

<span class="k">def</span> <span class="nf">polyCGR</span><span class="p">(</span><span class="n">N</span><span class="p">,</span> <span class="n">n</span><span class="p">,</span> <span class="o">*</span><span class="n">start</span><span class="p">):</span>
    <span class="k">if</span> <span class="n">start</span><span class="p">:</span>
        <span class="n">shape_coord</span> <span class="o">=</span> <span class="n">n_gon</span><span class="p">(</span><span class="n">n</span><span class="p">,</span> <span class="n">start</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">shape_coord</span> <span class="o">=</span> <span class="n">n_gon</span><span class="p">(</span><span class="n">n</span><span class="p">)</span>
    <span class="n">dataPoints</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="mi">2</span><span class="p">,</span> <span class="n">N</span><span class="p">))</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">N</span><span class="p">):</span>
        <span class="n">r</span> <span class="o">=</span> <span class="n">random</span><span class="o">.</span><span class="n">random</span><span class="p">()</span>
        <span class="n">dataPoints</span><span class="p">[:,</span> <span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="mf">0.5</span><span class="o">*</span><span class="p">(</span><span class="n">dataPoints</span><span class="p">[:,</span> <span class="n">i</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">+</span> <span class="n">shape_coord</span><span class="p">[:,</span> <span class="n">math</span><span class="o">.</span><span class="n">floor</span><span class="p">(</span><span class="n">r</span><span class="o">*</span><span class="n">n</span><span class="p">)])</span>
    <span class="k">return</span><span class="p">(</span><span class="n">shape_coord</span><span class="p">,</span> <span class="n">dataPoints</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">N</span> <span class="o">=</span> <span class="mi">10000</span>

<span class="n">fig</span><span class="p">,</span> <span class="p">((</span><span class="n">ax1</span><span class="p">,</span> <span class="n">ax2</span><span class="p">,</span> <span class="n">ax3</span><span class="p">),</span> <span class="p">(</span><span class="n">ax4</span><span class="p">,</span> <span class="n">ax5</span><span class="p">,</span> <span class="n">ax6</span><span class="p">))</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">12</span><span class="p">,</span> <span class="mi">8</span><span class="p">))</span>
<span class="n">fig</span><span class="o">.</span><span class="n">suptitle</span><span class="p">(</span><span class="s1">&#39;Chaos game of polygons with different number of vertices&#39;</span><span class="p">)</span>

<span class="p">(</span><span class="n">coord</span><span class="p">,</span> <span class="n">points</span><span class="p">)</span> <span class="o">=</span> <span class="n">polyCGR</span><span class="p">(</span><span class="n">N</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">45</span><span class="p">)</span>
<span class="n">ax1</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">coord</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="p">:],</span> <span class="n">coord</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="p">:],</span> <span class="s1">&#39;k&#39;</span><span class="p">)</span>
<span class="n">ax1</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">points</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="p">:],</span> <span class="n">points</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="p">:],</span> <span class="s1">&#39;k.&#39;</span><span class="p">,</span> <span class="n">markersize</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="n">ax1</span><span class="o">.</span><span class="n">axis</span><span class="p">(</span><span class="s1">&#39;off&#39;</span><span class="p">)</span>
<span class="n">ax1</span><span class="o">.</span><span class="n">set_aspect</span><span class="p">(</span><span class="s1">&#39;equal&#39;</span><span class="p">,</span> <span class="s1">&#39;box&#39;</span><span class="p">)</span>
<span class="n">ax1</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;Square&#39;</span><span class="p">)</span>

<span class="p">(</span><span class="n">coord</span><span class="p">,</span> <span class="n">points</span><span class="p">)</span> <span class="o">=</span> <span class="n">polyCGR</span><span class="p">(</span><span class="n">N</span><span class="p">,</span> <span class="mi">5</span><span class="p">,</span> <span class="mi">90</span><span class="p">)</span>
<span class="n">ax2</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">coord</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="p">:],</span> <span class="n">coord</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="p">:],</span> <span class="s1">&#39;k&#39;</span><span class="p">)</span>
<span class="n">ax2</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">points</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="p">:],</span> <span class="n">points</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="p">:],</span> <span class="s1">&#39;k.&#39;</span><span class="p">,</span> <span class="n">markersize</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="n">ax2</span><span class="o">.</span><span class="n">axis</span><span class="p">(</span><span class="s1">&#39;off&#39;</span><span class="p">)</span>
<span class="n">ax2</span><span class="o">.</span><span class="n">set_aspect</span><span class="p">(</span><span class="s1">&#39;equal&#39;</span><span class="p">,</span> <span class="s1">&#39;box&#39;</span><span class="p">)</span>
<span class="n">ax2</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;Pentagon&#39;</span><span class="p">)</span>

<span class="p">(</span><span class="n">coord</span><span class="p">,</span> <span class="n">points</span><span class="p">)</span> <span class="o">=</span> <span class="n">polyCGR</span><span class="p">(</span><span class="n">N</span><span class="p">,</span> <span class="mi">6</span><span class="p">)</span>
<span class="n">ax3</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">coord</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="p">:],</span> <span class="n">coord</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="p">:],</span> <span class="s1">&#39;k&#39;</span><span class="p">)</span>
<span class="n">ax3</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">points</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="p">:],</span> <span class="n">points</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="p">:],</span> <span class="s1">&#39;k.&#39;</span><span class="p">,</span> <span class="n">markersize</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="n">ax3</span><span class="o">.</span><span class="n">axis</span><span class="p">(</span><span class="s1">&#39;off&#39;</span><span class="p">)</span>
<span class="n">ax3</span><span class="o">.</span><span class="n">set_aspect</span><span class="p">(</span><span class="s1">&#39;equal&#39;</span><span class="p">,</span> <span class="s1">&#39;box&#39;</span><span class="p">)</span>
<span class="n">ax3</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;Hexagon&#39;</span><span class="p">)</span>

<span class="p">(</span><span class="n">coord</span><span class="p">,</span> <span class="n">points</span><span class="p">)</span> <span class="o">=</span> <span class="n">polyCGR</span><span class="p">(</span><span class="n">N</span><span class="p">,</span> <span class="mi">7</span><span class="p">,</span> <span class="mi">90</span><span class="p">)</span>
<span class="n">ax4</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">coord</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="p">:],</span> <span class="n">coord</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="p">:],</span> <span class="s1">&#39;k&#39;</span><span class="p">)</span>
<span class="n">ax4</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">points</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="p">:],</span> <span class="n">points</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="p">:],</span> <span class="s1">&#39;k.&#39;</span><span class="p">,</span> <span class="n">markersize</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="n">ax4</span><span class="o">.</span><span class="n">axis</span><span class="p">(</span><span class="s1">&#39;off&#39;</span><span class="p">)</span>
<span class="n">ax4</span><span class="o">.</span><span class="n">set_aspect</span><span class="p">(</span><span class="s1">&#39;equal&#39;</span><span class="p">,</span> <span class="s1">&#39;box&#39;</span><span class="p">)</span>
<span class="n">ax4</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;Heptagon&#39;</span><span class="p">)</span>

<span class="p">(</span><span class="n">coord</span><span class="p">,</span> <span class="n">points</span><span class="p">)</span> <span class="o">=</span> <span class="n">polyCGR</span><span class="p">(</span><span class="n">N</span><span class="p">,</span> <span class="mi">8</span><span class="p">)</span>
<span class="n">ax5</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">coord</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="p">:],</span> <span class="n">coord</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="p">:],</span> <span class="s1">&#39;k&#39;</span><span class="p">)</span>
<span class="n">ax5</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">points</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="p">:],</span> <span class="n">points</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="p">:],</span> <span class="s1">&#39;k.&#39;</span><span class="p">,</span> <span class="n">markersize</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="n">ax5</span><span class="o">.</span><span class="n">axis</span><span class="p">(</span><span class="s1">&#39;off&#39;</span><span class="p">)</span>
<span class="n">ax5</span><span class="o">.</span><span class="n">set_aspect</span><span class="p">(</span><span class="s1">&#39;equal&#39;</span><span class="p">,</span> <span class="s1">&#39;box&#39;</span><span class="p">)</span>
<span class="n">ax5</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;Octagon&#39;</span><span class="p">)</span>

<span class="p">(</span><span class="n">coord</span><span class="p">,</span> <span class="n">points</span><span class="p">)</span> <span class="o">=</span> <span class="n">polyCGR</span><span class="p">(</span><span class="n">N</span><span class="p">,</span> <span class="mi">9</span><span class="p">,</span> <span class="mi">90</span><span class="p">)</span>
<span class="n">ax6</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">coord</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="p">:],</span> <span class="n">coord</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="p">:],</span> <span class="s1">&#39;k&#39;</span><span class="p">)</span>
<span class="n">ax6</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">points</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="p">:],</span> <span class="n">points</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="p">:],</span> <span class="s1">&#39;k.&#39;</span><span class="p">,</span> <span class="n">markersize</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="n">ax6</span><span class="o">.</span><span class="n">axis</span><span class="p">(</span><span class="s1">&#39;off&#39;</span><span class="p">)</span>
<span class="n">ax6</span><span class="o">.</span><span class="n">set_aspect</span><span class="p">(</span><span class="s1">&#39;equal&#39;</span><span class="p">,</span> <span class="s1">&#39;box&#39;</span><span class="p">)</span>
<span class="n">ax6</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;Nonagon&#39;</span><span class="p">)</span>

<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/chaos-game-representation_25_0.png" src="../_images/chaos-game-representation_25_0.png" />
</div>
</div>
</section>
<section id="dividing-rate">
<span id="sec-dividingrate"></span><h4>Dividing Rate<a class="headerlink" href="#dividing-rate" title="Permalink to this headline">#</a></h4>
<p>In the previous subsection, we saw different patterns appear for different shapes. We saw that shapes with more than <span class="math notranslate nohighlight">\(4\)</span> vertices seemed to have several parts of the attractor overlapping, producing visibly nonuniform density (some areas were visibly darker than others). Here we will show that this attractor overlapping also happens when choosing a <em>small enough</em> dividing rate <span class="math notranslate nohighlight">\(r\)</span>.</p>
<p>To be explicit, the chaos game is not restricted to the “rule” in which the new point has to be placed halfway between the current point and the corresponding vertex as depicted in the previous examples. The new point can be placed anywhere within the line segment created by the two points of reference (or even outside, if you are willing to let the points land outside the convex hull of the vertices). The placement of the new point affects how the attractor of the CGR looks. We will see this in the following example. To quantify the placement of the point, we take the proportion of which the distance between the new point and current point is from the distance between the current point and the corresponding vertex. This is called the <em>dividing rate</em>, <span class="math notranslate nohighlight">\(r\)</span>. The new point <span class="math notranslate nohighlight">\(p_{i+1}\)</span> is thus <span class="math notranslate nohighlight">\(p_{i+1} = p_i + r(V-p_i)\)</span>. Therefore, when the new point is placed halfway between the current point and the vertex, the dividing rate is <span class="math notranslate nohighlight">\(r = 0.5\)</span>, and if the new point is placed closer to the location of the current point, then the dividing rate <span class="math notranslate nohighlight">\(r &lt; 0.5\)</span>, whereas if the point is closer to the vertex, then the dividing rate <span class="math notranslate nohighlight">\(r &gt; 0.5\)</span>.</p>
<p>To continue this discussion, let us vary <span class="math notranslate nohighlight">\(r\)</span> for the square-shaped CGR of a sequence of random integers. As we have seen in the previous subsection, the square-shaped CGR did not exhibit any fractal patterns when <span class="math notranslate nohighlight">\(r = 0.5\)</span>, so we wondered if we varied <span class="math notranslate nohighlight">\(r\)</span>, would this change? The following code gives the square CGRs for <span class="math notranslate nohighlight">\(r = 0.1, 0.25, 0.4, 0.6, 0.75,\)</span> and <span class="math notranslate nohighlight">\(0.9\)</span>. From these figures, we see how <span class="math notranslate nohighlight">\(r\)</span> affects the visualization.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">dividingRateCGR</span><span class="p">(</span><span class="n">N</span><span class="p">,</span> <span class="n">r</span><span class="p">):</span>
    <span class="n">shape_coord</span> <span class="o">=</span> <span class="n">n_gon</span><span class="p">(</span><span class="mi">4</span><span class="p">,</span> <span class="mi">45</span><span class="p">)</span>
    <span class="n">dataPoints</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="mi">2</span><span class="p">,</span> <span class="n">N</span><span class="p">))</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">N</span><span class="p">):</span>
        <span class="n">rand</span> <span class="o">=</span> <span class="n">random</span><span class="o">.</span><span class="n">random</span><span class="p">()</span>
        <span class="n">dataPoints</span><span class="p">[:,</span> <span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="n">dataPoints</span><span class="p">[:,</span> <span class="n">i</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">+</span> <span class="p">(</span><span class="n">shape_coord</span><span class="p">[:,</span> <span class="n">math</span><span class="o">.</span><span class="n">floor</span><span class="p">(</span><span class="mi">4</span><span class="o">*</span><span class="n">rand</span><span class="p">)]</span> <span class="o">-</span> <span class="n">dataPoints</span><span class="p">[:,</span> <span class="n">i</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span><span class="o">*</span><span class="n">r</span>
    <span class="k">return</span><span class="p">(</span><span class="n">shape_coord</span><span class="p">,</span> <span class="n">dataPoints</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">N</span> <span class="o">=</span> <span class="mi">10000</span>

<span class="n">fig</span><span class="p">,</span> <span class="p">((</span><span class="n">ax1</span><span class="p">,</span> <span class="n">ax2</span><span class="p">,</span> <span class="n">ax3</span><span class="p">),</span> <span class="p">(</span><span class="n">ax4</span><span class="p">,</span> <span class="n">ax5</span><span class="p">,</span> <span class="n">ax6</span><span class="p">))</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">12</span><span class="p">,</span> <span class="mi">8</span><span class="p">))</span>
<span class="n">fig</span><span class="o">.</span><span class="n">suptitle</span><span class="p">(</span><span class="s1">&#39;Illustration of different dividing rates on a square CGR&#39;</span><span class="p">)</span>

<span class="p">(</span><span class="n">coord</span><span class="p">,</span> <span class="n">points</span><span class="p">)</span> <span class="o">=</span> <span class="n">dividingRateCGR</span><span class="p">(</span><span class="n">N</span><span class="p">,</span> <span class="mf">0.1</span><span class="p">)</span>
<span class="n">ax1</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">coord</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="p">:],</span> <span class="n">coord</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="p">:],</span> <span class="s1">&#39;k&#39;</span><span class="p">)</span>
<span class="n">ax1</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">points</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="p">:],</span> <span class="n">points</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="p">:],</span> <span class="s1">&#39;k.&#39;</span><span class="p">,</span> <span class="n">markersize</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="n">ax1</span><span class="o">.</span><span class="n">axis</span><span class="p">(</span><span class="s1">&#39;off&#39;</span><span class="p">)</span>
<span class="n">ax1</span><span class="o">.</span><span class="n">set_aspect</span><span class="p">(</span><span class="s1">&#39;equal&#39;</span><span class="p">,</span> <span class="s1">&#39;box&#39;</span><span class="p">)</span>
<span class="n">ax1</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;r = 0.1&#39;</span><span class="p">)</span>

<span class="p">(</span><span class="n">coord</span><span class="p">,</span> <span class="n">points</span><span class="p">)</span> <span class="o">=</span> <span class="n">dividingRateCGR</span><span class="p">(</span><span class="n">N</span><span class="p">,</span> <span class="mf">0.25</span><span class="p">)</span>
<span class="n">ax2</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">coord</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="p">:],</span> <span class="n">coord</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="p">:],</span> <span class="s1">&#39;k&#39;</span><span class="p">)</span>
<span class="n">ax2</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">points</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="p">:],</span> <span class="n">points</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="p">:],</span> <span class="s1">&#39;k.&#39;</span><span class="p">,</span> <span class="n">markersize</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="n">ax2</span><span class="o">.</span><span class="n">axis</span><span class="p">(</span><span class="s1">&#39;off&#39;</span><span class="p">)</span>
<span class="n">ax2</span><span class="o">.</span><span class="n">set_aspect</span><span class="p">(</span><span class="s1">&#39;equal&#39;</span><span class="p">,</span> <span class="s1">&#39;box&#39;</span><span class="p">)</span>
<span class="n">ax2</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;r = 0.25&#39;</span><span class="p">)</span>

<span class="p">(</span><span class="n">coord</span><span class="p">,</span> <span class="n">points</span><span class="p">)</span> <span class="o">=</span> <span class="n">dividingRateCGR</span><span class="p">(</span><span class="n">N</span><span class="p">,</span> <span class="mf">0.4</span><span class="p">)</span>
<span class="n">ax3</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">coord</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="p">:],</span> <span class="n">coord</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="p">:],</span> <span class="s1">&#39;k&#39;</span><span class="p">)</span>
<span class="n">ax3</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">points</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="p">:],</span> <span class="n">points</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="p">:],</span> <span class="s1">&#39;k.&#39;</span><span class="p">,</span> <span class="n">markersize</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="n">ax3</span><span class="o">.</span><span class="n">axis</span><span class="p">(</span><span class="s1">&#39;off&#39;</span><span class="p">)</span>
<span class="n">ax3</span><span class="o">.</span><span class="n">set_aspect</span><span class="p">(</span><span class="s1">&#39;equal&#39;</span><span class="p">,</span> <span class="s1">&#39;box&#39;</span><span class="p">)</span>
<span class="n">ax3</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;r = 0.4&#39;</span><span class="p">)</span>

<span class="p">(</span><span class="n">coord</span><span class="p">,</span> <span class="n">points</span><span class="p">)</span> <span class="o">=</span> <span class="n">dividingRateCGR</span><span class="p">(</span><span class="n">N</span><span class="p">,</span> <span class="mf">0.6</span><span class="p">)</span>
<span class="n">ax4</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">coord</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="p">:],</span> <span class="n">coord</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="p">:],</span> <span class="s1">&#39;k&#39;</span><span class="p">)</span>
<span class="n">ax4</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">points</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="p">:],</span> <span class="n">points</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="p">:],</span> <span class="s1">&#39;k.&#39;</span><span class="p">,</span> <span class="n">markersize</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="n">ax4</span><span class="o">.</span><span class="n">axis</span><span class="p">(</span><span class="s1">&#39;off&#39;</span><span class="p">)</span>
<span class="n">ax4</span><span class="o">.</span><span class="n">set_aspect</span><span class="p">(</span><span class="s1">&#39;equal&#39;</span><span class="p">,</span> <span class="s1">&#39;box&#39;</span><span class="p">)</span>
<span class="n">ax4</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;r = 0.6&#39;</span><span class="p">)</span>

<span class="p">(</span><span class="n">coord</span><span class="p">,</span> <span class="n">points</span><span class="p">)</span> <span class="o">=</span> <span class="n">dividingRateCGR</span><span class="p">(</span><span class="n">N</span><span class="p">,</span> <span class="mf">0.75</span><span class="p">)</span>
<span class="n">ax5</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">coord</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="p">:],</span> <span class="n">coord</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="p">:],</span> <span class="s1">&#39;k&#39;</span><span class="p">)</span>
<span class="n">ax5</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">points</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="p">:],</span> <span class="n">points</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="p">:],</span> <span class="s1">&#39;k.&#39;</span><span class="p">,</span> <span class="n">markersize</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="n">ax5</span><span class="o">.</span><span class="n">axis</span><span class="p">(</span><span class="s1">&#39;off&#39;</span><span class="p">)</span>
<span class="n">ax5</span><span class="o">.</span><span class="n">set_aspect</span><span class="p">(</span><span class="s1">&#39;equal&#39;</span><span class="p">,</span> <span class="s1">&#39;box&#39;</span><span class="p">)</span>
<span class="n">ax5</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;r = 0.75&#39;</span><span class="p">)</span>

<span class="p">(</span><span class="n">coord</span><span class="p">,</span> <span class="n">points</span><span class="p">)</span> <span class="o">=</span> <span class="n">dividingRateCGR</span><span class="p">(</span><span class="n">N</span><span class="p">,</span> <span class="mf">0.9</span><span class="p">)</span>
<span class="n">ax6</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">coord</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="p">:],</span> <span class="n">coord</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="p">:],</span> <span class="s1">&#39;k&#39;</span><span class="p">)</span>
<span class="n">ax6</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">points</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="p">:],</span> <span class="n">points</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="p">:],</span> <span class="s1">&#39;k.&#39;</span><span class="p">,</span> <span class="n">markersize</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="n">ax6</span><span class="o">.</span><span class="n">axis</span><span class="p">(</span><span class="s1">&#39;off&#39;</span><span class="p">)</span>
<span class="n">ax6</span><span class="o">.</span><span class="n">set_aspect</span><span class="p">(</span><span class="s1">&#39;equal&#39;</span><span class="p">,</span> <span class="s1">&#39;box&#39;</span><span class="p">)</span>
<span class="n">ax6</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;r = 0.9&#39;</span><span class="p">)</span>

<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/chaos-game-representation_28_0.png" src="../_images/chaos-game-representation_28_0.png" />
</div>
</div>
<p>For <span class="math notranslate nohighlight">\(r &lt; 0.5\)</span>, we can see that some patterns appear, especially those with a dividing rate close to <span class="math notranslate nohighlight">\(r = 0.5\)</span>. More importantly, we see a nonuniformity of distribution in the square.  In contrast, we can see for <span class="math notranslate nohighlight">\(r = 0.6\)</span> that the points fill the corners of the plot area, but they are unevenly-spaced (in comparison to what we have seen for <span class="math notranslate nohighlight">\(r = 0.5\)</span>). This suggests that when <span class="math notranslate nohighlight">\(r &lt; 0.5\)</span>, the attractors are overlapping, indicated by the darker regions of the plot due to the points being more densely packed, and that when <span class="math notranslate nohighlight">\(r&gt;0.5\)</span> the attractor lies on a fractal and is apparently uniformly distributed thereon. For <span class="math notranslate nohighlight">\(r\)</span> significantly less than <span class="math notranslate nohighlight">\(1/2\)</span>, we can see that the CGR becomes more circular with the points spaced more closely together. From this, it infers that as <span class="math notranslate nohighlight">\(r\)</span> decreases, more of the attractors overlap.</p>
<p>For <span class="math notranslate nohighlight">\(r &gt; 0.5\)</span>, we can see that a different pattern appears (compared to those with a dividing rate <span class="math notranslate nohighlight">\(r &lt; 0.5\)</span>). The patterns shown for the CGRs with a dividing ratio <span class="math notranslate nohighlight">\(r &gt; 0.5\)</span> visibly resemble each other. The <span class="math notranslate nohighlight">\(r = 0.9\)</span> case looks like a smaller version of the <span class="math notranslate nohighlight">\(r = 0.75\)</span> case, which similarly, looks like a sparser version of the <span class="math notranslate nohighlight">\(r = 0.6\)</span> case. In contrast to the CGR with the dividing rate <span class="math notranslate nohighlight">\(r = 0.5\)</span>, we can see that the points are spread apart from each other. This can possibly indicate that the attractors do not overlap, and even do not touch each other. Although this can be used to distinguish any nonrandomness, it would be much more difficult to see with the human eye than using the dividing rate <span class="math notranslate nohighlight">\(r = 0.5\)</span>.</p>
<p>For applications which need a different number of vertices than <span class="math notranslate nohighlight">\(m=4\)</span>, it turns out to be optimal to find a dividing rate <span class="math notranslate nohighlight">\(r\)</span> for different polygons that can offer maximum packing of nested, non-overlapping attractors, in order to help find a pattern in a given sequence. In the next subsection, we look at this optimal value of <span class="math notranslate nohighlight">\(r\)</span> as reported in the literature.</p>
</section>
<section id="sec-generalization">
<span id="id18"></span><h4>Generalization of CGR<a class="headerlink" href="#sec-generalization" title="Permalink to this headline">#</a></h4>
<p>In <span id="id19">[<a class="reference internal" href="../references.html#id22" title="A. Fiser, G. E. Tusnady, and I. Simon. Chaos game representation of protein structures. Journal of molecular graphics, 12(4):302–304, 1994.">Fiser <em>et al.</em>, 1994</a>]</span>, Fiser et al. generalized CGR to be applicable for sequences of any number of elements <span class="math notranslate nohighlight">\(m &gt; 2\)</span>. There are many ways of generalizing CGR, some of which are highlighted in <span id="id20">[<a class="reference internal" href="../references.html#id9" title="J. S. Almeida and S. Vinga. Biological sequences as pictures–a generic two dimensional solution for iterated maps. BMC Bioinformatics, 10(1):100, 2009.">Almeida and Vinga, 2009</a>]</span>, but the generalization the authors of the earlier article chose was to use an <span class="math notranslate nohighlight">\(m\)</span>-sided polygon (which they refer to as an <span class="math notranslate nohighlight">\(m\)</span>-gon), where <span class="math notranslate nohighlight">\(m\)</span> was the number of elements in a sequence that should be represented <span id="id21">[<a class="reference internal" href="../references.html#id22" title="A. Fiser, G. E. Tusnady, and I. Simon. Chaos game representation of protein structures. Journal of molecular graphics, 12(4):302–304, 1994.">Fiser <em>et al.</em>, 1994</a>]</span>. As we saw previously in <a class="reference internal" href="#sec-polygons"><span class="std std-ref">Different Polygons</span></a>, the attractors of the <span class="math notranslate nohighlight">\(m\)</span>-gons for when <span class="math notranslate nohighlight">\(m &gt; 4\)</span> and <span class="math notranslate nohighlight">\(r=0.5\)</span> overlapped as shown by the uneven distribution of the points. Because of this, the picture is ambiguous and is not as useful to serve as a way to identity patterns in sequences. To combat this, Fiser et al. introduced a formula to calculate the dividing rate for <span class="math notranslate nohighlight">\(m\)</span>-gons for different values of <span class="math notranslate nohighlight">\(m\)</span>: by putting each attractor piece inside a non-overlapping circle, they derived</p>
<div class="amsmath math notranslate nohighlight">
\[\begin{equation*}
    r = \left(1 + \sin\left(\frac{\pi}{m}\right)\right)^{-1} .
\end{equation*}\]</div>
<p>Although the dividing rate calculated from this formula does prevent attractors from overlapping, Almeida and Vinga noticed in <span id="id22">[<a class="reference internal" href="../references.html#id9" title="J. S. Almeida and S. Vinga. Biological sequences as pictures–a generic two dimensional solution for iterated maps. BMC Bioinformatics, 10(1):100, 2009.">Almeida and Vinga, 2009</a>]</span> that the resulting attractors are not optimally packed. For example, for <span class="math notranslate nohighlight">\(m = 4\)</span>, the solution gives a dividing rate <span class="math notranslate nohighlight">\(r = 0.585786\)</span> instead of <span class="math notranslate nohighlight">\(r = 0.5\)</span>, which works perfectly well (and indeed is maybe better). In <span id="id23">[<a class="reference internal" href="../references.html#id9" title="J. S. Almeida and S. Vinga. Biological sequences as pictures–a generic two dimensional solution for iterated maps. BMC Bioinformatics, 10(1):100, 2009.">Almeida and Vinga, 2009</a>]</span>, the authors gave a more complicated formula for the dividing rate, which they say they derived by using basic trigonometry to pack polygons (instead of circles) into the outer <span class="math notranslate nohighlight">\(m\)</span>-gon.
We have simplified their formula considerably to get</p>
<div class="amsmath math notranslate nohighlight">
\[\begin{equation*}
r = 
\frac{\sin \! \left(\frac{(2  k -1)\pi}{m}\right)}{\sin \! \left(\frac{(2 k -1)\pi}{m}\right)+\sin \! \left(\frac{\pi}{m}\right)}
\end{equation*}\]</div>
<p>where</p>
<div class="amsmath math notranslate nohighlight">
\[\begin{equation*}
    k = \mathrm{round}\left(\frac{m+2}{4}\right) .
\end{equation*}\]</div>
<p>If we use <span class="math notranslate nohighlight">\(k= (m+2)/4\)</span> without rounding to an integer, this formula simplifies to the previous one. This fact was not noted in <span id="id24">[<a class="reference internal" href="../references.html#id9" title="J. S. Almeida and S. Vinga. Biological sequences as pictures–a generic two dimensional solution for iterated maps. BMC Bioinformatics, 10(1):100, 2009.">Almeida and Vinga, 2009</a>]</span>.
We note some special values: when <span class="math notranslate nohighlight">\(m=3\)</span> or <span class="math notranslate nohighlight">\(m=4\)</span>, this formula gives <span class="math notranslate nohighlight">\(r=0.5\)</span>.  When <span class="math notranslate nohighlight">\(m=5\)</span>, this formula gives <span class="math notranslate nohighlight">\(r = (\sqrt{5}-1)/2 = 0.618\ldots\)</span>. Various other <span class="math notranslate nohighlight">\(m\)</span> give algebraic values. When <span class="math notranslate nohighlight">\((m+2)/4\)</span> is half-way between integers, either choice of rounding for <span class="math notranslate nohighlight">\(k\)</span> gives the same <span class="math notranslate nohighlight">\(r\)</span>.</p>
<p>The figure created from the following code (the figure appears immediately after the code) shows the generalized CGRs for various <span class="math notranslate nohighlight">\(m\)</span>-gons using the division formula from <span id="id25">[<a class="reference internal" href="../references.html#id22" title="A. Fiser, G. E. Tusnady, and I. Simon. Chaos game representation of protein structures. Journal of molecular graphics, 12(4):302–304, 1994.">Fiser <em>et al.</em>, 1994</a>]</span> (left) and <span id="id26">[<a class="reference internal" href="../references.html#id9" title="J. S. Almeida and S. Vinga. Biological sequences as pictures–a generic two dimensional solution for iterated maps. BMC Bioinformatics, 10(1):100, 2009.">Almeida and Vinga, 2009</a>]</span> (right).</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">dividingRateFiser</span><span class="p">(</span><span class="n">n</span><span class="p">):</span>
    <span class="k">return</span> <span class="p">(</span><span class="mi">1</span> <span class="o">+</span> <span class="n">math</span><span class="o">.</span><span class="n">sin</span><span class="p">(</span><span class="n">math</span><span class="o">.</span><span class="n">pi</span><span class="o">/</span><span class="n">n</span><span class="p">))</span><span class="o">**</span><span class="p">(</span><span class="o">-</span><span class="mi">1</span><span class="p">)</span>

<span class="k">def</span> <span class="nf">dividingRateAlmeida</span><span class="p">(</span><span class="n">n</span><span class="p">):</span>
    <span class="n">k</span> <span class="o">=</span> <span class="nb">round</span><span class="p">((</span><span class="n">n</span><span class="o">+</span><span class="mf">2.</span><span class="p">)</span><span class="o">/</span><span class="mf">4.</span><span class="p">)</span>
    <span class="n">s_num</span> <span class="o">=</span> <span class="mi">2</span><span class="o">*</span><span class="n">math</span><span class="o">.</span><span class="n">cos</span><span class="p">(</span><span class="n">math</span><span class="o">.</span><span class="n">pi</span><span class="o">*</span><span class="p">((</span><span class="mi">1</span><span class="o">/</span><span class="mi">2</span><span class="p">)</span> <span class="o">-</span> <span class="p">(</span><span class="n">k</span><span class="o">/</span><span class="n">n</span><span class="p">)))</span> <span class="o">-</span> <span class="mi">2</span><span class="o">*</span><span class="n">math</span><span class="o">.</span><span class="n">cos</span><span class="p">(</span><span class="n">math</span><span class="o">.</span><span class="n">pi</span><span class="o">*</span><span class="p">((</span><span class="mi">1</span><span class="o">/</span><span class="mi">2</span><span class="p">)</span><span class="o">-</span><span class="p">(</span><span class="mi">1</span><span class="o">/</span><span class="p">(</span><span class="mi">2</span><span class="o">*</span><span class="n">n</span><span class="p">))))</span><span class="o">*</span><span class="n">math</span><span class="o">.</span><span class="n">cos</span><span class="p">((</span><span class="mi">2</span><span class="o">*</span><span class="n">k</span><span class="o">-</span><span class="mi">1</span><span class="p">)</span><span class="o">*</span><span class="p">(</span><span class="n">math</span><span class="o">.</span><span class="n">pi</span><span class="o">/</span><span class="p">(</span><span class="mi">2</span><span class="o">*</span><span class="n">n</span><span class="p">)))</span><span class="o">*</span><span class="p">(</span><span class="mi">1</span> <span class="o">+</span> <span class="p">(</span><span class="n">math</span><span class="o">.</span><span class="n">tan</span><span class="p">((</span><span class="mi">2</span><span class="o">*</span><span class="n">k</span><span class="o">-</span><span class="mi">1</span><span class="p">)</span><span class="o">*</span><span class="p">(</span><span class="n">math</span><span class="o">.</span><span class="n">pi</span><span class="o">/</span><span class="p">(</span><span class="mi">2</span><span class="o">*</span><span class="n">n</span><span class="p">))))</span><span class="o">/</span><span class="p">(</span><span class="n">math</span><span class="o">.</span><span class="n">tan</span><span class="p">(</span><span class="n">math</span><span class="o">.</span><span class="n">pi</span><span class="o">-</span><span class="p">((</span><span class="n">n</span><span class="o">+</span><span class="mi">2</span><span class="o">*</span><span class="n">k</span><span class="o">-</span><span class="mi">2</span><span class="p">)</span><span class="o">*</span><span class="p">(</span><span class="n">math</span><span class="o">.</span><span class="n">pi</span><span class="o">/</span><span class="p">(</span><span class="mi">2</span><span class="o">*</span><span class="n">n</span><span class="p">))))))</span>
    <span class="n">s_den</span> <span class="o">=</span> <span class="mi">2</span><span class="o">*</span><span class="n">math</span><span class="o">.</span><span class="n">cos</span><span class="p">(</span><span class="n">math</span><span class="o">.</span><span class="n">pi</span><span class="o">*</span><span class="p">((</span><span class="mi">1</span><span class="o">/</span><span class="mi">2</span><span class="p">)</span> <span class="o">-</span> <span class="p">(</span><span class="n">k</span><span class="o">/</span><span class="n">n</span><span class="p">)))</span>
    <span class="k">return</span> <span class="n">s_num</span><span class="o">/</span><span class="n">s_den</span>

<span class="k">def</span> <span class="nf">generalizedCGR</span><span class="p">(</span><span class="n">N</span><span class="p">,</span> <span class="n">divide_func</span><span class="p">,</span> <span class="n">n</span><span class="p">,</span> <span class="o">*</span><span class="n">start</span><span class="p">):</span>
    <span class="k">if</span> <span class="n">start</span><span class="p">:</span>
        <span class="n">shape_coord</span> <span class="o">=</span> <span class="n">n_gon</span><span class="p">(</span><span class="n">n</span><span class="p">,</span> <span class="n">start</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">shape_coord</span> <span class="o">=</span> <span class="n">n_gon</span><span class="p">(</span><span class="n">n</span><span class="p">)</span>
    <span class="n">r</span> <span class="o">=</span> <span class="n">divide_func</span><span class="p">(</span><span class="n">n</span><span class="p">)</span>
    <span class="n">dataPoints</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="mi">2</span><span class="p">,</span> <span class="n">N</span><span class="p">))</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">N</span><span class="p">):</span>
        <span class="n">rand</span> <span class="o">=</span> <span class="n">random</span><span class="o">.</span><span class="n">random</span><span class="p">()</span>
        <span class="n">dataPoints</span><span class="p">[:,</span> <span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="n">dataPoints</span><span class="p">[:,</span> <span class="n">i</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">+</span> <span class="p">(</span><span class="n">shape_coord</span><span class="p">[:,</span> <span class="n">math</span><span class="o">.</span><span class="n">floor</span><span class="p">(</span><span class="n">n</span><span class="o">*</span><span class="n">rand</span><span class="p">)]</span> <span class="o">-</span> <span class="n">dataPoints</span><span class="p">[:,</span> <span class="n">i</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span><span class="o">*</span><span class="n">r</span>
    <span class="k">return</span><span class="p">(</span><span class="n">r</span><span class="p">,</span> <span class="n">shape_coord</span><span class="p">,</span> <span class="n">dataPoints</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">N</span> <span class="o">=</span> <span class="mi">10000</span>

<span class="n">fig</span><span class="p">,</span> <span class="p">((</span><span class="n">ax1</span><span class="p">,</span> <span class="n">ax2</span><span class="p">),</span> <span class="p">(</span><span class="n">ax3</span><span class="p">,</span> <span class="n">ax4</span><span class="p">),</span> <span class="p">(</span><span class="n">ax5</span><span class="p">,</span> <span class="n">ax6</span><span class="p">))</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">3</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">8</span><span class="p">))</span>
<span class="n">fig</span><span class="o">.</span><span class="n">suptitle</span><span class="p">(</span><span class="s1">&#39;Graphical comparison of proposed CGR generalization (Fiser 1994 (left) vs. Almeida 2009 (right))&#39;</span><span class="p">)</span>

<span class="p">(</span><span class="n">r</span><span class="p">,</span> <span class="n">coord</span><span class="p">,</span> <span class="n">points</span><span class="p">)</span> <span class="o">=</span> <span class="n">generalizedCGR</span><span class="p">(</span><span class="n">N</span><span class="p">,</span> <span class="n">dividingRateFiser</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">90</span><span class="p">)</span>
<span class="n">ax1</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">coord</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="p">:],</span> <span class="n">coord</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="p">:],</span> <span class="s1">&#39;k&#39;</span><span class="p">)</span>
<span class="n">ax1</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">points</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="p">:],</span> <span class="n">points</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="p">:],</span> <span class="s1">&#39;k.&#39;</span><span class="p">,</span> <span class="n">markersize</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="n">ax1</span><span class="o">.</span><span class="n">axis</span><span class="p">(</span><span class="s1">&#39;off&#39;</span><span class="p">)</span>
<span class="n">ax1</span><span class="o">.</span><span class="n">set_aspect</span><span class="p">(</span><span class="s1">&#39;equal&#39;</span><span class="p">,</span> <span class="s1">&#39;box&#39;</span><span class="p">)</span>
<span class="n">ax1</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;r = </span><span class="si">{}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">r</span><span class="p">))</span>

<span class="p">(</span><span class="n">r</span><span class="p">,</span> <span class="n">coord</span><span class="p">,</span> <span class="n">points</span><span class="p">)</span> <span class="o">=</span> <span class="n">generalizedCGR</span><span class="p">(</span><span class="n">N</span><span class="p">,</span> <span class="n">dividingRateAlmeida</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">90</span><span class="p">)</span>
<span class="n">ax2</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">coord</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="p">:],</span> <span class="n">coord</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="p">:],</span> <span class="s1">&#39;k&#39;</span><span class="p">)</span>
<span class="n">ax2</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">points</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="p">:],</span> <span class="n">points</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="p">:],</span> <span class="s1">&#39;k.&#39;</span><span class="p">,</span> <span class="n">markersize</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="n">ax2</span><span class="o">.</span><span class="n">axis</span><span class="p">(</span><span class="s1">&#39;off&#39;</span><span class="p">)</span>
<span class="n">ax2</span><span class="o">.</span><span class="n">set_aspect</span><span class="p">(</span><span class="s1">&#39;equal&#39;</span><span class="p">,</span> <span class="s1">&#39;box&#39;</span><span class="p">)</span>
<span class="n">ax2</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;r = </span><span class="si">{}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">r</span><span class="p">))</span>

<span class="p">(</span><span class="n">r</span><span class="p">,</span> <span class="n">coord</span><span class="p">,</span> <span class="n">points</span><span class="p">)</span> <span class="o">=</span> <span class="n">generalizedCGR</span><span class="p">(</span><span class="n">N</span><span class="p">,</span> <span class="n">dividingRateFiser</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">45</span><span class="p">)</span>
<span class="n">ax3</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">coord</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="p">:],</span> <span class="n">coord</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="p">:],</span> <span class="s1">&#39;k&#39;</span><span class="p">)</span>
<span class="n">ax3</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">points</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="p">:],</span> <span class="n">points</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="p">:],</span> <span class="s1">&#39;k.&#39;</span><span class="p">,</span> <span class="n">markersize</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="n">ax3</span><span class="o">.</span><span class="n">axis</span><span class="p">(</span><span class="s1">&#39;off&#39;</span><span class="p">)</span>
<span class="n">ax3</span><span class="o">.</span><span class="n">set_aspect</span><span class="p">(</span><span class="s1">&#39;equal&#39;</span><span class="p">,</span> <span class="s1">&#39;box&#39;</span><span class="p">)</span>
<span class="n">ax3</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;r = </span><span class="si">{}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">r</span><span class="p">))</span>

<span class="p">(</span><span class="n">r</span><span class="p">,</span> <span class="n">coord</span><span class="p">,</span> <span class="n">points</span><span class="p">)</span> <span class="o">=</span> <span class="n">generalizedCGR</span><span class="p">(</span><span class="n">N</span><span class="p">,</span> <span class="n">dividingRateAlmeida</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">45</span><span class="p">)</span>
<span class="n">ax4</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">coord</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="p">:],</span> <span class="n">coord</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="p">:],</span> <span class="s1">&#39;k&#39;</span><span class="p">)</span>
<span class="n">ax4</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">points</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="p">:],</span> <span class="n">points</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="p">:],</span> <span class="s1">&#39;k.&#39;</span><span class="p">,</span> <span class="n">markersize</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="n">ax4</span><span class="o">.</span><span class="n">axis</span><span class="p">(</span><span class="s1">&#39;off&#39;</span><span class="p">)</span>
<span class="n">ax4</span><span class="o">.</span><span class="n">set_aspect</span><span class="p">(</span><span class="s1">&#39;equal&#39;</span><span class="p">,</span> <span class="s1">&#39;box&#39;</span><span class="p">)</span>
<span class="n">ax4</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;r = </span><span class="si">{}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">r</span><span class="p">))</span>

<span class="p">(</span><span class="n">r</span><span class="p">,</span> <span class="n">coord</span><span class="p">,</span> <span class="n">points</span><span class="p">)</span> <span class="o">=</span> <span class="n">generalizedCGR</span><span class="p">(</span><span class="n">N</span><span class="p">,</span> <span class="n">dividingRateFiser</span><span class="p">,</span> <span class="mi">5</span><span class="p">,</span> <span class="mi">90</span><span class="p">)</span>
<span class="n">ax5</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">coord</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="p">:],</span> <span class="n">coord</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="p">:],</span> <span class="s1">&#39;k&#39;</span><span class="p">)</span>
<span class="n">ax5</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">points</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="p">:],</span> <span class="n">points</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="p">:],</span> <span class="s1">&#39;k.&#39;</span><span class="p">,</span> <span class="n">markersize</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="n">ax5</span><span class="o">.</span><span class="n">axis</span><span class="p">(</span><span class="s1">&#39;off&#39;</span><span class="p">)</span>
<span class="n">ax5</span><span class="o">.</span><span class="n">set_aspect</span><span class="p">(</span><span class="s1">&#39;equal&#39;</span><span class="p">,</span> <span class="s1">&#39;box&#39;</span><span class="p">)</span>
<span class="n">ax5</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;r = </span><span class="si">{}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">r</span><span class="p">))</span>

<span class="p">(</span><span class="n">r</span><span class="p">,</span> <span class="n">coord</span><span class="p">,</span> <span class="n">points</span><span class="p">)</span> <span class="o">=</span> <span class="n">generalizedCGR</span><span class="p">(</span><span class="n">N</span><span class="p">,</span> <span class="n">dividingRateAlmeida</span><span class="p">,</span> <span class="mi">5</span><span class="p">,</span> <span class="mi">90</span><span class="p">)</span>
<span class="n">ax6</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">coord</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="p">:],</span> <span class="n">coord</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="p">:],</span> <span class="s1">&#39;k&#39;</span><span class="p">)</span>
<span class="n">ax6</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">points</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="p">:],</span> <span class="n">points</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="p">:],</span> <span class="s1">&#39;k.&#39;</span><span class="p">,</span> <span class="n">markersize</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="n">ax6</span><span class="o">.</span><span class="n">axis</span><span class="p">(</span><span class="s1">&#39;off&#39;</span><span class="p">)</span>
<span class="n">ax6</span><span class="o">.</span><span class="n">set_aspect</span><span class="p">(</span><span class="s1">&#39;equal&#39;</span><span class="p">,</span> <span class="s1">&#39;box&#39;</span><span class="p">)</span>
<span class="n">ax6</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;r = </span><span class="si">{}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">r</span><span class="p">))</span>

<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/chaos-game-representation_32_0.png" src="../_images/chaos-game-representation_32_0.png" />
</div>
</div>
<div class="tip admonition" id="cgractivity-3">
<p class="admonition-title">CGR Activity 3</p>
<p>Using the provided code (or if you’re brave, write your own), create a generalized CGR of a 6, 9, and 12-sided polygon (using the dividing rate formula by Almeida). <a class="reference internal" href="../Solutions/Reports%20on%20Chaos%20Game%20Activities.html#cgractivityreport-3"><span class="std std-ref">[What happened when we did this]</span></a></p>
</div>
<p>In the next section, we will look at applications of CGR, mainly biological and mathematical. For the following examples, unless otherwise specified, assume that <span class="math notranslate nohighlight">\(r = 0.5\)</span>.</p>
</section>
</section>
</section>
<section id="cgr-of-biological-sequences">
<span id="sec-dna"></span><h2>CGR of Biological Sequences<a class="headerlink" href="#cgr-of-biological-sequences" title="Permalink to this headline">#</a></h2>
<p>The chaos game representation has been applied to biological areas. In this section, we will look at how the chaos game has been applied to DNA and protein (amino acids) sequencing and visualizations. We also will give a brief overview of the <em>Map of Life</em>, an extension of the visualization of DNA sequences. The Map of Life shows potential in allowing researchers to quantitatively classify species that was once unclear in its taxonomy.</p>
<section id="nucleotide-sequences">
<h3>Nucleotide Sequences<a class="headerlink" href="#nucleotide-sequences" title="Permalink to this headline">#</a></h3>
<p>H. J. Jeffrey was the first to propose using the chaos game representation as a novel way of visualizing nucleotide sequences. This revealed previously unknown patterns in certain proteins <span id="id27">[<a class="reference internal" href="../references.html#id27" title="H. J. Jeffrey. Chaos game representation of gene structure. Nucleic Acids Research, 18(8):2163–2170, 1990.">Jeffrey, 1990</a>, <a class="reference internal" href="../references.html#id28" title="H. J. Jeffrey. Chaos game visualization of sequences. Computers &amp; Graphics, 16(1):25–33, 1992.">Jeffrey, 1992</a>]</span>. The genetic sequence is made up of four bases: adenine (A), guanine (G), cytosine (C), and either thymine (T) or uracil (U) for DNA or RNA, respectively. Using a square-shaped CGR, we label the four corners with the name of each base. In this paper, A is in the bottom-left corner, C in the top-left, G in the top-right, and U/T in the bottom right. Rather than use a random number generator to determine which map to use for each iteration as we did in <a class="reference internal" href="#sec-chaos"><span class="std std-ref">Chaos Game</span></a>, we follow the genetic sequence that we want to create a CGR for.</p>
<p>To demonstrate the chaos game for DNA sequences, let us walk through plotting the first five bases of the DNA sequence <a class="reference external" href="https://www.ncbi.nlm.nih.gov/nuccore/U01317.1?report=fasta">HUMHBB</a> (human beta globin region, chromosome 11), “GAATT”, shown in <a class="reference internal" href="#cgr-globin"><span class="std std-numref">Fig. 8</span></a>. We first mark the center as our initial point. The first base in the sequence is “G” so, we plot a point half way between our initial point and the “G” corner. The next base in the sequence is “A” so we plot a point half way between the point that we just plotted and the “A” corner. The following is the code to recreate the CGR.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">parse_sequence</span><span class="p">(</span><span class="n">filename</span><span class="p">):</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">filename</span><span class="p">)</span> <span class="k">as</span> <span class="n">inputfile</span><span class="p">:</span>
        <span class="nb">next</span><span class="p">(</span><span class="n">inputfile</span><span class="p">)</span>
        <span class="n">results</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">csv</span><span class="o">.</span><span class="n">reader</span><span class="p">(</span><span class="n">inputfile</span><span class="p">))</span>
    <span class="n">seq</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span>
    <span class="k">for</span> <span class="n">r</span> <span class="ow">in</span> <span class="n">results</span><span class="p">[:</span><span class="o">-</span><span class="mi">1</span><span class="p">]:</span>
        <span class="n">seq</span> <span class="o">=</span> <span class="n">seq</span> <span class="o">+</span> <span class="n">r</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
    <span class="k">return</span> <span class="n">seq</span>

<span class="k">def</span> <span class="nf">dna_cgr</span><span class="p">(</span><span class="n">seq</span><span class="p">):</span>
    <span class="n">N</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span>
    <span class="n">dataPoints</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="mi">2</span><span class="p">,</span> <span class="n">N</span><span class="o">+</span><span class="mi">1</span><span class="p">))</span>
    <span class="n">dataPoints</span><span class="p">[:,</span> <span class="mi">0</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="mf">0.5</span><span class="p">,</span> <span class="mf">0.5</span><span class="p">])</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">N</span><span class="o">+</span><span class="mi">1</span><span class="p">):</span>
        <span class="k">if</span> <span class="n">seq</span><span class="p">[</span><span class="n">i</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;A&#39;</span><span class="p">:</span>
            <span class="n">corner</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">])</span>
        <span class="k">elif</span> <span class="n">seq</span><span class="p">[</span><span class="n">i</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;C&#39;</span><span class="p">:</span>
            <span class="n">corner</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">])</span>
        <span class="k">elif</span> <span class="n">seq</span><span class="p">[</span><span class="n">i</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;G&#39;</span><span class="p">:</span>
            <span class="n">corner</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">])</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">corner</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">])</span>
        <span class="n">dataPoints</span><span class="p">[:,</span> <span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="mf">0.5</span><span class="o">*</span><span class="p">(</span><span class="n">dataPoints</span><span class="p">[:,</span> <span class="n">i</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">+</span> <span class="n">corner</span><span class="p">)</span>
    <span class="k">return</span><span class="p">(</span><span class="n">dataPoints</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">humhbb_dna</span> <span class="o">=</span> <span class="n">parse_sequence</span><span class="p">(</span><span class="s1">&#39;../Supplementary Material/CGR/FASTA Files/HUMHBB.fasta&#39;</span><span class="p">)</span>
<span class="n">coord</span> <span class="o">=</span> <span class="n">dna_cgr</span><span class="p">(</span><span class="n">humhbb_dna</span><span class="p">[:</span><span class="mi">5</span><span class="p">])</span>

<span class="n">fig</span><span class="p">,</span> <span class="p">((</span><span class="n">ax1</span><span class="p">,</span> <span class="n">ax2</span><span class="p">,</span> <span class="n">ax3</span><span class="p">),</span> <span class="p">(</span><span class="n">ax4</span><span class="p">,</span> <span class="n">ax5</span><span class="p">,</span> <span class="n">ax6</span><span class="p">))</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">12</span><span class="p">,</span> <span class="mi">8</span><span class="p">))</span>
<span class="n">fig</span><span class="o">.</span><span class="n">suptitle</span><span class="p">(</span><span class="s1">&#39;First five steps of producing CGR using HUMHBB Data&#39;</span><span class="p">)</span>

<span class="n">ax1</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">coord</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="p">:</span><span class="mi">1</span><span class="p">],</span> <span class="n">coord</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="p">:</span><span class="mi">1</span><span class="p">],</span> <span class="s1">&#39;k.&#39;</span><span class="p">)</span>
<span class="n">ax1</span><span class="o">.</span><span class="n">axes</span><span class="o">.</span><span class="n">xaxis</span><span class="o">.</span><span class="n">set_visible</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
<span class="n">ax1</span><span class="o">.</span><span class="n">axes</span><span class="o">.</span><span class="n">yaxis</span><span class="o">.</span><span class="n">set_visible</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
<span class="n">ax1</span><span class="o">.</span><span class="n">set_xlim</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">])</span>
<span class="n">ax1</span><span class="o">.</span><span class="n">set_ylim</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">])</span>
<span class="n">ax1</span><span class="o">.</span><span class="n">set_aspect</span><span class="p">(</span><span class="s1">&#39;equal&#39;</span><span class="p">,</span> <span class="s1">&#39;box&#39;</span><span class="p">)</span>
<span class="n">ax1</span><span class="o">.</span><span class="n">text</span><span class="p">(</span><span class="mf">0.025</span><span class="p">,</span> <span class="mf">0.025</span><span class="p">,</span> <span class="s1">&#39;A&#39;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">12</span><span class="p">)</span>
<span class="n">ax1</span><span class="o">.</span><span class="n">text</span><span class="p">(</span><span class="mf">0.025</span><span class="p">,</span> <span class="mf">0.925</span><span class="p">,</span> <span class="s1">&#39;C&#39;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">12</span><span class="p">)</span>
<span class="n">ax1</span><span class="o">.</span><span class="n">text</span><span class="p">(</span><span class="mf">0.94</span><span class="p">,</span> <span class="mf">0.925</span><span class="p">,</span> <span class="s1">&#39;G&#39;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">12</span><span class="p">)</span>
<span class="n">ax1</span><span class="o">.</span><span class="n">text</span><span class="p">(</span><span class="mf">0.94</span><span class="p">,</span> <span class="mf">0.025</span><span class="p">,</span> <span class="s1">&#39;T&#39;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">12</span><span class="p">)</span>
<span class="n">ax1</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;Initial Point&#39;</span><span class="p">)</span>

<span class="n">ax2</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">coord</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="p">:</span><span class="mi">2</span><span class="p">],</span> <span class="n">coord</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="p">:</span><span class="mi">2</span><span class="p">],</span> <span class="s1">&#39;k.&#39;</span><span class="p">)</span>
<span class="n">ax2</span><span class="o">.</span><span class="n">axes</span><span class="o">.</span><span class="n">xaxis</span><span class="o">.</span><span class="n">set_visible</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
<span class="n">ax2</span><span class="o">.</span><span class="n">axes</span><span class="o">.</span><span class="n">yaxis</span><span class="o">.</span><span class="n">set_visible</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
<span class="n">ax2</span><span class="o">.</span><span class="n">set_xlim</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">])</span>
<span class="n">ax2</span><span class="o">.</span><span class="n">set_ylim</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">])</span>
<span class="n">ax2</span><span class="o">.</span><span class="n">set_aspect</span><span class="p">(</span><span class="s1">&#39;equal&#39;</span><span class="p">,</span> <span class="s1">&#39;box&#39;</span><span class="p">)</span>
<span class="n">ax2</span><span class="o">.</span><span class="n">text</span><span class="p">(</span><span class="mf">0.025</span><span class="p">,</span> <span class="mf">0.025</span><span class="p">,</span> <span class="s1">&#39;A&#39;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">12</span><span class="p">)</span>
<span class="n">ax2</span><span class="o">.</span><span class="n">text</span><span class="p">(</span><span class="mf">0.025</span><span class="p">,</span> <span class="mf">0.925</span><span class="p">,</span> <span class="s1">&#39;C&#39;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">12</span><span class="p">)</span>
<span class="n">ax2</span><span class="o">.</span><span class="n">text</span><span class="p">(</span><span class="mf">0.94</span><span class="p">,</span> <span class="mf">0.925</span><span class="p">,</span> <span class="s1">&#39;G&#39;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">12</span><span class="p">)</span>
<span class="n">ax2</span><span class="o">.</span><span class="n">text</span><span class="p">(</span><span class="mf">0.94</span><span class="p">,</span> <span class="mf">0.025</span><span class="p">,</span> <span class="s1">&#39;T&#39;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">12</span><span class="p">)</span>
<span class="n">ax2</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;G&#39;</span><span class="p">)</span>

<span class="n">ax3</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">coord</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="p">:</span><span class="mi">3</span><span class="p">],</span> <span class="n">coord</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="p">:</span><span class="mi">3</span><span class="p">],</span> <span class="s1">&#39;k.&#39;</span><span class="p">)</span>
<span class="n">ax3</span><span class="o">.</span><span class="n">axes</span><span class="o">.</span><span class="n">xaxis</span><span class="o">.</span><span class="n">set_visible</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
<span class="n">ax3</span><span class="o">.</span><span class="n">axes</span><span class="o">.</span><span class="n">yaxis</span><span class="o">.</span><span class="n">set_visible</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
<span class="n">ax3</span><span class="o">.</span><span class="n">set_xlim</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">])</span>
<span class="n">ax3</span><span class="o">.</span><span class="n">set_ylim</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">])</span>
<span class="n">ax3</span><span class="o">.</span><span class="n">set_aspect</span><span class="p">(</span><span class="s1">&#39;equal&#39;</span><span class="p">,</span> <span class="s1">&#39;box&#39;</span><span class="p">)</span>
<span class="n">ax3</span><span class="o">.</span><span class="n">text</span><span class="p">(</span><span class="mf">0.025</span><span class="p">,</span> <span class="mf">0.025</span><span class="p">,</span> <span class="s1">&#39;A&#39;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">12</span><span class="p">)</span>
<span class="n">ax3</span><span class="o">.</span><span class="n">text</span><span class="p">(</span><span class="mf">0.025</span><span class="p">,</span> <span class="mf">0.925</span><span class="p">,</span> <span class="s1">&#39;C&#39;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">12</span><span class="p">)</span>
<span class="n">ax3</span><span class="o">.</span><span class="n">text</span><span class="p">(</span><span class="mf">0.94</span><span class="p">,</span> <span class="mf">0.925</span><span class="p">,</span> <span class="s1">&#39;G&#39;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">12</span><span class="p">)</span>
<span class="n">ax3</span><span class="o">.</span><span class="n">text</span><span class="p">(</span><span class="mf">0.94</span><span class="p">,</span> <span class="mf">0.025</span><span class="p">,</span> <span class="s1">&#39;T&#39;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">12</span><span class="p">)</span>
<span class="n">ax3</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;GA&#39;</span><span class="p">)</span>

<span class="n">ax4</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">coord</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="p">:</span><span class="mi">4</span><span class="p">],</span> <span class="n">coord</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="p">:</span><span class="mi">4</span><span class="p">],</span> <span class="s1">&#39;k.&#39;</span><span class="p">)</span>
<span class="n">ax4</span><span class="o">.</span><span class="n">axes</span><span class="o">.</span><span class="n">xaxis</span><span class="o">.</span><span class="n">set_visible</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
<span class="n">ax4</span><span class="o">.</span><span class="n">axes</span><span class="o">.</span><span class="n">yaxis</span><span class="o">.</span><span class="n">set_visible</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
<span class="n">ax4</span><span class="o">.</span><span class="n">set_xlim</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">])</span>
<span class="n">ax4</span><span class="o">.</span><span class="n">set_ylim</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">])</span>
<span class="n">ax4</span><span class="o">.</span><span class="n">set_aspect</span><span class="p">(</span><span class="s1">&#39;equal&#39;</span><span class="p">,</span> <span class="s1">&#39;box&#39;</span><span class="p">)</span>
<span class="n">ax4</span><span class="o">.</span><span class="n">text</span><span class="p">(</span><span class="mf">0.025</span><span class="p">,</span> <span class="mf">0.025</span><span class="p">,</span> <span class="s1">&#39;A&#39;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">12</span><span class="p">)</span>
<span class="n">ax4</span><span class="o">.</span><span class="n">text</span><span class="p">(</span><span class="mf">0.025</span><span class="p">,</span> <span class="mf">0.925</span><span class="p">,</span> <span class="s1">&#39;C&#39;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">12</span><span class="p">)</span>
<span class="n">ax4</span><span class="o">.</span><span class="n">text</span><span class="p">(</span><span class="mf">0.94</span><span class="p">,</span> <span class="mf">0.925</span><span class="p">,</span> <span class="s1">&#39;G&#39;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">12</span><span class="p">)</span>
<span class="n">ax4</span><span class="o">.</span><span class="n">text</span><span class="p">(</span><span class="mf">0.94</span><span class="p">,</span> <span class="mf">0.025</span><span class="p">,</span> <span class="s1">&#39;T&#39;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">12</span><span class="p">)</span>
<span class="n">ax4</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;GAA&#39;</span><span class="p">)</span>

<span class="n">ax5</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">coord</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="p">:</span><span class="mi">5</span><span class="p">],</span> <span class="n">coord</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="p">:</span><span class="mi">5</span><span class="p">],</span> <span class="s1">&#39;k.&#39;</span><span class="p">)</span>
<span class="n">ax5</span><span class="o">.</span><span class="n">axes</span><span class="o">.</span><span class="n">xaxis</span><span class="o">.</span><span class="n">set_visible</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
<span class="n">ax5</span><span class="o">.</span><span class="n">axes</span><span class="o">.</span><span class="n">yaxis</span><span class="o">.</span><span class="n">set_visible</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
<span class="n">ax5</span><span class="o">.</span><span class="n">set_xlim</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">])</span>
<span class="n">ax5</span><span class="o">.</span><span class="n">set_ylim</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">])</span>
<span class="n">ax5</span><span class="o">.</span><span class="n">set_aspect</span><span class="p">(</span><span class="s1">&#39;equal&#39;</span><span class="p">,</span> <span class="s1">&#39;box&#39;</span><span class="p">)</span>
<span class="n">ax5</span><span class="o">.</span><span class="n">text</span><span class="p">(</span><span class="mf">0.025</span><span class="p">,</span> <span class="mf">0.025</span><span class="p">,</span> <span class="s1">&#39;A&#39;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">12</span><span class="p">)</span>
<span class="n">ax5</span><span class="o">.</span><span class="n">text</span><span class="p">(</span><span class="mf">0.025</span><span class="p">,</span> <span class="mf">0.925</span><span class="p">,</span> <span class="s1">&#39;C&#39;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">12</span><span class="p">)</span>
<span class="n">ax5</span><span class="o">.</span><span class="n">text</span><span class="p">(</span><span class="mf">0.94</span><span class="p">,</span> <span class="mf">0.925</span><span class="p">,</span> <span class="s1">&#39;G&#39;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">12</span><span class="p">)</span>
<span class="n">ax5</span><span class="o">.</span><span class="n">text</span><span class="p">(</span><span class="mf">0.94</span><span class="p">,</span> <span class="mf">0.025</span><span class="p">,</span> <span class="s1">&#39;T&#39;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">12</span><span class="p">)</span>
<span class="n">ax5</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;GAAT&#39;</span><span class="p">)</span>

<span class="n">ax6</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">coord</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="p">:</span><span class="mi">6</span><span class="p">],</span> <span class="n">coord</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="p">:</span><span class="mi">6</span><span class="p">],</span> <span class="s1">&#39;k.&#39;</span><span class="p">)</span>
<span class="n">ax6</span><span class="o">.</span><span class="n">axes</span><span class="o">.</span><span class="n">xaxis</span><span class="o">.</span><span class="n">set_visible</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
<span class="n">ax6</span><span class="o">.</span><span class="n">axes</span><span class="o">.</span><span class="n">yaxis</span><span class="o">.</span><span class="n">set_visible</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
<span class="n">ax6</span><span class="o">.</span><span class="n">set_xlim</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">])</span>
<span class="n">ax6</span><span class="o">.</span><span class="n">set_ylim</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">])</span>
<span class="n">ax6</span><span class="o">.</span><span class="n">set_aspect</span><span class="p">(</span><span class="s1">&#39;equal&#39;</span><span class="p">,</span> <span class="s1">&#39;box&#39;</span><span class="p">)</span>
<span class="n">ax6</span><span class="o">.</span><span class="n">text</span><span class="p">(</span><span class="mf">0.025</span><span class="p">,</span> <span class="mf">0.025</span><span class="p">,</span> <span class="s1">&#39;A&#39;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">12</span><span class="p">)</span>
<span class="n">ax6</span><span class="o">.</span><span class="n">text</span><span class="p">(</span><span class="mf">0.025</span><span class="p">,</span> <span class="mf">0.925</span><span class="p">,</span> <span class="s1">&#39;C&#39;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">12</span><span class="p">)</span>
<span class="n">ax6</span><span class="o">.</span><span class="n">text</span><span class="p">(</span><span class="mf">0.94</span><span class="p">,</span> <span class="mf">0.925</span><span class="p">,</span> <span class="s1">&#39;G&#39;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">12</span><span class="p">)</span>
<span class="n">ax6</span><span class="o">.</span><span class="n">text</span><span class="p">(</span><span class="mf">0.94</span><span class="p">,</span> <span class="mf">0.025</span><span class="p">,</span> <span class="s1">&#39;T&#39;</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">12</span><span class="p">)</span>
<span class="n">ax6</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;GAATT&#39;</span><span class="p">)</span>

<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/chaos-game-representation_38_0.png" src="../_images/chaos-game-representation_38_0.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">coord</span> <span class="o">=</span> <span class="n">dna_cgr</span><span class="p">(</span><span class="n">humhbb_dna</span><span class="p">)</span>

<span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">coord</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="p">:],</span> <span class="n">coord</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="p">:],</span> <span class="s1">&#39;k.&#39;</span><span class="p">,</span> <span class="n">markersize</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">gca</span><span class="p">()</span><span class="o">.</span><span class="n">set_aspect</span><span class="p">(</span><span class="s1">&#39;equal&#39;</span><span class="p">,</span> <span class="n">adjustable</span><span class="o">=</span><span class="s1">&#39;box&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">axis</span><span class="p">(</span><span class="s1">&#39;off&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/chaos-game-representation_39_0.png" src="../_images/chaos-game-representation_39_0.png" />
</div>
</div>
<p>The finished CGR is shown in <a class="reference internal" href="#cgr-globin"><span class="std std-numref">Fig. 8</span></a> at the start of this section, which certainly has an identifiable fractal pattern. The most prominent repeated pattern is the unshaded area in the top right quadrant of the figure, referred to as the “double scoop” appearance by <span id="id28">[<a class="reference internal" href="../references.html#id27" title="H. J. Jeffrey. Chaos game representation of gene structure. Nucleic Acids Research, 18(8):2163–2170, 1990.">Jeffrey, 1990</a>]</span>, which actually appears in almost all vertebrate DNA sequences. This pattern is due to the fact that there is a comparative sparseness of guanine following cytosine in the gene sequence since CG dinucleotides are prone to methylation and subsequently mutation.</p>
<p>To fully understand the double scoop pattern, we must understand the biological meaning of the CGR. Each point plotted in the CGR corresponds to a base, and depending on where it is placed, we can trace back and figure out parts of the sequence we are examining <span id="id29">[<a class="reference internal" href="../references.html#id24" title="N. Goldman. Nucleotide, dinucleotide and trinucleotide frequencies explain patterns observed in chaos game representations of DNA sequences. Nucleic Acids Research, 21(10):2487–2491, 1993.">Goldman, 1993</a>]</span>. <a class="reference internal" href="#cgr-quadrants"><span class="std std-numref">Fig. 10</span></a> shows the relationship corresponding area of the CGR and the DNA sequence. In reference to this figure, we can see that for any point that corresponds with base G, it will be located in the upper right quadrant of the CGR plot. To see what the previous base is, we can divide the quadrant into sub-quadrants (labeling them in the same order as the quadrants), and depending on where the point is, we can determine what the previous base of the sequence is. We can repeat this step again and again to find the order in which the bases appear in the sequence. <a class="reference internal" href="#cgr-no-cg"><span class="std std-numref">Fig. 11</span></a> shows a CGR square where all the CG quadrants are unfilled. We can see from this figure that even though it is not a real CGR (since we did not use a sequence to produce it), we still get the same double scoop pattern found in <a class="reference internal" href="#cgr-globin"><span class="std std-numref">Fig. 8</span></a>, displayed at the beginning of the unit. By identifying regions of the CGR square in this way, it is possible to identify features of DNA sequences that correspond to patterns of the CGR.</p>
<figure class="align-default" id="cgr-quadrants">
<a class="reference internal image-reference" href="../_images/Generalized_CGR.png"><img alt="../_images/Generalized_CGR.png" src="../_images/Generalized_CGR.png" style="height: 300px;" /></a>
<figcaption>
<p><span class="caption-number">Fig. 10 </span><span class="caption-text">Correspondence between DNA sequences and areas of the CGR of DNA sequences</span><a class="headerlink" href="#cgr-quadrants" title="Permalink to this image">#</a></p>
</figcaption>
</figure>
<figure class="align-default" id="cgr-no-cg">
<a class="reference internal image-reference" href="../_images/HUMHBB_CGR.png"><img alt="../_images/HUMHBB_CGR.png" src="../_images/HUMHBB_CGR.png" style="height: 300px;" /></a>
<figcaption>
<p><span class="caption-number">Fig. 11 </span><span class="caption-text">Explanation of the double scoop pattern—plot of CGR square with all CG quadrants unfilled.</span><a class="headerlink" href="#cgr-no-cg" title="Permalink to this image">#</a></p>
</figcaption>
</figure>
<p>The following shows other examples of CGRs for DNA sequences. We can see that the CGR of DNA sequence of human herpesvirus strain also exhibit the double scoop pattern that we have seen for HUMHBB. This agrees with what was mentioned earlier in that it is common to see the lack of C and G dinucleotides together within the human genome. The CGR of the DNA sequence of the chloroplast of quinoa plant, on the other hand, does not show a double scoop pattern. We can see from using the CGR of DNA sequences, we are able to distinguish between different species very easily.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">herpesvirus_dna</span> <span class="o">=</span> <span class="n">parse_sequence</span><span class="p">(</span><span class="s1">&#39;../Supplementary Material/CGR/FASTA Files/KF021605.fasta&#39;</span><span class="p">)</span>
<span class="n">herpesvirus_coord</span> <span class="o">=</span> <span class="n">dna_cgr</span><span class="p">(</span><span class="n">herpesvirus_dna</span><span class="p">)</span>

<span class="n">chloroplast_dna</span> <span class="o">=</span> <span class="n">parse_sequence</span><span class="p">(</span><span class="s1">&#39;../Supplementary Material/CGR/FASTA Files/CM008430.fasta&#39;</span><span class="p">)</span>
<span class="n">chloroplast_coord</span> <span class="o">=</span> <span class="n">dna_cgr</span><span class="p">(</span><span class="n">chloroplast_dna</span><span class="p">)</span>

<span class="n">fig</span><span class="p">,</span> <span class="p">(</span><span class="n">ax1</span><span class="p">,</span> <span class="n">ax2</span><span class="p">)</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">16</span><span class="p">,</span> <span class="mi">6</span><span class="p">))</span>
<span class="n">fig</span><span class="o">.</span><span class="n">suptitle</span><span class="p">(</span><span class="s1">&#39;More example of CGRs of DNA sequences&#39;</span><span class="p">)</span>
<span class="n">ax1</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">herpesvirus_coord</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="p">:],</span> <span class="n">herpesvirus_coord</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="p">:],</span> <span class="s1">&#39;k.&#39;</span><span class="p">,</span> <span class="n">markersize</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="n">ax1</span><span class="o">.</span><span class="n">axes</span><span class="o">.</span><span class="n">xaxis</span><span class="o">.</span><span class="n">set_visible</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
<span class="n">ax1</span><span class="o">.</span><span class="n">axes</span><span class="o">.</span><span class="n">yaxis</span><span class="o">.</span><span class="n">set_visible</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
<span class="n">ax1</span><span class="o">.</span><span class="n">set_xlim</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">])</span>
<span class="n">ax1</span><span class="o">.</span><span class="n">set_ylim</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">])</span>
<span class="n">ax1</span><span class="o">.</span><span class="n">set_aspect</span><span class="p">(</span><span class="s1">&#39;equal&#39;</span><span class="p">,</span> <span class="s1">&#39;box&#39;</span><span class="p">)</span>
<span class="n">ax1</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;Human herpesvirus 5 strain TR, complete genome&#39;</span><span class="p">)</span>

<span class="n">ax2</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">chloroplast_coord</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="p">:],</span> <span class="n">chloroplast_coord</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="p">:],</span> <span class="s1">&#39;k.&#39;</span><span class="p">,</span> <span class="n">markersize</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="n">ax2</span><span class="o">.</span><span class="n">axes</span><span class="o">.</span><span class="n">xaxis</span><span class="o">.</span><span class="n">set_visible</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
<span class="n">ax2</span><span class="o">.</span><span class="n">axes</span><span class="o">.</span><span class="n">yaxis</span><span class="o">.</span><span class="n">set_visible</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
<span class="n">ax2</span><span class="o">.</span><span class="n">set_xlim</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">])</span>
<span class="n">ax2</span><span class="o">.</span><span class="n">set_ylim</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">])</span>
<span class="n">ax2</span><span class="o">.</span><span class="n">set_aspect</span><span class="p">(</span><span class="s1">&#39;equal&#39;</span><span class="p">,</span> <span class="s1">&#39;box&#39;</span><span class="p">)</span>
<span class="n">ax2</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;Chenopodium quinoa cultivar Real Blanca cholorplast, complete sequence&#39;</span><span class="p">)</span>

<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/chaos-game-representation_43_0.png" src="../_images/chaos-game-representation_43_0.png" />
</div>
</div>
<div class="tip admonition" id="cgractivity-4">
<p class="admonition-title">CGR Activity 4</p>
<p>Produce your own CGR of DNA sequences of your chosing. Pick sequences that are from difference species, such as plants, fungi, protists, etc. It is recommended that the length of the sequence does not exceed 50000 bp. Compare the CGRs of the difference species visually? Are there any species in which their CGRs are similar to one another? How about species in which their CGRs are most different from each other? <a class="reference internal" href="../Solutions/Reports%20on%20Chaos%20Game%20Activities.html#cgractivityreport-4"><span class="std std-ref">[What happened when we did this]</span></a></p>
</div>
</section>
<section id="map-of-life">
<h3>Map of Life<a class="headerlink" href="#map-of-life" title="Permalink to this headline">#</a></h3>
<p>Taking DNA sequence visualization one step further, the authors of <span id="id30">[<a class="reference internal" href="../references.html#id32" title="L. Kari, K. A. Hill, A. S. Sayem, N. Bryans, K. Davis, and N. S. Dattani. Map of life: measuring and visualizing species' relatedness with &quot;molecular distance maps&quot;. arXiv preprint arXiv:1307.3755, 2013.">Kari <em>et al.</em>, 2013</a>]</span> proposed a novel combination of methods to create <em>Molecular Distance Maps</em>. Molecular distance maps visually illustrate the quantitative relationships and patterns of proximities among the given genetic sequences and among the species they represent. To compute and visually display relationships between DNA sequences, the three techniques that were used include chaos game representation, structural dissimilarity index (DSSIM), and multi-dimensional scaling. In the paper <span id="id31">[<a class="reference internal" href="../references.html#id32" title="L. Kari, K. A. Hill, A. S. Sayem, N. Bryans, K. Davis, and N. S. Dattani. Map of life: measuring and visualizing species' relatedness with &quot;molecular distance maps&quot;. arXiv preprint arXiv:1307.3755, 2013.">Kari <em>et al.</em>, 2013</a>]</span>, this method was applied to a variety of cases that have been historically controversial and was there demonstrated to have the potential for taxonomical clarification. In the following, we will discuss the methods used from this paper, more particularly the structural dissimilarity index and the multi-dimensional scaling as we have already looked into the chaos game representation for nucleotides.</p>
<p>To understand the structural dissimilarity index, we will first have to explain what the structural similarity index is. The structural similarity (SSIM) index is a method for measuring the similarity between two images based on the computation of three terms, namely luminance distortion, contrast distortion, and linear correlation <span id="id32">[<a class="reference internal" href="../references.html#id45" title="Z. Wang, A. C. Bovik, H. R. Sheikh, and E. P. Simoncelli. Image quality assessment: from error visibility to structural similarity. IEEE trans. image process., 13(4):600–612, 2004.">Wang <em>et al.</em>, 2004</a>]</span>. It was designed to perform similarly to the human visual system, which is highly adapted to extract structural information. The overall index is a multiplicative combination of the three terms:</p>
<div class="amsmath math notranslate nohighlight">
\[\begin{equation*}
	\operatorname{SSIM}(x, y) = [l(x, y)]^\alpha\cdot[c(x, y)]^\beta\cdot[s(x, y)]^\gamma ,
\end{equation*}\]</div>
<p>where</p>
<div class="math notranslate nohighlight">
\[\begin{split}
\begin{align}
	l(x, y) &amp;= \dfrac{2\mu_{x}\mu_{y}+C_1}{\mu_{x}^{2} + \mu_{y}^{2} + C_1} \\
	c(x, y) &amp;= \dfrac{2\sigma_{x}\sigma_{y}+C_{2}}{\sigma_{x}^{2} + \sigma_{y}^{2} + C_2} \\
	s(x, y) &amp;= \dfrac{\sigma_{xy} + C_{3}}{\sigma_{x}\sigma_{y} + C_{3}} ,
\end{align}
\end{split}\]</div>
<p>where <span class="math notranslate nohighlight">\(\mu_{x}\)</span>, <span class="math notranslate nohighlight">\(\mu_{y}\)</span>, <span class="math notranslate nohighlight">\(\sigma_{x}\)</span>, <span class="math notranslate nohighlight">\(\sigma_{y}\)</span> and <span class="math notranslate nohighlight">\(\sigma_{xy}\)</span> are the local means, standard deviations, and cross-covariance for images <span class="math notranslate nohighlight">\(x\)</span> and <span class="math notranslate nohighlight">\(y\)</span>. <span class="math notranslate nohighlight">\(C_{1}\)</span>, <span class="math notranslate nohighlight">\(C_{2}\)</span>, and <span class="math notranslate nohighlight">\(C_{3}\)</span> are the regularization constants for the luminance, contrast, and structural terms, respectively. If <span class="math notranslate nohighlight">\(\alpha = \beta = \gamma = 1\)</span>, and <span class="math notranslate nohighlight">\(C_{3} = \frac{C_{2}}{2}\)</span>, the index simplifies to:</p>
<div class="amsmath math notranslate nohighlight">
\[\begin{equation*}
	\operatorname{SSIM}(x, y) = \dfrac{(2\mu_{x}\mu_{y} + C_{1})(2\sigma_{xy} + C_{2})}{(\mu_{x}^{2} + \mu_{y}^{2} + C_{1})(\sigma_{x}^{2} + \sigma_{y}^{2} + C_{2})} .
\end{equation*}\]</div>
<p>The theoretical range of <span class="math notranslate nohighlight">\(\operatorname{SSIM}(x, y) \in [-1, 1]\)</span>, where the high value indicates high similarity. In <span id="id33">[<a class="reference internal" href="../references.html#id32" title="L. Kari, K. A. Hill, A. S. Sayem, N. Bryans, K. Davis, and N. S. Dattani. Map of life: measuring and visualizing species' relatedness with &quot;molecular distance maps&quot;. arXiv preprint arXiv:1307.3755, 2013.">Kari <em>et al.</em>, 2013</a>]</span>, instead of calculating the overall SSIM, they computed the local SSIM value for each pixel in the image <span class="math notranslate nohighlight">\(x\)</span>. They refer to this as a distance matrix. The SSIM index can be computed in Python by using the function <code class="docutils literal notranslate"><span class="pre">structural_similarity</span></code> from the <code class="docutils literal notranslate"><span class="pre">skimage.metrics</span></code> library.</p>
<p>Now that we know how to compute the SSIM index, we are able to compute the structural dissimilarity (DSSIM) index:</p>
<div class="amsmath math notranslate nohighlight">
\[\begin{equation*}
	\operatorname{DSSIM}(x, y) = 1 - \operatorname{SSIM}(x, y) ,
\end{equation*}\]</div>
<p>whose theoretical range is <span class="math notranslate nohighlight">\([0, 2]\)</span> with the distance being 0 between two identical images, and 2 if the two images are negatively correlated. An example of two images that are negatively correlated would be if one is completely white, while the other is completely black. Typically, the range that DSSIM falls with is <span class="math notranslate nohighlight">\([0, 1]\)</span>—the authors from <span id="id34">[<a class="reference internal" href="../references.html#id33" title="L. Kari, K. A. Hill, A. S. Sayem, R. Karamichalis, N. Bryans, K. Davis, and N. S. Dattani. Mapping the space of genomic signatures. PloS one, 10(5):e0119815, 2015.">Kari <em>et al.</em>, 2015</a>]</span> noted that almost all (over 5 million) distances that found were between <span class="math notranslate nohighlight">\(0\)</span> and <span class="math notranslate nohighlight">\(1\)</span>, with only half a dozen exceptions of distances between <span class="math notranslate nohighlight">\(1\)</span> and <span class="math notranslate nohighlight">\(1.0033\)</span>.</p>
<p>To demonstrate this, let us compare the “anatomically modern” human (Homo sapiens sapiens) mitochondrial DNA and the neanderthal (Homo sapiens neanderthalensis) mitochondrial DNA (mtDNA). As seen from their scientific classification, we can see that these two species are from the same genus, which suggests that their DNA should be similar. The following shows the CGRs of the “anatomically modern” human (left) and of the neanderthal (right). We can see that the two CGRs are quite similar (which is agrees with the fact that the two species are from the same genus) and therefore, we expect the the DSSIM to be small.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">human_dna</span> <span class="o">=</span> <span class="n">parse_sequence</span><span class="p">(</span><span class="s1">&#39;../Supplementary Material/CGR/FASTA Files/NC_012920.fasta&#39;</span><span class="p">)</span>
<span class="n">human_coord</span> <span class="o">=</span> <span class="n">dna_cgr</span><span class="p">(</span><span class="n">human_dna</span><span class="p">)</span>

<span class="n">neanderthal_dna</span> <span class="o">=</span> <span class="n">parse_sequence</span><span class="p">(</span><span class="s1">&#39;../Supplementary Material/CGR/FASTA Files/NC_011137.fasta&#39;</span><span class="p">)</span>
<span class="n">neanderthal_coord</span> <span class="o">=</span> <span class="n">dna_cgr</span><span class="p">(</span><span class="n">neanderthal_dna</span><span class="p">)</span>

<span class="n">fig</span><span class="p">,</span> <span class="p">(</span><span class="n">ax1</span><span class="p">,</span> <span class="n">ax2</span><span class="p">)</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">16</span><span class="p">,</span> <span class="mi">6</span><span class="p">))</span>
<span class="n">fig</span><span class="o">.</span><span class="n">suptitle</span><span class="p">(</span><span class="s1">&#39;CGRs of human vs neanderthal mitochondrial DNA sequences&#39;</span><span class="p">)</span>
<span class="n">ax1</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">human_coord</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="p">:],</span> <span class="n">human_coord</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="p">:],</span> <span class="s1">&#39;k.&#39;</span><span class="p">,</span> <span class="n">markersize</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="n">ax1</span><span class="o">.</span><span class="n">axes</span><span class="o">.</span><span class="n">xaxis</span><span class="o">.</span><span class="n">set_visible</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
<span class="n">ax1</span><span class="o">.</span><span class="n">axes</span><span class="o">.</span><span class="n">yaxis</span><span class="o">.</span><span class="n">set_visible</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
<span class="n">ax1</span><span class="o">.</span><span class="n">set_xlim</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">])</span>
<span class="n">ax1</span><span class="o">.</span><span class="n">set_ylim</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">])</span>
<span class="n">ax1</span><span class="o">.</span><span class="n">set_aspect</span><span class="p">(</span><span class="s1">&#39;equal&#39;</span><span class="p">,</span> <span class="s1">&#39;box&#39;</span><span class="p">)</span>
<span class="n">ax1</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;Human (Homo sapiens) mitochondrial DNA&#39;</span><span class="p">)</span>

<span class="n">ax2</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">neanderthal_coord</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="p">:],</span> <span class="n">neanderthal_coord</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="p">:],</span> <span class="s1">&#39;k.&#39;</span><span class="p">,</span> <span class="n">markersize</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="n">ax2</span><span class="o">.</span><span class="n">axes</span><span class="o">.</span><span class="n">xaxis</span><span class="o">.</span><span class="n">set_visible</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
<span class="n">ax2</span><span class="o">.</span><span class="n">axes</span><span class="o">.</span><span class="n">yaxis</span><span class="o">.</span><span class="n">set_visible</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
<span class="n">ax2</span><span class="o">.</span><span class="n">set_xlim</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">])</span>
<span class="n">ax2</span><span class="o">.</span><span class="n">set_ylim</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">])</span>
<span class="n">ax2</span><span class="o">.</span><span class="n">set_aspect</span><span class="p">(</span><span class="s1">&#39;equal&#39;</span><span class="p">,</span> <span class="s1">&#39;box&#39;</span><span class="p">)</span>
<span class="n">ax2</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;Neanderthal (Homo sapiens neanderthalensis) mitochondrial DNA&#39;</span><span class="p">)</span>

<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/chaos-game-representation_46_0.png" src="../_images/chaos-game-representation_46_0.png" />
</div>
</div>
<p>In Python, in order to compare these two plots, we first need to convert each plot into a matrix. The size of the matrix is determined by the dimension of the plot (number of pixels), and each element within the matrix represents each pixel in the plot, where <span class="math notranslate nohighlight">\(0\)</span> represents a white pixel, <span class="math notranslate nohighlight">\(1\)</span> represents a black pixel, and any value in between represents a pixel that is a shade of grey. The simplest way to convert the plot to a matrix is to first save the image as a .png file. In the following code block, we show the way that we like to save the figures—the additional settings in the <code class="docutils literal notranslate"><span class="pre">savefig()</span></code> call removes the white border around the plot.</p>
<div class="cell tag_remove-output docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># In order to find SSIM, need to save figures as an image file. </span>
<span class="c1"># Below is an example of how you would save your figures as a png file.</span>
<span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">human_coord</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="p">:],</span> <span class="n">human_coord</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="p">:],</span> <span class="s1">&#39;k.&#39;</span><span class="p">,</span> <span class="n">markersize</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xticks</span><span class="p">([])</span>
<span class="n">plt</span><span class="o">.</span><span class="n">yticks</span><span class="p">([])</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xlim</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">])</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylim</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">])</span>
<span class="n">plt</span><span class="o">.</span><span class="n">gca</span><span class="p">()</span><span class="o">.</span><span class="n">set_aspect</span><span class="p">(</span><span class="s1">&#39;equal&#39;</span><span class="p">,</span> <span class="s1">&#39;box&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="s1">&#39;../Supplementary Material/CGR/Output/human.png&#39;</span><span class="p">,</span> <span class="n">bbox_inches</span><span class="o">=</span><span class="s1">&#39;tight&#39;</span><span class="p">,</span> <span class="n">pad_inches</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># In order to find SSIM, need to save figures as an image file. </span>
<span class="c1"># Below is an example of how you would save your figures as a png file.</span>
<span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">neanderthal_coord</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="p">:],</span> <span class="n">neanderthal_coord</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="p">:],</span> <span class="s1">&#39;k.&#39;</span><span class="p">,</span> <span class="n">markersize</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xticks</span><span class="p">([])</span>
<span class="n">plt</span><span class="o">.</span><span class="n">yticks</span><span class="p">([])</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xlim</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">])</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylim</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">])</span>
<span class="n">plt</span><span class="o">.</span><span class="n">gca</span><span class="p">()</span><span class="o">.</span><span class="n">set_aspect</span><span class="p">(</span><span class="s1">&#39;equal&#39;</span><span class="p">,</span> <span class="s1">&#39;box&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="s1">&#39;../Supplementary Material/CGR/Output/neanderthal.png&#39;</span><span class="p">,</span> <span class="n">bbox_inches</span><span class="o">=</span><span class="s1">&#39;tight&#39;</span><span class="p">,</span> <span class="n">pad_inches</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/chaos-game-representation_49_0.png" src="../_images/chaos-game-representation_49_0.png" />
</div>
</div>
<p>Once we have saved the images as a .png file, we use Python’s <code class="docutils literal notranslate"><span class="pre">skimage.io.imread()</span></code> function to read in the images as matrices. We also specify in the parameters of this function that this image is greyscale, which returns an <span class="math notranslate nohighlight">\(m\times n\)</span> matrix (<span class="math notranslate nohighlight">\(m\)</span> and <span class="math notranslate nohighlight">\(n\)</span> are dependent on the dimensions of the image). If the parameter <code class="docutils literal notranslate"><span class="pre">as_gray</span></code> was not set to <code class="docutils literal notranslate"><span class="pre">True</span></code>, then the output would be a <span class="math notranslate nohighlight">\(m \times n \times 3\)</span>, which accounts for the different color channels (RGB).</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">human_CGR</span> <span class="o">=</span> <span class="n">skimage</span><span class="o">.</span><span class="n">io</span><span class="o">.</span><span class="n">imread</span><span class="p">(</span><span class="s1">&#39;../Supplementary Material/CGR/Output/human.png&#39;</span><span class="p">,</span> <span class="n">as_gray</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="n">neanderthal_CGR</span> <span class="o">=</span> <span class="n">skimage</span><span class="o">.</span><span class="n">io</span><span class="o">.</span><span class="n">imread</span><span class="p">(</span><span class="s1">&#39;../Supplementary Material/CGR/Output/neanderthal.png&#39;</span><span class="p">,</span> <span class="n">as_gray</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>Now that the two images have been converted to matrices, we use Python’s <code class="docutils literal notranslate"><span class="pre">structural_similarity</span></code> function from the <code class="docutils literal notranslate"><span class="pre">skimage.metrics</span></code> library (aliased as <code class="docutils literal notranslate"><span class="pre">ssim</span></code> below) and calculate the structural similarity index between the two images.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">human_neanderthal_ssim</span> <span class="o">=</span> <span class="n">ssim</span><span class="p">(</span><span class="n">human_CGR</span><span class="p">,</span> <span class="n">neanderthal_CGR</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">human_neanderthal_ssim</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>0.8968926645376538
</pre></div>
</div>
</div>
</div>
<p>and therefore, the structural dissimilarity index is then</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="mi">1</span> <span class="o">-</span> <span class="n">human_neanderthal_ssim</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>0.10310733546234618
</pre></div>
</div>
</div>
</div>
<p>which agrees with our expectations.</p>
<p>In the following, we look at another example that compares the mtDNA of humans to other species, such as the great spotted kiwi (left) and pearlfish (right).</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">kiwi_dna</span> <span class="o">=</span> <span class="n">parse_sequence</span><span class="p">(</span><span class="s1">&#39;../Supplementary Material/CGR/FASTA Files/NC_002782.fasta&#39;</span><span class="p">)</span>
<span class="n">kiwi_coord</span> <span class="o">=</span> <span class="n">dna_cgr</span><span class="p">(</span><span class="n">kiwi_dna</span><span class="p">)</span>

<span class="n">pearlfish_dna</span> <span class="o">=</span> <span class="n">parse_sequence</span><span class="p">(</span><span class="s1">&#39;../Supplementary Material/CGR/FASTA Files/NC_004373.fasta&#39;</span><span class="p">)</span>
<span class="n">pearlfish_coord</span> <span class="o">=</span> <span class="n">dna_cgr</span><span class="p">(</span><span class="n">pearlfish_dna</span><span class="p">)</span>

<span class="n">fig</span><span class="p">,</span> <span class="p">(</span><span class="n">ax1</span><span class="p">,</span> <span class="n">ax2</span><span class="p">)</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">16</span><span class="p">,</span> <span class="mi">6</span><span class="p">))</span>
<span class="n">fig</span><span class="o">.</span><span class="n">suptitle</span><span class="p">(</span><span class="s1">&#39;CGRs of Great Spotted Kiwi vs Pearlfish mitochondrial DNA sequences&#39;</span><span class="p">)</span>
<span class="n">ax1</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">kiwi_coord</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="p">:],</span> <span class="n">kiwi_coord</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="p">:],</span> <span class="s1">&#39;k.&#39;</span><span class="p">,</span> <span class="n">markersize</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="n">ax1</span><span class="o">.</span><span class="n">axes</span><span class="o">.</span><span class="n">xaxis</span><span class="o">.</span><span class="n">set_visible</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
<span class="n">ax1</span><span class="o">.</span><span class="n">axes</span><span class="o">.</span><span class="n">yaxis</span><span class="o">.</span><span class="n">set_visible</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
<span class="n">ax1</span><span class="o">.</span><span class="n">set_xlim</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">])</span>
<span class="n">ax1</span><span class="o">.</span><span class="n">set_ylim</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">])</span>
<span class="n">ax1</span><span class="o">.</span><span class="n">set_aspect</span><span class="p">(</span><span class="s1">&#39;equal&#39;</span><span class="p">,</span> <span class="s1">&#39;box&#39;</span><span class="p">)</span>
<span class="n">ax1</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;Great spotted kiwi (Apteryx haastii) mitochondrial DNA&#39;</span><span class="p">)</span>

<span class="n">ax2</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">pearlfish_coord</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="p">:],</span> <span class="n">pearlfish_coord</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="p">:],</span> <span class="s1">&#39;k.&#39;</span><span class="p">,</span> <span class="n">markersize</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="n">ax2</span><span class="o">.</span><span class="n">axes</span><span class="o">.</span><span class="n">xaxis</span><span class="o">.</span><span class="n">set_visible</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
<span class="n">ax2</span><span class="o">.</span><span class="n">axes</span><span class="o">.</span><span class="n">yaxis</span><span class="o">.</span><span class="n">set_visible</span><span class="p">(</span><span class="kc">False</span><span class="p">)</span>
<span class="n">ax2</span><span class="o">.</span><span class="n">set_xlim</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">])</span>
<span class="n">ax2</span><span class="o">.</span><span class="n">set_ylim</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">])</span>
<span class="n">ax2</span><span class="o">.</span><span class="n">set_aspect</span><span class="p">(</span><span class="s1">&#39;equal&#39;</span><span class="p">,</span> <span class="s1">&#39;box&#39;</span><span class="p">)</span>
<span class="n">ax2</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;Pearlfish (Carapus bermudensis) mitochondrial DNA&#39;</span><span class="p">)</span>

<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/chaos-game-representation_58_0.png" src="../_images/chaos-game-representation_58_0.png" />
</div>
</div>
<p>Just visually, the CGR of the mtDNA of the great spotted kiwi looks more similar to the CGRs of the human and neanderthal than the CGR of the pearlfish. Using all of the DSSIM, we can produce what is called a distance matrix, which is a square, real, symmetric matrix where each element is the structural dissimilarity index of two species with corresponds to the row and column. The distance matrix for the four species that we have already produced the CGR for is</p>
<!-- $$
    \mathrm{DSSIM}(x, y) =
    \begin{bmatrix}
        0 & 0.1223 & 0.7823 & 0.8541 \\
        0.1223 & 0 & 0.7821 & 0.8533 \\
        0.7823 & 0.7821 & 0 & 0.8519 \\
        0.8541 & 0.8533 & 0.8519 & 0
    \end{bmatrix} ,
$$ --><div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">dist_matrix</span><span class="p">(</span><span class="n">cgr_list</span><span class="p">):</span>
    <span class="n">n</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">cgr_list</span><span class="p">)</span>
    <span class="n">dmat</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="n">n</span><span class="p">,</span> <span class="n">n</span><span class="p">))</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">n</span><span class="p">):</span>
        <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">i</span><span class="o">+</span><span class="mi">1</span><span class="p">,</span> <span class="n">n</span><span class="p">):</span>
            <span class="n">dmat</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="n">j</span><span class="p">]</span> <span class="o">=</span> <span class="mi">1</span> <span class="o">-</span> <span class="n">ssim</span><span class="p">(</span><span class="n">cgr_list</span><span class="p">[</span><span class="n">i</span><span class="p">],</span> <span class="n">cgr_list</span><span class="p">[</span><span class="n">j</span><span class="p">])</span>
            <span class="n">dmat</span><span class="p">[</span><span class="n">j</span><span class="p">,</span> <span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="n">dmat</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="n">j</span><span class="p">]</span>
    <span class="k">return</span><span class="p">(</span><span class="n">dmat</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">kiwi_CGR</span> <span class="o">=</span> <span class="n">skimage</span><span class="o">.</span><span class="n">io</span><span class="o">.</span><span class="n">imread</span><span class="p">(</span><span class="s1">&#39;../Supplementary Material/CGR/Output/greatspottedkiwi.png&#39;</span><span class="p">,</span> <span class="n">as_gray</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="n">pearlfish_CGR</span> <span class="o">=</span> <span class="n">skimage</span><span class="o">.</span><span class="n">io</span><span class="o">.</span><span class="n">imread</span><span class="p">(</span><span class="s1">&#39;../Supplementary Material/CGR/Output/pearlfish.png&#39;</span><span class="p">,</span> <span class="n">as_gray</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

<span class="n">CGR_list</span> <span class="o">=</span> <span class="p">[</span><span class="n">human_CGR</span><span class="p">,</span> <span class="n">neanderthal_CGR</span><span class="p">,</span> <span class="n">kiwi_CGR</span><span class="p">,</span> <span class="n">pearlfish_CGR</span><span class="p">]</span>

<span class="n">dssim</span> <span class="o">=</span> <span class="n">dist_matrix</span><span class="p">(</span><span class="n">CGR_list</span><span class="p">)</span>
<span class="n">dssim</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>array([[0.        , 0.10310734, 0.76524434, 0.85537312],
       [0.10310734, 0.        , 0.76727659, 0.85644889],
       [0.76524434, 0.76727659, 0.        , 0.85483255],
       [0.85537312, 0.85644889, 0.85483255, 0.        ]])
</pre></div>
</div>
</div>
</div>
<p>where the first row/column represents humans, the second row/column represents the neanderthals, the third row/column represents the great spotted kiwi, and lastly, the fourth row/column represents the pearlfish. Obviously, the diagonal of the matrix would be <span class="math notranslate nohighlight">\(0\)</span>, since the DSSIM of two of the same images would produce a result of 0. As we have seen previously, the <span class="math notranslate nohighlight">\(\mathrm{DSSIM}(\mathrm{human},\ \mathrm{neanderthal}) = 0.10310734\)</span>, and vice versa. From this matrix, we can see that the CGR of the mtDNA of pearlfish is most different than the rest—this agrees with what we have observed.</p>
<p>Multi-dimensional scaling (MDS) is a means of visualizing the level of similarity or dissimilarity of individual case of a dataset. It has been used in various fields such as cognitive science, information science, pychometrics, marketing, ecology, social science and other areas of study <span id="id35">[<a class="reference internal" href="../references.html#id16" title="I. Borg and P. J. F. Groenen. Modern multidimensional scaling: Theory and applications. Springer Science &amp; Business Media, 2005.">Borg and Groenen, 2005</a>]</span>. The goal of MDS is to find a spatial configuration of objects when all that is known is some measure of their general similarity or dissimilarity <span id="id36">[<a class="reference internal" href="../references.html#id46" title="F. Wickelmaier. An introduction to MDS. Sound Quality Research Unit, Aalborg University, Denmark, 2003.">Wickelmaier, 2003</a>]</span>. Here, the MDS takes in the distance matrix (of size <span class="math notranslate nohighlight">\(n\)</span> by <span class="math notranslate nohighlight">\(n\)</span>) and outputs a <span class="math notranslate nohighlight">\(n\)</span> by <span class="math notranslate nohighlight">\(p\)</span> configuration matrix, where each row of this matrix represents the coordinates of the item in <span class="math notranslate nohighlight">\(p\)</span> dimension. In this unit, we will only be looking at <span class="math notranslate nohighlight">\(p = 2\)</span> and <span class="math notranslate nohighlight">\(p = 3\)</span> as it is difficult to create visualization in dimensions greater than 3. In <span id="id37">[<a class="reference internal" href="../references.html#id32" title="L. Kari, K. A. Hill, A. S. Sayem, N. Bryans, K. Davis, and N. S. Dattani. Map of life: measuring and visualizing species' relatedness with &quot;molecular distance maps&quot;. arXiv preprint arXiv:1307.3755, 2013.">Kari <em>et al.</em>, 2013</a>]</span>, the authors used a classical MDS, which assumes that all the distances (from the distance matrix) are Euclidean. For the algorithm for classical MDS, refer to <span id="id38">[<a class="reference internal" href="../references.html#id46" title="F. Wickelmaier. An introduction to MDS. Sound Quality Research Unit, Aalborg University, Denmark, 2003.">Wickelmaier, 2003</a>]</span>(p. 10). Although Python’s <code class="docutils literal notranslate"><span class="pre">scikit-learn</span></code> library has a MDS function (<code class="docutils literal notranslate"><span class="pre">sklearn.manifold.MDS()</span></code>), we provide a different function that performs classical multidimensional scaling below. The following function computes the coordinates of the points under the new map much quicker than the MDS function in the <code class="docutils literal notranslate"><span class="pre">scikit-learn</span></code> library.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Function taken from https://pyseer.readthedocs.io/en/master/_modules/pyseer/cmdscale.html</span>
<span class="k">def</span> <span class="nf">cmdscale</span><span class="p">(</span><span class="n">D</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Classical multidimensional scaling (MDS)</span>

<span class="sd">    Args:</span>
<span class="sd">        D (numpy.array)</span>
<span class="sd">            Symmetric distance matrix (n, n)</span>

<span class="sd">    Returns:</span>
<span class="sd">        Y (numpy.array)</span>
<span class="sd">            Configuration matrix (n, p). Each column represents a dimension. Only the</span>
<span class="sd">            p dimensions corresponding to positive eigenvalues of B are returned.</span>
<span class="sd">            Note that each dimension is only determined up to an overall sign,</span>
<span class="sd">            corresponding to a reflection.</span>
<span class="sd">        e (numpy.array)</span>
<span class="sd">            Eigenvalues of B (n, 1)</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Number of points</span>
    <span class="n">n</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">D</span><span class="p">)</span>

    <span class="c1"># Centering matrix</span>
    <span class="n">H</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">eye</span><span class="p">(</span><span class="n">n</span><span class="p">)</span> <span class="o">-</span> <span class="n">np</span><span class="o">.</span><span class="n">ones</span><span class="p">((</span><span class="n">n</span><span class="p">,</span> <span class="n">n</span><span class="p">))</span><span class="o">/</span><span class="n">n</span>

    <span class="c1"># YY^T</span>
    <span class="n">B</span> <span class="o">=</span> <span class="o">-</span><span class="n">H</span><span class="o">.</span><span class="n">dot</span><span class="p">(</span><span class="n">D</span><span class="o">**</span><span class="mi">2</span><span class="p">)</span><span class="o">.</span><span class="n">dot</span><span class="p">(</span><span class="n">H</span><span class="p">)</span><span class="o">/</span><span class="mi">2</span>

    <span class="c1"># Diagonalize</span>
    <span class="n">evals</span><span class="p">,</span> <span class="n">evecs</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">linalg</span><span class="o">.</span><span class="n">eigh</span><span class="p">(</span><span class="n">B</span><span class="p">)</span>

    <span class="c1"># Sort by eigenvalue in descending order</span>
    <span class="n">idx</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">argsort</span><span class="p">(</span><span class="n">evals</span><span class="p">)[::</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
    <span class="n">evals</span> <span class="o">=</span> <span class="n">evals</span><span class="p">[</span><span class="n">idx</span><span class="p">]</span>
    <span class="n">evecs</span> <span class="o">=</span> <span class="n">evecs</span><span class="p">[:,</span> <span class="n">idx</span><span class="p">]</span>

    <span class="c1"># Compute the coordinates using positive-eigenvalued components only</span>
    <span class="n">w</span><span class="p">,</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="n">evals</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">)</span>
    <span class="n">L</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">diag</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">sqrt</span><span class="p">(</span><span class="n">evals</span><span class="p">[</span><span class="n">w</span><span class="p">]))</span>
    <span class="n">V</span> <span class="o">=</span> <span class="n">evecs</span><span class="p">[:,</span> <span class="n">w</span><span class="p">]</span>
    <span class="n">Y</span> <span class="o">=</span> <span class="n">V</span><span class="o">.</span><span class="n">dot</span><span class="p">(</span><span class="n">L</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">Y</span><span class="p">,</span> <span class="n">evals</span><span class="p">[</span><span class="n">evals</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">]</span>
</pre></div>
</div>
</div>
</div>
<p>We perform the classical MDS on the dissimilarity matrix <code class="docutils literal notranslate"><span class="pre">dssim</span></code> calculated above. Even though the <code class="docutils literal notranslate"><span class="pre">cmdscale</span></code> function returns two outputs, we only require the first output (the configuration matrix) for our visualization. We can also converted the matrix into a pandas dataframe to clarify which row corresponds which animal.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">Y</span><span class="p">,</span> <span class="n">_</span> <span class="o">=</span> <span class="n">cmdscale</span><span class="p">(</span><span class="n">dssim</span><span class="p">)</span>
<span class="n">Y_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">Y</span><span class="p">,</span> <span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;x&#39;</span><span class="p">,</span> <span class="s1">&#39;y&#39;</span><span class="p">,</span> <span class="s1">&#39;z&#39;</span><span class="p">])</span>
<span class="n">Y_df</span><span class="p">[</span><span class="s1">&#39;animal&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;human&#39;</span><span class="p">,</span> <span class="s1">&#39;neanderthal&#39;</span><span class="p">,</span> <span class="s1">&#39;great spotted kiwi&#39;</span><span class="p">,</span> <span class="s1">&#39;pearlfish&#39;</span><span class="p">]</span>
<span class="n">Y_df</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>x</th>
      <th>y</th>
      <th>z</th>
      <th>animal</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>-0.326245</td>
      <td>0.107564</td>
      <td>0.051616</td>
      <td>human</td>
    </tr>
    <tr>
      <th>1</th>
      <td>-0.327661</td>
      <td>0.109078</td>
      <td>-0.051471</td>
      <td>neanderthal</td>
    </tr>
    <tr>
      <th>2</th>
      <td>0.143462</td>
      <td>-0.494346</td>
      <td>-0.000133</td>
      <td>great spotted kiwi</td>
    </tr>
    <tr>
      <th>3</th>
      <td>0.510445</td>
      <td>0.277704</td>
      <td>-0.000012</td>
      <td>pearlfish</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<p>To plot the points in in 2 dimensions, we use the first two columns as our <span class="math notranslate nohighlight">\(x\)</span> and <span class="math notranslate nohighlight">\(y\)</span> coordinates, respectively. Here, we decide to use the <code class="docutils literal notranslate"><span class="pre">seaborn</span></code> data visualization library (which is based on <code class="docutils literal notranslate"><span class="pre">matplotlib</span></code>) as this library specializes in dealing with data stored in a dataframe.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">()</span>
<span class="n">fig</span><span class="o">.</span><span class="n">set_size_inches</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">6</span><span class="p">)</span>
<span class="n">sns</span><span class="o">.</span><span class="n">scatterplot</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="s1">&#39;x&#39;</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="s1">&#39;y&#39;</span><span class="p">,</span> <span class="n">hue</span><span class="o">=</span><span class="s1">&#39;animal&#39;</span><span class="p">,</span> <span class="n">data</span><span class="o">=</span><span class="n">Y_df</span><span class="p">)</span>
<span class="n">legend</span> <span class="o">=</span> <span class="n">ax</span><span class="o">.</span><span class="n">legend</span><span class="p">()</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/chaos-game-representation_68_0.png" src="../_images/chaos-game-representation_68_0.png" />
</div>
</div>
<p>From the plot above, we are able to observe the degree of difference between the four CGRs by the distance between the four points. Since the human and neanderthal points overlap each other, this indicates that their CGRs are very similar, thus implying that the mtDNA of human and neanderthals are also similar. If we were interested in knowing whether humans are more similar to the great spotted kiwi or pearlfish, we can easily see from the plot above that humans are closer related to the great spotted kiwi genetically since the distance between the human and great spotted kiwi points are closer than the distance between the human and pearlfish points.</p>
<p>Now that we understand the idea of using classical multidimensional scaling, we can apply this method to a large set of animals (in this case 4844). This will allow us to visualize how similar the animals are genetically to each other, which could potentially help with classifying animals in which their class is ambiguous.</p>
<p>Since it takes an extremely long time to calculate the dissimilarity matrix for 4844 vertebrate animals (it took us around 8 hours to compute the dissimilarity matrix in parallel with 12 threads, using an AMD Ryzen 5 3600X 6-Core Processor), we have already calculated it for you and provided the data file, which can be found in the <code class="docutils literal notranslate"><span class="pre">Supplementary</span> <span class="pre">Material</span></code> folder in our Github repository. The code used to create the dissimilarity matrix is provided in the <code class="docutils literal notranslate"><span class="pre">code</span></code> folder of our repository (read the <code class="docutils literal notranslate"><span class="pre">README.md</span></code> file for instructions on the order to run the code). Now, let us apply the classical MDS to the dissimilarity matrix provided.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Load list of animals and their classes. This will be used to colour the points on the scatter plot</span>
<span class="n">file</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="s2">&quot;../Supplementary Material/CGR/Data Files/animal_class.pkl&quot;</span><span class="p">,</span> <span class="s2">&quot;rb&quot;</span><span class="p">)</span>
<span class="n">animal_list</span> <span class="o">=</span> <span class="n">pickle</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">file</span><span class="p">)</span>
<span class="n">file</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
<span class="n">animal_class</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">animal_list</span><span class="o">.</span><span class="n">values</span><span class="p">())</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">DSSIM</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="s1">&#39;../Supplementary Material/CGR/Data Files/DSSIM_matrix.dat&#39;</span><span class="p">,</span> <span class="n">allow_pickle</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="n">coord</span><span class="p">,</span> <span class="n">_</span> <span class="o">=</span> <span class="n">cmdscale</span><span class="p">(</span><span class="n">DSSIM</span><span class="p">)</span>
<span class="n">coord_splice</span> <span class="o">=</span> <span class="n">coord</span><span class="p">[:,</span> <span class="p">:</span><span class="mi">3</span><span class="p">]</span>
<span class="n">cmdscaling_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">data</span> <span class="o">=</span> <span class="n">coord_splice</span><span class="p">,</span> <span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;x&#39;</span><span class="p">,</span> <span class="s1">&#39;y&#39;</span><span class="p">,</span> <span class="s1">&#39;z&#39;</span><span class="p">])</span>
<span class="n">cmdscaling_df</span><span class="p">[</span><span class="s1">&#39;class&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">animal_class</span>
<span class="n">cmdscaling_df</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>x</th>
      <th>y</th>
      <th>z</th>
      <th>class</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>-0.044784</td>
      <td>-0.017598</td>
      <td>0.023091</td>
      <td>amphibians</td>
    </tr>
    <tr>
      <th>1</th>
      <td>-0.040420</td>
      <td>0.012079</td>
      <td>0.012869</td>
      <td>amphibians</td>
    </tr>
    <tr>
      <th>2</th>
      <td>-0.044732</td>
      <td>-0.027890</td>
      <td>0.028080</td>
      <td>amphibians</td>
    </tr>
    <tr>
      <th>3</th>
      <td>-0.023975</td>
      <td>0.015180</td>
      <td>0.030752</td>
      <td>amphibians</td>
    </tr>
    <tr>
      <th>4</th>
      <td>-0.053848</td>
      <td>-0.068480</td>
      <td>0.015756</td>
      <td>amphibians</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">fig</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">11</span><span class="p">,</span> <span class="mi">8</span><span class="p">))</span>
<span class="n">ax</span> <span class="o">=</span> <span class="n">sns</span><span class="o">.</span><span class="n">scatterplot</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="s1">&#39;x&#39;</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="s1">&#39;y&#39;</span><span class="p">,</span> <span class="n">hue</span><span class="o">=</span><span class="s1">&#39;class&#39;</span><span class="p">,</span> <span class="n">data</span><span class="o">=</span><span class="n">cmdscaling_df</span><span class="p">)</span>
<span class="n">legend</span> <span class="o">=</span> <span class="n">ax</span><span class="o">.</span><span class="n">legend</span><span class="p">()</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/chaos-game-representation_72_0.png" src="../_images/chaos-game-representation_72_0.png" />
</div>
</div>
<p>The plot above shows a molecular distance map of 4844 animals from various classes from the chordate phylum (meaning that all these animals have a vertebrate), coloured according to their class. The sequences were taken from NCBI Reference Sequence Database (RefSeq). We can see from this figure that this method (Map of Life) does quite a good job in sorting species into different categories. One particular aspect that we want to point out in this figure is where the points representing the lungfish are—they are in the area of where the bony fish, cartilage fish, and amphibians touch. This is truly remarkable as lung fish has qualities of all these classes.</p>
<p>Since the points overlap one another, we could also plot the points in 3 dimensions.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Since for 3D scatter plot, we need to plot each class of animals separately, we pre-assign colours to each animal class</span>
<span class="n">animal_class_unique</span> <span class="o">=</span> <span class="n">cmdscaling_df</span><span class="p">[</span><span class="s1">&#39;class&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">()</span>
<span class="n">cmap</span> <span class="o">=</span> <span class="n">sns</span><span class="o">.</span><span class="n">color_palette</span><span class="p">(</span><span class="s2">&quot;husl&quot;</span><span class="p">,</span> <span class="mi">12</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># %matplotlib notebook</span>
<span class="n">fig</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">11</span><span class="p">,</span> <span class="mi">8</span><span class="p">))</span>
<span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">axes</span><span class="p">(</span><span class="n">projection</span> <span class="o">=</span> <span class="s1">&#39;3d&#39;</span><span class="p">)</span>
<span class="n">fig</span><span class="o">.</span><span class="n">add_axes</span><span class="p">(</span><span class="n">ax</span><span class="p">)</span>
<span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">animal_class_unique</span><span class="p">)):</span>
    <span class="n">data</span> <span class="o">=</span> <span class="n">cmdscaling_df</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">cmdscaling_df</span><span class="p">[</span><span class="s1">&#39;class&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">animal_class_unique</span><span class="p">[</span><span class="n">i</span><span class="p">]]</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">data</span><span class="p">[</span><span class="s1">&#39;x&#39;</span><span class="p">],</span> <span class="n">data</span><span class="p">[</span><span class="s1">&#39;y&#39;</span><span class="p">],</span> <span class="n">data</span><span class="p">[</span><span class="s1">&#39;z&#39;</span><span class="p">],</span> <span class="n">label</span><span class="o">=</span><span class="n">animal_class_unique</span><span class="p">[</span><span class="n">i</span><span class="p">],</span> <span class="n">color</span><span class="o">=</span><span class="n">cmap</span><span class="p">[</span><span class="n">i</span><span class="p">])</span>
<span class="n">ax</span><span class="o">.</span><span class="n">view_init</span><span class="p">(</span><span class="mi">30</span><span class="p">,</span> <span class="o">-</span><span class="mi">125</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">loc</span><span class="o">=</span><span class="p">(</span><span class="mf">1.04</span><span class="p">,</span><span class="mf">0.5</span><span class="p">))</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/chaos-game-representation_75_0.png" src="../_images/chaos-game-representation_75_0.png" />
</div>
</div>
<p><em>If you are viewing this in jupyter notebook, uncomment the first line of the previous code block and rerun the code block. This will allow you to rotate the 3D plot above! (Before moving on to the following code blocks run the line <code class="docutils literal notranslate"><span class="pre">%matplotlib</span> <span class="pre">inline</span></code> again so that the following plots will render within the notebook.)</em></p>
<p>If you are curious as to how to use the <code class="docutils literal notranslate"><span class="pre">sklearn.manifold.MDS()</span></code> function, the code for multidimensional scaling for the 4844 vertebrates is given in the following code block. <em>Warning:</em> if you are not performing the following code block in parallel, it will take around half an hour to complete. Additionally, since our <code class="docutils literal notranslate"><span class="pre">DSSIM</span></code> variable is already a dissimilarity matrix, we set the parameter <code class="docutils literal notranslate"><span class="pre">dissimilarity</span></code> in our <code class="docutils literal notranslate"><span class="pre">sklearn.manifold.MDS()</span></code> class to <code class="docutils literal notranslate"><span class="pre">precomputed</span></code>. However, be aware that the dissimilarity matrix <code class="docutils literal notranslate"><span class="pre">DSSIM</span></code> is not Euclidean, and therefore, the results using the <code class="docutils literal notranslate"><span class="pre">sklearn.maifold.MDS()</span></code> class will differ from the results shown previously.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">md_scaling</span> <span class="o">=</span> <span class="n">manifold</span><span class="o">.</span><span class="n">MDS</span><span class="p">(</span><span class="n">n_components</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">dissimilarity</span><span class="o">=</span><span class="s1">&#39;precomputed&#39;</span><span class="p">,</span> <span class="n">random_state</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">n_jobs</span><span class="o">=</span><span class="n">mp</span><span class="o">.</span><span class="n">cpu_count</span><span class="p">())</span>
<span class="n">coord</span> <span class="o">=</span> <span class="n">md_scaling</span><span class="o">.</span><span class="n">fit_transform</span><span class="p">(</span><span class="n">DSSIM</span><span class="p">)</span>
<span class="n">mdscaling_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">data</span> <span class="o">=</span> <span class="n">coord</span><span class="p">,</span> <span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;x&#39;</span><span class="p">,</span> <span class="s1">&#39;y&#39;</span><span class="p">])</span>
<span class="n">mdscaling_df</span><span class="p">[</span><span class="s1">&#39;class&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">animal_class</span>
<span class="n">mdscaling_df</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>x</th>
      <th>y</th>
      <th>class</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>-0.425241</td>
      <td>0.569110</td>
      <td>amphibians</td>
    </tr>
    <tr>
      <th>1</th>
      <td>0.712073</td>
      <td>0.134961</td>
      <td>amphibians</td>
    </tr>
    <tr>
      <th>2</th>
      <td>-0.660362</td>
      <td>-0.025253</td>
      <td>amphibians</td>
    </tr>
    <tr>
      <th>3</th>
      <td>0.703996</td>
      <td>0.048820</td>
      <td>amphibians</td>
    </tr>
    <tr>
      <th>4</th>
      <td>-0.589360</td>
      <td>0.044000</td>
      <td>amphibians</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">fig</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">()</span>
<span class="n">fig</span><span class="o">.</span><span class="n">set_size_inches</span><span class="p">(</span><span class="mi">11</span><span class="p">,</span> <span class="mi">8</span><span class="p">)</span>
<span class="n">ax</span> <span class="o">=</span> <span class="n">sns</span><span class="o">.</span><span class="n">scatterplot</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="s1">&#39;x&#39;</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="s1">&#39;y&#39;</span><span class="p">,</span> <span class="n">hue</span><span class="o">=</span><span class="s1">&#39;class&#39;</span><span class="p">,</span> <span class="n">data</span><span class="o">=</span><span class="n">mdscaling_df</span><span class="p">)</span>
<span class="n">legend</span> <span class="o">=</span> <span class="n">ax</span><span class="o">.</span><span class="n">legend</span><span class="p">()</span>
<span class="n">plt</span><span class="o">.</span><span class="n">axis</span><span class="p">(</span><span class="s1">&#39;equal&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/chaos-game-representation_79_0.png" src="../_images/chaos-game-representation_79_0.png" />
</div>
</div>
<p>Although the results from using the <code class="docutils literal notranslate"><span class="pre">sklearn.manifold.MDS()</span></code> class differs from the result from the <code class="docutils literal notranslate"><span class="pre">cmdscale()</span></code> function, we observe that this plot is still able to group the animals by class quite well.</p>
<div class="tip admonition" id="cgractivity-5">
<p class="admonition-title">CGR Activity 5</p>
<p>Create a molecular distance map of the mitochondrion DNA of molluscs. The Fasta files are provided to you and can be found in the <code class="docutils literal notranslate"><span class="pre">Supplementary</span> <span class="pre">Material/CGR/Activities/Molluscs/Fasta</span> <span class="pre">Files</span></code> folder. <a class="reference internal" href="../Solutions/Reports%20on%20Chaos%20Game%20Activities.html#cgractivityreport-5"><span class="std std-ref">[What happened when we did this]</span></a></p>
</div>
</section>
<section id="protein-visualization">
<h3>Protein Visualization<a class="headerlink" href="#protein-visualization" title="Permalink to this headline">#</a></h3>
<p>CGR has also been applied to visualizing and analyzing both the primary and secondary structures of proteins. The primary structure of a protein is simply an amino acid sequence. To analyze the primary structure of proteins, the authors from <span id="id39">[<a class="reference internal" href="../references.html#id22" title="A. Fiser, G. E. Tusnady, and I. Simon. Chaos game representation of protein structures. Journal of molecular graphics, 12(4):302–304, 1994.">Fiser <em>et al.</em>, 1994</a>]</span> using an <span class="math notranslate nohighlight">\(20\)</span>-sided regular polygon, each representing an amino acid. <a class="reference internal" href="#amino-acid"><span class="std std-numref">Table 2</span></a> shows all 20 amino acids along with their 3-letter code, as well as their 1-letter code. To avoid the attractor from overlapping itself, we will use the dividing rate for a <span class="math notranslate nohighlight">\(20\)</span>-gon shown from <span id="id40">[<a class="reference internal" href="../references.html#id9" title="J. S. Almeida and S. Vinga. Biological sequences as pictures–a generic two dimensional solution for iterated maps. BMC Bioinformatics, 10(1):100, 2009.">Almeida and Vinga, 2009</a>]</span>, <span class="math notranslate nohighlight">\(r = 0.863271\)</span>. Some CGRs for protein visualizations can be found in the following.</p>
<table class="colwidths-auto table" id="amino-acid">
<caption><span class="caption-number">Table 2 </span><span class="caption-text">Table of all 20 amino acids</span><a class="headerlink" href="#amino-acid" title="Permalink to this table">#</a></caption>
<thead>
<tr class="row-odd"><th class="head"><p>Name</p></th>
<th class="head"><p>3 letter code</p></th>
<th class="head"><p>1 letter code</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>alanine</p></td>
<td><p>ala</p></td>
<td><p>A</p></td>
</tr>
<tr class="row-odd"><td><p>arginine</p></td>
<td><p>arg</p></td>
<td><p>R</p></td>
</tr>
<tr class="row-even"><td><p>asparagine</p></td>
<td><p>asn</p></td>
<td><p>N</p></td>
</tr>
<tr class="row-odd"><td><p>aspartic acid</p></td>
<td><p>asp</p></td>
<td><p>D</p></td>
</tr>
<tr class="row-even"><td><p>cysteine</p></td>
<td><p>cys</p></td>
<td><p>C</p></td>
</tr>
<tr class="row-odd"><td><p>glutamine</p></td>
<td><p>gln</p></td>
<td><p>Q</p></td>
</tr>
<tr class="row-even"><td><p>glutamic acid</p></td>
<td><p>glu</p></td>
<td><p>E</p></td>
</tr>
<tr class="row-odd"><td><p>glycine</p></td>
<td><p>gly</p></td>
<td><p>G</p></td>
</tr>
<tr class="row-even"><td><p>histidine</p></td>
<td><p>his</p></td>
<td><p>H</p></td>
</tr>
<tr class="row-odd"><td><p>isoleucine</p></td>
<td><p>ile</p></td>
<td><p>I</p></td>
</tr>
<tr class="row-even"><td><p>leucine</p></td>
<td><p>leu</p></td>
<td><p>L</p></td>
</tr>
<tr class="row-odd"><td><p>lysine</p></td>
<td><p>lys</p></td>
<td><p>K</p></td>
</tr>
<tr class="row-even"><td><p>methionine</p></td>
<td><p>met</p></td>
<td><p>M</p></td>
</tr>
<tr class="row-odd"><td><p>phenylalanine</p></td>
<td><p>phe</p></td>
<td><p>F</p></td>
</tr>
<tr class="row-even"><td><p>proline</p></td>
<td><p>pro</p></td>
<td><p>P</p></td>
</tr>
<tr class="row-odd"><td><p>serine</p></td>
<td><p>ser</p></td>
<td><p>S</p></td>
</tr>
<tr class="row-even"><td><p>threonine</p></td>
<td><p>thre</p></td>
<td><p>T</p></td>
</tr>
<tr class="row-odd"><td><p>tryptophan</p></td>
<td><p>trp</p></td>
<td><p>W</p></td>
</tr>
<tr class="row-even"><td><p>tyrosine</p></td>
<td><p>tyr</p></td>
<td><p>Y</p></td>
</tr>
<tr class="row-odd"><td><p>valine</p></td>
<td><p>val</p></td>
<td><p>V</p></td>
</tr>
</tbody>
</table>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">protein_cgr</span><span class="p">(</span><span class="n">seq</span><span class="p">):</span>
    <span class="n">N</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span>
    <span class="n">AA</span> <span class="o">=</span> <span class="s1">&#39;ARNDCQEGHILKMFPSTWYVBZ&#39;</span>
    <span class="c1"># B = Aspartic acid (D) or Asparagine (N)</span>
    <span class="c1"># Z = Glutamic acid (E) or Glutamine (Q)</span>
    <span class="n">vertices</span> <span class="o">=</span> <span class="n">n_gon</span><span class="p">(</span><span class="mi">20</span><span class="p">,</span> <span class="mi">90</span><span class="p">)</span>
    <span class="n">r</span> <span class="o">=</span> <span class="n">dividingRateAlmeida</span><span class="p">(</span><span class="mi">20</span><span class="p">)</span>
    <span class="n">dataPoints</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="mi">2</span><span class="p">,</span> <span class="n">N</span><span class="o">+</span><span class="mi">1</span><span class="p">))</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">N</span><span class="o">+</span><span class="mi">1</span><span class="p">):</span>
        <span class="n">index</span> <span class="o">=</span> <span class="n">AA</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="n">seq</span><span class="p">[</span><span class="n">i</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
        <span class="k">if</span> <span class="n">index</span> <span class="o">==</span> <span class="mi">20</span><span class="p">:</span>
            <span class="n">r</span> <span class="o">=</span> <span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">r</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
                <span class="n">index</span> <span class="o">=</span> <span class="n">AA</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="s1">&#39;D&#39;</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">index</span> <span class="o">=</span> <span class="n">AA</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="s1">&#39;N&#39;</span><span class="p">)</span>
        <span class="k">elif</span> <span class="n">index</span> <span class="o">==</span> <span class="mi">21</span><span class="p">:</span>
            <span class="n">r</span> <span class="o">=</span> <span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">r</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
                <span class="n">index</span> <span class="o">=</span> <span class="n">AA</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="s1">&#39;E&#39;</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">index</span> <span class="o">=</span> <span class="n">AA</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="s1">&#39;Q&#39;</span><span class="p">)</span>
        <span class="n">dataPoints</span><span class="p">[:,</span> <span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="n">dataPoints</span><span class="p">[:,</span> <span class="n">i</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">+</span> <span class="p">(</span><span class="n">vertices</span><span class="p">[:,</span> <span class="n">index</span><span class="p">]</span> <span class="o">-</span> <span class="n">dataPoints</span><span class="p">[:,</span> <span class="n">i</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span><span class="o">*</span><span class="n">r</span>
    <span class="k">return</span><span class="p">(</span><span class="n">vertices</span><span class="p">,</span> <span class="n">dataPoints</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">fig</span><span class="p">,</span> <span class="p">((</span><span class="n">ax1</span><span class="p">,</span> <span class="n">ax2</span><span class="p">),</span> <span class="p">(</span><span class="n">ax3</span><span class="p">,</span> <span class="n">ax4</span><span class="p">))</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">12</span><span class="p">,</span> <span class="mi">11</span><span class="p">))</span>
<span class="n">fig</span><span class="o">.</span><span class="n">suptitle</span><span class="p">(</span><span class="s1">&#39;Examples of CGR of protein sequences&#39;</span><span class="p">)</span>

<span class="n">human_cytochrome_b_seq</span> <span class="o">=</span> <span class="n">parse_sequence</span><span class="p">(</span><span class="s1">&#39;../Supplementary Material/CGR/FASTA Files/ASY00349.fasta&#39;</span><span class="p">)</span>
<span class="p">(</span><span class="n">coord</span><span class="p">,</span> <span class="n">points</span><span class="p">)</span> <span class="o">=</span> <span class="n">protein_cgr</span><span class="p">(</span><span class="n">human_cytochrome_b_seq</span><span class="p">)</span>
<span class="n">ax1</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">coord</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="p">:],</span> <span class="n">coord</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="p">:],</span> <span class="s1">&#39;k&#39;</span><span class="p">)</span>
<span class="n">ax1</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">points</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="p">:],</span> <span class="n">points</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="p">:],</span> <span class="s1">&#39;k.&#39;</span><span class="p">,</span> <span class="n">markersize</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="n">ax1</span><span class="o">.</span><span class="n">axis</span><span class="p">(</span><span class="s1">&#39;off&#39;</span><span class="p">)</span>
<span class="n">ax1</span><span class="o">.</span><span class="n">set_aspect</span><span class="p">(</span><span class="s1">&#39;equal&#39;</span><span class="p">,</span> <span class="s1">&#39;box&#39;</span><span class="p">)</span>
<span class="n">ax1</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;cytochrome b (mitochondrion) [Homo sapiens]&#39;</span><span class="p">)</span>

<span class="n">mouse_cytochrome_b_seq</span> <span class="o">=</span> <span class="n">parse_sequence</span><span class="p">(</span><span class="s1">&#39;../Supplementary Material/CGR/FASTA Files/NP_904340.fasta&#39;</span><span class="p">)</span>
<span class="p">(</span><span class="n">coord</span><span class="p">,</span> <span class="n">points</span><span class="p">)</span> <span class="o">=</span> <span class="n">protein_cgr</span><span class="p">(</span><span class="n">mouse_cytochrome_b_seq</span><span class="p">)</span>
<span class="n">ax2</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">coord</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="p">:],</span> <span class="n">coord</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="p">:],</span> <span class="s1">&#39;k&#39;</span><span class="p">)</span>
<span class="n">ax2</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">points</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="p">:],</span> <span class="n">points</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="p">:],</span> <span class="s1">&#39;k.&#39;</span><span class="p">,</span> <span class="n">markersize</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="n">ax2</span><span class="o">.</span><span class="n">axis</span><span class="p">(</span><span class="s1">&#39;off&#39;</span><span class="p">)</span>
<span class="n">ax2</span><span class="o">.</span><span class="n">set_aspect</span><span class="p">(</span><span class="s1">&#39;equal&#39;</span><span class="p">,</span> <span class="s1">&#39;box&#39;</span><span class="p">)</span>
<span class="n">ax2</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;cytochrome b (mitochondrion) [Mus musculus]&#39;</span><span class="p">)</span>

<span class="n">major_vault_protein_seq</span> <span class="o">=</span> <span class="n">parse_sequence</span><span class="p">(</span><span class="s1">&#39;../Supplementary Material/CGR/FASTA Files/NP_073206.fasta&#39;</span><span class="p">)</span>
<span class="p">(</span><span class="n">coord</span><span class="p">,</span> <span class="n">points</span><span class="p">)</span> <span class="o">=</span> <span class="n">protein_cgr</span><span class="p">(</span><span class="n">major_vault_protein_seq</span><span class="p">)</span>
<span class="n">ax3</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">coord</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="p">:],</span> <span class="n">coord</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="p">:],</span> <span class="s1">&#39;k&#39;</span><span class="p">)</span>
<span class="n">ax3</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">points</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="p">:],</span> <span class="n">points</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="p">:],</span> <span class="s1">&#39;k.&#39;</span><span class="p">,</span> <span class="n">markersize</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="n">ax3</span><span class="o">.</span><span class="n">axis</span><span class="p">(</span><span class="s1">&#39;off&#39;</span><span class="p">)</span>
<span class="n">ax3</span><span class="o">.</span><span class="n">set_aspect</span><span class="p">(</span><span class="s1">&#39;equal&#39;</span><span class="p">,</span> <span class="s1">&#39;box&#39;</span><span class="p">)</span>
<span class="n">ax3</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;major vault protein [Rattus norvegicus]&#39;</span><span class="p">)</span>

<span class="n">NADH_dehydrogenase_seq</span> <span class="o">=</span> <span class="n">parse_sequence</span><span class="p">(</span><span class="s1">&#39;../Supplementary Material/CGR/FASTA Files/NP_904340.fasta&#39;</span><span class="p">)</span>
<span class="p">(</span><span class="n">coord</span><span class="p">,</span> <span class="n">points</span><span class="p">)</span> <span class="o">=</span> <span class="n">protein_cgr</span><span class="p">(</span><span class="n">NADH_dehydrogenase_seq</span><span class="p">)</span>
<span class="n">ax4</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">coord</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="p">:],</span> <span class="n">coord</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="p">:],</span> <span class="s1">&#39;k&#39;</span><span class="p">)</span>
<span class="n">ax4</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">points</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="p">:],</span> <span class="n">points</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="p">:],</span> <span class="s1">&#39;k.&#39;</span><span class="p">,</span> <span class="n">markersize</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="n">ax4</span><span class="o">.</span><span class="n">axis</span><span class="p">(</span><span class="s1">&#39;off&#39;</span><span class="p">)</span>
<span class="n">ax4</span><span class="o">.</span><span class="n">set_aspect</span><span class="p">(</span><span class="s1">&#39;equal&#39;</span><span class="p">,</span> <span class="s1">&#39;box&#39;</span><span class="p">)</span>
<span class="n">ax4</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;NADH dehydrogenase subunit 5 (mitochondrion) [Mus musculus]&#39;</span><span class="p">)</span>

<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/chaos-game-representation_84_0.png" src="../_images/chaos-game-representation_84_0.png" />
</div>
</div>
<div class="tip admonition" id="cgractivity-6">
<p class="admonition-title">CGR Activity 6</p>
<p>Create a molecular distance map of the cytochrome b (mitochondrion) protein in all land plants. Fasta files are located in the <code class="docutils literal notranslate"><span class="pre">Supplementary</span> <span class="pre">Material/CGR/Activities/Land</span> <span class="pre">Plants/Fasta</span> <span class="pre">Files</span></code> folder. <a class="reference internal" href="../Solutions/Reports%20on%20Chaos%20Game%20Activities.html#cgractivityreport-6"><span class="std std-ref">[What happened when we did this]</span></a></p>
</div>
<p>According to Basu et al. <span id="id41">[<a class="reference internal" href="../references.html#id15" title="S. Basu, A. Pan, C. Dutta, and J. Das. Chaos game representation of proteins. Journal of Molecular Graphics and Modelling, 15(5):279–289, 1997.">Basu <em>et al.</em>, 1997</a>]</span>, there are two serious limitations for using a 20-vertex CGR to identify patterns. The first limitation the authors mentioned about using the 20-vertex is that it is hard to visualize different protein families since one would have to plot the sequences in different polygons rather than plotting a random sampling of proteins of different functions and origins in a single CGR <span id="id42">[<a class="reference internal" href="../references.html#id22" title="A. Fiser, G. E. Tusnady, and I. Simon. Chaos game representation of protein structures. Journal of molecular graphics, 12(4):302–304, 1994.">Fiser <em>et al.</em>, 1994</a>]</span>. The second limitation is that the amino acid residues in different positions are often replaced by similar amino acids <span id="id43">[<a class="reference internal" href="../references.html#id15" title="S. Basu, A. Pan, C. Dutta, and J. Das. Chaos game representation of proteins. Journal of Molecular Graphics and Modelling, 15(5):279–289, 1997.">Basu <em>et al.</em>, 1997</a>]</span>. This means that the 20-vertex CGR cannot be used to differentiate between similar and dissimilar residues and the visualization could be different for proteins within the same family.</p>
<p>To fix the second limitation mentioned in <span id="id44">[<a class="reference internal" href="../references.html#id15" title="S. Basu, A. Pan, C. Dutta, and J. Das. Chaos game representation of proteins. Journal of Molecular Graphics and Modelling, 15(5):279–289, 1997.">Basu <em>et al.</em>, 1997</a>]</span>, Fiser et al. <span id="id45">[<a class="reference internal" href="../references.html#id22" title="A. Fiser, G. E. Tusnady, and I. Simon. Chaos game representation of protein structures. Journal of molecular graphics, 12(4):302–304, 1994.">Fiser <em>et al.</em>, 1994</a>]</span> introduced using CGR to study 3D structures of proteins. A chain of amino acid is what is called a polypeptide. Due to each amino acid having a specific structure, which contributes to the amino acid’s properties, such as hydrophobicity or hydrogen-bonding, depending on the sequence of the amino acid chain, the polypeptide chain will fold into its lowest energy configuration <span id="id46">[<a class="reference internal" href="../references.html#id47" title="I. Simon, L. Glasser, and H. A. Scheraga. Calculation of protein conformation as an assembly of stable overlapping segments: application to bovine pancreatic trypsin inhibitor. Proceedings of the National Academy of Sciences, 88(9):3661–3665, 1991.">Simon <em>et al.</em>, 1991</a>]</span>.
This is known as the secondary structure. The two most common structures that appear in this stage are the <span class="math notranslate nohighlight">\(\alpha\)</span>-helices and the <span class="math notranslate nohighlight">\(\beta\)</span>-sheets. Using the chaos game on the secondary structure can indicate any non-randomness of the structural elements in proteins. One way to achieve this is to divide the 20 amino acids into 4 groups based on different properties and assign each group to a corner. Some properties that have been considered for CGR include hydrophobicity, molecular weight, isoelectric point (pI), <span class="math notranslate nohighlight">\(\alpha\)</span> propensity, and <span class="math notranslate nohighlight">\(\beta\)</span> propensity. For details, see <span id="id47">[<a class="reference internal" href="../references.html#id15" title="S. Basu, A. Pan, C. Dutta, and J. Das. Chaos game representation of proteins. Journal of Molecular Graphics and Modelling, 15(5):279–289, 1997.">Basu <em>et al.</em>, 1997</a>, <a class="reference internal" href="../references.html#id22" title="A. Fiser, G. E. Tusnady, and I. Simon. Chaos game representation of protein structures. Journal of molecular graphics, 12(4):302–304, 1994.">Fiser <em>et al.</em>, 1994</a>]</span>.</p>
</section>
</section>
<section id="cgr-of-mathematical-sequences">
<span id="sec-math"></span><h2>CGR of Mathematical Sequences<a class="headerlink" href="#cgr-of-mathematical-sequences" title="Permalink to this headline">#</a></h2>
<p>In the first-year course we gave in 2015 <span id="id48">[<a class="reference internal" href="../references.html#id17" title="E. Y. S. Chan and R. M. Corless. A random walk through experimental mathematics. In Springer Proceedings in Mathematics &amp; Statistics, pages 203–226. Springer International Publishing, 2020. URL: https://doi.org/10.1007/978-3-030-36568-4_14, doi:10.1007/978-3-030-36568-4_14.">Chan and Corless, 2020</a>]</span>, we applied the chaos game representation to mathematical sequences found on the Online Encyclopedia of Integer Sequences (OEIS) <span id="id49">[<a class="reference internal" href="../references.html#id43" title="OEIS Foundation Inc. The on-line encyclopedia of integer sequences. https://oeis.org/, 2019.">Inc., 2019</a>]</span>. In the following, we show some CGR experiments that came from the first-year course. Some sequences that were used from the Online Encyclopedia of Integer Sequences (OEIS) included digits of <span class="math notranslate nohighlight">\(\pi\)</span> (A000796), Fibonacci Numbers (A000045), prime numbers (A000040), and some from the continued fractions section (which were introduced at the beginning of the course). We will not only look at the visualizations created in the first-year course, but we will also determine whether these experiments are considered to be good visual representations.</p>
<!-- Note that for the following 4-vertex examples, the vertices are labeled from "0" to "3" starting from the bottom left corner of the plot going clockwise (i.e. "0" corresponds to the coordinate point $(-1, -1)$, "1" corresponds to the point $(-1, 1)$, etc.).  -->
<section id="digits-of-pi">
<h3>Digits of Pi<a class="headerlink" href="#digits-of-pi" title="Permalink to this headline">#</a></h3>
<p>We first applied CGR to the digits of <span class="math notranslate nohighlight">\(\pi\)</span> (A000796 from the <a class="reference external" href="https://oeis.org/A000796">OEIS</a>), since we believe that the sequence of the digits are random<a class="footnote-reference brackets" href="#id72" id="id50">6</a>. Since we focused on the 4-vertex CGR in the course, we naïvely took the digits of <span class="math notranslate nohighlight">\(\pi\)</span> modulo 4 and applied it to the 4-vertex CGR. The result of this is shown in the left figure below. Since we assumed that the digits of <span class="math notranslate nohighlight">\(\pi\)</span> are random, this means that the visualization should be similar to the square-shaped CGR of random integer sequences from <a class="reference internal" href="#sec-polygons"><span class="std std-ref">Different Polygons</span></a>. However, we can see that this is not the case; instead of a uniform covering of the space, we can see a pattern of vertical lines occurring. Quantitatively, the DSSIM index between the CGR of the digits of <span class="math notranslate nohighlight">\(\pi\)</span> modulo 4 and the CGR of random digits is <span class="math notranslate nohighlight">\(0.9851\)</span>, which indicates that they are not similar. The reason for this is because when we take <span class="math notranslate nohighlight">\(10\bmod{4}\)</span>, the values are distributed to the corners unevenly: our 0 and 1 vertices would have one extra value each, and thus why our CGR of the digits of <span class="math notranslate nohighlight">\(\pi\)</span> does not match the CGR of random integers even though we know for a fact that the digits of <span class="math notranslate nohighlight">\(\pi\)</span> are indeed random. Due to this reason, the left figure below is not a good visual representation of the digits of <span class="math notranslate nohighlight">\(\pi\)</span>.</p>
<p>However, there are solutions to this problem. Through personal communication, Lila Kari suggested reading 2 digits of <span class="math notranslate nohighlight">\(\pi\)</span> at a time (i.e. the first few 2-digit number of <span class="math notranslate nohighlight">\(\pi\)</span> would be <span class="math notranslate nohighlight">\(14\)</span>, <span class="math notranslate nohighlight">\(15\)</span>, <span class="math notranslate nohighlight">\(92\)</span>, etc.). As the smallest and largest possible number would be 0 and 99, respectively, this means that there are 100 possible combinations, and therefore 100 different numbers. Therefore, since 100 is divisible by 4, this means that we can group the numbers evenly into 4 groups, whether it is by taking the modulo (base 4) of the numbers, or dividing the numbers by the groups 0-24, 25-49, 50-74, and 75-99.</p>
<p>Another solution to this is to visualize the digits of <span class="math notranslate nohighlight">\(\pi\)</span> is shown in the right figure below. We use a 10-vertex CGR (<span class="math notranslate nohighlight">\(r = 0.763932\)</span>) where each vertex represents an integer from 0 to 9. From this figure, we can see that the points produce a fractal pattern of the decagon, which suggests that the sequence is random.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">pi_digits</span><span class="p">(</span><span class="n">n</span><span class="p">):</span>
    <span class="n">S</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">sympy</span><span class="o">.</span><span class="n">N</span><span class="p">(</span><span class="n">sympy</span><span class="o">.</span><span class="n">pi</span><span class="p">,</span> <span class="n">n</span><span class="o">+</span><span class="mi">1</span><span class="p">))</span>
    <span class="k">return</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="nb">map</span><span class="p">(</span><span class="nb">int</span><span class="p">,</span> <span class="nb">list</span><span class="p">(</span><span class="n">S</span><span class="p">[</span><span class="mi">2</span><span class="p">:]))))</span>    
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">numeric_cgr</span><span class="p">(</span><span class="n">seq</span><span class="p">,</span> <span class="n">n_ver</span><span class="p">,</span> <span class="o">*</span><span class="n">start</span><span class="p">):</span>
    <span class="n">N</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">start</span><span class="p">:</span>
        <span class="n">vertices</span> <span class="o">=</span> <span class="n">n_gon</span><span class="p">(</span><span class="n">n_ver</span><span class="p">,</span> <span class="n">start</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">vertices</span> <span class="o">=</span> <span class="n">n_gon</span><span class="p">(</span><span class="n">n_ver</span><span class="p">,</span> <span class="mi">0</span><span class="p">)</span>
    <span class="n">r</span> <span class="o">=</span> <span class="n">dividingRateAlmeida</span><span class="p">(</span><span class="n">n_ver</span><span class="p">)</span>
    <span class="n">dataPoints</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="mi">2</span><span class="p">,</span> <span class="n">N</span><span class="o">+</span><span class="mi">1</span><span class="p">))</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">N</span><span class="o">+</span><span class="mi">1</span><span class="p">):</span>
        <span class="n">index</span> <span class="o">=</span> <span class="n">seq</span><span class="p">[</span><span class="n">i</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">%</span><span class="k">n_ver</span>
        <span class="n">dataPoints</span><span class="p">[:,</span> <span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="n">dataPoints</span><span class="p">[:,</span> <span class="n">i</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">+</span> <span class="p">(</span><span class="n">vertices</span><span class="p">[:,</span> <span class="n">index</span><span class="p">]</span> <span class="o">-</span> <span class="n">dataPoints</span><span class="p">[:,</span> <span class="n">i</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span><span class="o">*</span><span class="n">r</span>
    <span class="k">return</span><span class="p">(</span><span class="n">vertices</span><span class="p">,</span> <span class="n">dataPoints</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">N</span> <span class="o">=</span> <span class="mi">50000</span>
<span class="n">pi_seq</span> <span class="o">=</span> <span class="n">pi_digits</span><span class="p">(</span><span class="n">N</span><span class="p">)</span>

<span class="n">fig</span><span class="p">,</span> <span class="p">(</span><span class="n">ax1</span><span class="p">,</span> <span class="n">ax2</span><span class="p">)</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">16</span><span class="p">,</span> <span class="mi">6</span><span class="p">))</span>
<span class="n">fig</span><span class="o">.</span><span class="n">suptitle</span><span class="p">(</span><span class="s1">&#39;CGRs of the first </span><span class="si">{}</span><span class="s1"> digits of pi&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">N</span><span class="p">))</span>
<span class="p">(</span><span class="n">coord</span><span class="p">,</span> <span class="n">points</span><span class="p">)</span> <span class="o">=</span> <span class="n">numeric_cgr</span><span class="p">(</span><span class="n">pi_seq</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="o">-</span><span class="mi">135</span><span class="p">)</span>
<span class="n">ax1</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">coord</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="p">:],</span> <span class="n">coord</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="p">:],</span> <span class="s1">&#39;k&#39;</span><span class="p">)</span>
<span class="n">ax1</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">points</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="p">:],</span> <span class="n">points</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="p">:],</span> <span class="s1">&#39;k.&#39;</span><span class="p">,</span> <span class="n">markersize</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="n">ax1</span><span class="o">.</span><span class="n">axis</span><span class="p">(</span><span class="s1">&#39;off&#39;</span><span class="p">)</span>
<span class="n">ax1</span><span class="o">.</span><span class="n">set_aspect</span><span class="p">(</span><span class="s1">&#39;equal&#39;</span><span class="p">,</span> <span class="s1">&#39;box&#39;</span><span class="p">)</span>
<span class="n">ax1</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;Digits of pi modulo 4&#39;</span><span class="p">)</span>

<span class="p">(</span><span class="n">coord</span><span class="p">,</span> <span class="n">points</span><span class="p">)</span> <span class="o">=</span> <span class="n">numeric_cgr</span><span class="p">(</span><span class="n">pi_seq</span><span class="p">,</span> <span class="mi">10</span><span class="p">)</span>
<span class="n">ax2</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">coord</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="p">:],</span> <span class="n">coord</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="p">:],</span> <span class="s1">&#39;k&#39;</span><span class="p">)</span>
<span class="n">ax2</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">points</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="p">:],</span> <span class="n">points</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="p">:],</span> <span class="s1">&#39;k.&#39;</span><span class="p">,</span> <span class="n">markersize</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="n">ax2</span><span class="o">.</span><span class="n">axis</span><span class="p">(</span><span class="s1">&#39;off&#39;</span><span class="p">)</span>
<span class="n">ax2</span><span class="o">.</span><span class="n">set_aspect</span><span class="p">(</span><span class="s1">&#39;equal&#39;</span><span class="p">,</span> <span class="s1">&#39;box&#39;</span><span class="p">)</span>
<span class="n">ax2</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;Digits of pi modulo 10&#39;</span><span class="p">)</span>

<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/chaos-game-representation_90_0.png" src="../_images/chaos-game-representation_90_0.png" />
</div>
</div>
<div class="tip admonition" id="cgractivity-7">
<p class="admonition-title">CGR Activity 7</p>
<p>Create a square CGR for the digits of <span class="math notranslate nohighlight">\(pi\)</span> by reading 2 digits at a time (as explained above). Assign the two-digit numbers a corner by the following two methods:</p>
<p>a) taking the modulo (base 4) and assigning the corners 0, 1, 2, 3 <a class="reference internal" href="../Solutions/Reports%20on%20Chaos%20Game%20Activities.html#cgractivityreport-7a"><span class="std std-ref">[What happened when we did this]</span></a></p>
<p>b) assigning the corners to cover the values 0-24, 25-49, 50-74, and 75-99, respectively. <a class="reference internal" href="../Solutions/Reports%20on%20Chaos%20Game%20Activities.html#cgractivityreport-7b"><span class="std std-ref">[What happened when we did this]</span></a></p>
<p>c) Do the CGRs look different from one another? What conclusions can you draw from the two CGRs? <a class="reference internal" href="../Solutions/Reports%20on%20Chaos%20Game%20Activities.html#cgractivityreport-7c"><span class="std std-ref">[What happened when we did this]</span></a></p>
</div>
<div class="tip admonition" id="cgractivity-8">
<p class="admonition-title">CGR Activity 8</p>
<p>Repeat the above activity with the digits of <span class="math notranslate nohighlight">\(e = 2.71828...\)</span> <a class="reference internal" href="../Solutions/Reports%20on%20Chaos%20Game%20Activities.html#cgractivityreport-8"><span class="std std-ref">[What happened when we did this]</span></a></p>
</div>
</section>
<section id="fibonacci-sequence">
<h3>Fibonacci Sequence<a class="headerlink" href="#fibonacci-sequence" title="Permalink to this headline">#</a></h3>
<p>We also created a visualization of the Fibonacci Sequence (A000045 from the OEIS). One example of a visualization of this sequence is a 10-vertex CGR of the Fibonacci sequence modulo 10; that is, a sequence of the final digits of <span class="math notranslate nohighlight">\(F_n\)</span>. It appears by the density of the even blocks in <a class="reference internal" href="#cgr-fibonacci-wrong"><span class="std std-numref">Fig. 12</span></a> that most (by far and away most) of the numbers of the Fibonacci sequence are even; this conclusion seems very strange, and deserves a closer look.</p>
<p>In fact the sequence is ultimately periodic (as of course it must be, because the set of decimal digits is finite), and indeed it has period <span class="math notranslate nohighlight">\(60\)</span>: The first <span class="math notranslate nohighlight">\(65\)</span> entries are <span class="math notranslate nohighlight">\(1, 1, 2, 3, 5, 8, 3, 1, \ldots, 9, 1, 0, 1, 1, 2, 3, 5\)</span> and it repeats from there.  Just by counting, <span class="math notranslate nohighlight">\(40\)</span> out of these <span class="math notranslate nohighlight">\(60\)</span> numbers are <em>odd</em>, and only <span class="math notranslate nohighlight">\(20\)</span> are even; so the conclusion that “almost all Fibonacci numbers are even” that we drew from <a class="reference internal" href="#cgr-fibonacci-wrong"><span class="std std-numref">Fig. 12</span></a> is completely wrong!</p>
<figure class="align-default" id="cgr-fibonacci-wrong">
<a class="reference internal image-reference" href="../_images/FibonacciCGRmod10crop.png"><img alt="../_images/FibonacciCGRmod10crop.png" src="../_images/FibonacciCGRmod10crop.png" style="height: 300px;" /></a>
<figcaption>
<p><span class="caption-number">Fig. 12 </span><span class="caption-text">CGR of the Fibonacci sequence calculated in floating-point arithmetic before taking the remainder mod 10 in Matlab</span><a class="headerlink" href="#cgr-fibonacci-wrong" title="Permalink to this image">#</a></p>
</figcaption>
</figure>
<p>The correct behaviour is shown below; this shows the periodicity by its sparseness.  It is important to do the modular reduction as you go, if you are working in floating-point arithmetic; once <span class="math notranslate nohighlight">\(F_n\)</span> gets large enough it won’t be possible to compute the trailing digit correctly and all one will see is rounding error, which we see in <a class="reference internal" href="#cgr-fibonacci-wrong"><span class="std std-numref">Fig. 12</span></a> looks pretty uniform on each <em>even</em> block. In contrast, while a pair of floating-point integers (called a “flints” in some places) remains small, arithmetic with it remains perfectly accurate. What we are seeing in the figure is an effect of rounding errors in <em>large</em> floating-point integers; more, a closer look at <span class="math notranslate nohighlight">\(f_{79}\)</span> shows the <em>round-to-even</em> rule for IEEE floating-point arithmetic<a class="footnote-reference brackets" href="#id74" id="id51">7</a>. Since <span class="math notranslate nohighlight">\(f_{78}\)</span> really is even, and <span class="math notranslate nohighlight">\(f_{79}\)</span> is rounded to an even number by the round-to-even rule, all computed floating-point Fibonacci numbers thereafter are even. Of course, once they got large enough that the units place falls completely off the significand, they would all be reported as “even” anyway.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">fibonacci</span><span class="p">(</span><span class="n">n</span><span class="p">,</span> <span class="n">mod</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
    <span class="k">if</span> <span class="n">n</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
        <span class="k">return</span><span class="p">([</span><span class="mi">0</span><span class="p">])</span>
    <span class="k">elif</span> <span class="n">n</span> <span class="o">==</span> <span class="mi">2</span><span class="p">:</span>
        <span class="k">return</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">])</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">seq</span> <span class="o">=</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">]</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="n">n</span><span class="p">):</span>
            <span class="n">n_element</span> <span class="o">=</span> <span class="n">seq</span><span class="p">[</span><span class="n">i</span><span class="o">-</span><span class="mi">2</span><span class="p">]</span> <span class="o">+</span> <span class="n">seq</span><span class="p">[</span><span class="n">i</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
            <span class="k">if</span> <span class="n">mod</span><span class="p">:</span>
                <span class="n">n_element</span> <span class="o">=</span> <span class="n">n_element</span><span class="o">%</span><span class="k">10</span>
            <span class="n">seq</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">n_element</span><span class="p">)</span>
        <span class="k">return</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">N</span> <span class="o">=</span> <span class="mi">10000</span>

<span class="n">fig</span><span class="p">,</span> <span class="p">(</span><span class="n">ax1</span><span class="p">,</span> <span class="n">ax2</span><span class="p">)</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">16</span><span class="p">,</span> <span class="mi">6</span><span class="p">))</span>
<span class="n">fig</span><span class="o">.</span><span class="n">suptitle</span><span class="p">(</span><span class="s1">&#39;CGRs of the first </span><span class="si">{}</span><span class="s1"> Fibonacci Sequence modulo 10&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">N</span><span class="p">))</span>
<span class="n">fib_seq</span> <span class="o">=</span> <span class="n">fibonacci</span><span class="p">(</span><span class="n">N</span><span class="p">)</span>
<span class="p">(</span><span class="n">coord</span><span class="p">,</span> <span class="n">points</span><span class="p">)</span> <span class="o">=</span> <span class="n">numeric_cgr</span><span class="p">(</span><span class="n">fib_seq</span><span class="p">,</span> <span class="mi">10</span><span class="p">)</span>
<span class="n">ax1</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">coord</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="p">:],</span> <span class="n">coord</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="p">:],</span> <span class="s1">&#39;k&#39;</span><span class="p">)</span>
<span class="n">ax1</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">points</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="p">:],</span> <span class="n">points</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="p">:],</span> <span class="s1">&#39;k.&#39;</span><span class="p">,</span> <span class="n">markersize</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="n">ax1</span><span class="o">.</span><span class="n">axis</span><span class="p">(</span><span class="s1">&#39;off&#39;</span><span class="p">)</span>
<span class="n">ax1</span><span class="o">.</span><span class="n">set_aspect</span><span class="p">(</span><span class="s1">&#39;equal&#39;</span><span class="p">,</span> <span class="s1">&#39;box&#39;</span><span class="p">)</span>
<span class="n">ax1</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;Taking modulo 10 after creating the Fibonacci sequence&#39;</span><span class="p">)</span>

<span class="n">fib_seq_mod10</span> <span class="o">=</span> <span class="n">fibonacci</span><span class="p">(</span><span class="n">N</span><span class="p">,</span> <span class="n">mod</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="p">(</span><span class="n">coord</span><span class="p">,</span> <span class="n">points</span><span class="p">)</span> <span class="o">=</span> <span class="n">numeric_cgr</span><span class="p">(</span><span class="n">fib_seq</span><span class="p">,</span> <span class="mi">10</span><span class="p">)</span>
<span class="n">ax2</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">coord</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="p">:],</span> <span class="n">coord</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="p">:],</span> <span class="s1">&#39;k&#39;</span><span class="p">)</span>
<span class="n">ax2</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">points</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="p">:],</span> <span class="n">points</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="p">:],</span> <span class="s1">&#39;k.&#39;</span><span class="p">,</span> <span class="n">markersize</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="n">ax2</span><span class="o">.</span><span class="n">axis</span><span class="p">(</span><span class="s1">&#39;off&#39;</span><span class="p">)</span>
<span class="n">ax2</span><span class="o">.</span><span class="n">set_aspect</span><span class="p">(</span><span class="s1">&#39;equal&#39;</span><span class="p">,</span> <span class="s1">&#39;box&#39;</span><span class="p">)</span>
<span class="n">ax2</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;Taking modulo 10 while making the Fibonacci sequence&#39;</span><span class="p">)</span>

<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/chaos-game-representation_97_0.png" src="../_images/chaos-game-representation_97_0.png" />
</div>
</div>
<p>Other visualizations for the Fibonacci sequence can be done, such as taking the remainder using a different divisor and using a CGR with different number of vertices.</p>
<p>However, since the Fibonacci sequence grows very quickly, like <span class="math notranslate nohighlight">\(\phi^n\)</span> where <span class="math notranslate nohighlight">\(\phi = (1+\sqrt{5})/2 \doteq 1.618\)</span> is the golden ratio, the students also learned about another limitation of floating point since they were unable to take more than the first 3000 Fibonacci numbers before the computation overflow, registering the number as infinity. Using big integers instead as is possible in Python, overflow wouldn’t happen (as seen above), but the arithmetic would get slower.</p>
</section>
<section id="prime-numbers">
<h3>Prime Numbers<a class="headerlink" href="#prime-numbers" title="Permalink to this headline">#</a></h3>
<p>Following the visualization of the digits of <span class="math notranslate nohighlight">\(\pi\)</span> and the Fibonacci sequence using the chaos game, we looked at another popular sequence: the sequence of prime numbers (<a class="reference external" href="https://oeis.org/A000040">A000040 from the OEIS</a>). We looked at four different ways of creating the CGR for prime numbers.</p>
<p>We first visualized the prime numbers in a similar way as what we did for the digits of <span class="math notranslate nohighlight">\(\pi\)</span> and the Fibonacci numbers by taking the sequence of prime numbers modulo 10. This is shown in the top left figure below. We notice that almost all of the prime numbers ends with a 1, 3, 7, or 9. Because of this observation, we decided to create a square CGR labelling the vertices with these values. The result is shown in the top right figure below.</p>
<p>The bottom left figure is the third visualization of prime numbers: a CGR of the <span class="math notranslate nohighlight">\(103+k\)</span>th prime numbers <span class="math notranslate nohighlight">\(\bmod\ 4\)</span>, a diagonal line which spans from the “1” corner to the “3” corner. As we all know, prime numbers larger than 2 will never be even, so the corner points “0” and “2” which are representative “even” values would never occur, thus creating a straight diagonal line. If we had reassigned the values the coordinate points represent, the plot would have turned out differently; instead of a diagonal line, it could possibly be a horizontal (spanning from coordinate points <span class="math notranslate nohighlight">\((-1, -1)\)</span> to <span class="math notranslate nohighlight">\((1, -1)\)</span> or from <span class="math notranslate nohighlight">\((-1, 1)\)</span> to <span class="math notranslate nohighlight">\((1, 1)\)</span>) or vertical line (spanning from <span class="math notranslate nohighlight">\((-1, 1)\)</span> to <span class="math notranslate nohighlight">\((-1, -1)\)</span> or <span class="math notranslate nohighlight">\((1, 1)\)</span> to <span class="math notranslate nohighlight">\((1, -1)\)</span>), or a diagonal line the spanning the other two vertices. One needs to be mindful of how the coordinate points are assigned.</p>
<p>Lastly, we can visualize the prime numbers in a different way, as seen in the bottom right figure. Here, we took the sequence of prime numbers less than one million and took <span class="math notranslate nohighlight">\(\bmod\ 8\)</span> of the numbers. We thought that this would be interesting because taking any prime number larger than <span class="math notranslate nohighlight">\(2\)</span> modulo <span class="math notranslate nohighlight">\(8\)</span> will only result in the following numbers: 1, 3, 5, and 7. Because of this, we relabelled the vertices to “1”, “3”, “5”, and “7” starting from the bottom left corner, going clockwise (i.e. the coordinate point <span class="math notranslate nohighlight">\((-1, -1)\)</span> is labelled “1”, <span class="math notranslate nohighlight">\((-1, 1)\)</span> is labelled “3”, etc.), and plotted the points according to the chaos game (<span class="math notranslate nohighlight">\(r = 0.5\)</span>). Surprisingly, this gives us a pattern, apparently meaning that there is a pattern in the sequence of prime numbers <span class="math notranslate nohighlight">\(\bmod\ 8\)</span>. However, compare this to a random-populated square CGR, which, though random, showed patterns. This “evidence of pattern” is not conclusive! This is possibly simply a Procrustean effect, and we’d have to know more about the distribution of primes to say more.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">primes</span><span class="p">(</span><span class="n">N</span><span class="p">):</span>
    <span class="c1"># returns list of prime numbers less than or equal to N</span>
    <span class="n">seq</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="n">N</span><span class="o">+</span><span class="mi">1</span><span class="p">):</span>
        <span class="n">divisor</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="n">i</span><span class="p">):</span>
            <span class="k">if</span> <span class="n">i</span><span class="o">%</span><span class="k">j</span> == 0:
                <span class="n">divisor</span> <span class="o">+=</span> <span class="mi">1</span>
        <span class="k">if</span> <span class="n">divisor</span> <span class="o">==</span> <span class="mi">0</span> <span class="ow">or</span> <span class="n">i</span> <span class="o">==</span> <span class="mi">2</span><span class="p">:</span>
            <span class="n">seq</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>
    <span class="k">return</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span>

<span class="k">def</span> <span class="nf">numeric_cgr_vertex</span><span class="p">(</span><span class="n">seq</span><span class="p">,</span> <span class="n">vertex</span><span class="p">,</span> <span class="o">*</span><span class="n">start</span><span class="p">):</span>
    <span class="n">N</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span>
    <span class="n">n_ver</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">vertex</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">start</span><span class="p">:</span>
        <span class="n">vertices</span> <span class="o">=</span> <span class="n">n_gon</span><span class="p">(</span><span class="n">n_ver</span><span class="p">,</span> <span class="n">start</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">vertices</span> <span class="o">=</span> <span class="n">n_gon</span><span class="p">(</span><span class="n">n_ver</span><span class="p">,</span> <span class="mi">0</span><span class="p">)</span>
    <span class="n">r</span> <span class="o">=</span> <span class="n">dividingRateAlmeida</span><span class="p">(</span><span class="n">n_ver</span><span class="p">)</span>
    <span class="n">dataPoints</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="mi">2</span><span class="p">,</span> <span class="n">N</span><span class="o">+</span><span class="mi">1</span><span class="p">))</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">N</span><span class="o">+</span><span class="mi">1</span><span class="p">):</span>
        <span class="n">index</span> <span class="o">=</span> <span class="n">vertex</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="n">seq</span><span class="p">[</span><span class="n">i</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
        <span class="n">dataPoints</span><span class="p">[:,</span> <span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="n">dataPoints</span><span class="p">[:,</span> <span class="n">i</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">+</span> <span class="p">(</span><span class="n">vertices</span><span class="p">[:,</span> <span class="n">index</span><span class="p">]</span> <span class="o">-</span> <span class="n">dataPoints</span><span class="p">[:,</span> <span class="n">i</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span><span class="o">*</span><span class="n">r</span>
    <span class="k">return</span><span class="p">(</span><span class="n">vertices</span><span class="p">,</span> <span class="n">dataPoints</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">N</span> <span class="o">=</span> <span class="mi">50000</span>

<span class="n">fig</span><span class="p">,</span> <span class="p">((</span><span class="n">ax1</span><span class="p">,</span> <span class="n">ax2</span><span class="p">),</span> <span class="p">(</span><span class="n">ax3</span><span class="p">,</span> <span class="n">ax4</span><span class="p">))</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">12</span><span class="p">,</span> <span class="mi">11</span><span class="p">))</span>
<span class="n">fig</span><span class="o">.</span><span class="n">suptitle</span><span class="p">(</span><span class="s1">&#39;Examples of CGR of prime numbers&#39;</span><span class="p">)</span>

<span class="n">prime_num</span> <span class="o">=</span> <span class="n">primes</span><span class="p">(</span><span class="n">N</span><span class="p">)</span>
<span class="p">(</span><span class="n">coord</span><span class="p">,</span> <span class="n">points</span><span class="p">)</span> <span class="o">=</span> <span class="n">numeric_cgr</span><span class="p">(</span><span class="n">prime_num</span><span class="p">,</span> <span class="mi">10</span><span class="p">)</span>
<span class="n">ax1</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">coord</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="p">:],</span> <span class="n">coord</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="p">:],</span> <span class="s1">&#39;k&#39;</span><span class="p">)</span>
<span class="n">ax1</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">points</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="p">:],</span> <span class="n">points</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="p">:],</span> <span class="s1">&#39;k.&#39;</span><span class="p">,</span> <span class="n">markersize</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="n">ax1</span><span class="o">.</span><span class="n">axis</span><span class="p">(</span><span class="s1">&#39;off&#39;</span><span class="p">)</span>
<span class="n">ax1</span><span class="o">.</span><span class="n">set_aspect</span><span class="p">(</span><span class="s1">&#39;equal&#39;</span><span class="p">,</span> <span class="s1">&#39;box&#39;</span><span class="p">)</span>
<span class="n">ax1</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;All prime numbers less than </span><span class="si">{}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">N</span><span class="p">))</span>

<span class="n">prime_num_mod10</span> <span class="o">=</span> <span class="p">[</span><span class="n">p</span><span class="o">%</span><span class="k">10</span> for p in prime_num]
<span class="p">(</span><span class="n">coord</span><span class="p">,</span> <span class="n">points</span><span class="p">)</span> <span class="o">=</span> <span class="n">numeric_cgr_vertex</span><span class="p">(</span><span class="n">prime_num_mod10</span><span class="p">[</span><span class="mi">3</span><span class="p">:],</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">7</span><span class="p">,</span> <span class="mi">9</span><span class="p">],</span> <span class="o">-</span><span class="mi">135</span><span class="p">)</span>
<span class="n">ax2</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">coord</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="p">:],</span> <span class="n">coord</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="p">:],</span> <span class="s1">&#39;k&#39;</span><span class="p">)</span>
<span class="n">ax2</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">points</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="p">:],</span> <span class="n">points</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="p">:],</span> <span class="s1">&#39;k.&#39;</span><span class="p">,</span> <span class="n">markersize</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="n">ax2</span><span class="o">.</span><span class="n">axis</span><span class="p">(</span><span class="s1">&#39;off&#39;</span><span class="p">)</span>
<span class="n">ax2</span><span class="o">.</span><span class="n">set_aspect</span><span class="p">(</span><span class="s1">&#39;equal&#39;</span><span class="p">,</span> <span class="s1">&#39;box&#39;</span><span class="p">)</span>
<span class="n">ax2</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;All prime numbers between 7 and </span><span class="si">{}</span><span class="s1"> taken mod 10, vertices = [1, 3, 7, 9]&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">N</span><span class="p">))</span>

<span class="n">prime_num_mod4</span> <span class="o">=</span> <span class="p">[</span><span class="n">p</span><span class="o">%</span><span class="k">4</span> for p in prime_num]
<span class="p">(</span><span class="n">coord</span><span class="p">,</span> <span class="n">points</span><span class="p">)</span> <span class="o">=</span> <span class="n">numeric_cgr</span><span class="p">(</span><span class="n">prime_num_mod4</span><span class="p">[</span><span class="mi">102</span><span class="p">:],</span> <span class="mi">4</span><span class="p">,</span> <span class="o">-</span><span class="mi">135</span><span class="p">)</span>
<span class="n">ax3</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">coord</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="p">:],</span> <span class="n">coord</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="p">:],</span> <span class="s1">&#39;k&#39;</span><span class="p">)</span>
<span class="n">ax3</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">points</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="p">:],</span> <span class="n">points</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="p">:],</span> <span class="s1">&#39;k.&#39;</span><span class="p">,</span> <span class="n">markersize</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="n">ax3</span><span class="o">.</span><span class="n">axis</span><span class="p">(</span><span class="s1">&#39;off&#39;</span><span class="p">)</span>
<span class="n">ax3</span><span class="o">.</span><span class="n">set_aspect</span><span class="p">(</span><span class="s1">&#39;equal&#39;</span><span class="p">,</span> <span class="s1">&#39;box&#39;</span><span class="p">)</span>
<span class="n">ax3</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;103+kth prime number mod 4&#39;</span><span class="p">)</span>

<span class="n">prime_num_mod8</span> <span class="o">=</span> <span class="p">[</span><span class="n">p</span><span class="o">%</span><span class="k">8</span> for p in prime_num]
<span class="p">(</span><span class="n">coord</span><span class="p">,</span> <span class="n">points</span><span class="p">)</span> <span class="o">=</span> <span class="n">numeric_cgr_vertex</span><span class="p">(</span><span class="n">prime_num_mod8</span><span class="p">[</span><span class="mi">3</span><span class="p">:],</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">5</span><span class="p">,</span> <span class="mi">7</span><span class="p">],</span> <span class="o">-</span><span class="mi">135</span><span class="p">)</span>
<span class="n">ax4</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">coord</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="p">:],</span> <span class="n">coord</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="p">:],</span> <span class="s1">&#39;k&#39;</span><span class="p">)</span>
<span class="n">ax4</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">points</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="p">:],</span> <span class="n">points</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="p">:],</span> <span class="s1">&#39;k.&#39;</span><span class="p">,</span> <span class="n">markersize</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="n">ax4</span><span class="o">.</span><span class="n">axis</span><span class="p">(</span><span class="s1">&#39;off&#39;</span><span class="p">)</span>
<span class="n">ax4</span><span class="o">.</span><span class="n">set_aspect</span><span class="p">(</span><span class="s1">&#39;equal&#39;</span><span class="p">,</span> <span class="s1">&#39;box&#39;</span><span class="p">)</span>
<span class="n">ax4</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;All prime numbers between 7 and </span><span class="si">{}</span><span class="s1"> taken mod 8, vertices = [1, 3, 5, 7]&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">N</span><span class="p">))</span>

<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/chaos-game-representation_101_0.png" src="../_images/chaos-game-representation_101_0.png" />
</div>
</div>
</section>
<section id="partial-quotients-of-continued-fractions">
<h3>Partial Quotients of Continued Fractions<a class="headerlink" href="#partial-quotients-of-continued-fractions" title="Permalink to this headline">#</a></h3>
<p>We then explored visualizing the sequences taken from the partial quotients of continued fractions, which were taught at the beginning of the course. Continued fractions are fractions written in the following general form <span id="id52">[<a class="reference internal" href="../references.html#id3" title="R. M. Corless. Continued fractions and chaos. The American Mathematical Monthly, 99(3):203–215, March 1992. doi:10.1080/00029890.1992.11995835.">Corless, 1992</a>, <a class="reference internal" href="../references.html#id39" title="C. D. Olds. Continued fractions. Volume 18. Random House New York, 1963.">Olds, 1963</a>]</span>:</p>
<div class="amsmath math notranslate nohighlight" id="equation-e6b3e39c-6292-4cdf-92b2-957d5234247f">
<span class="eqno">(73)<a class="headerlink" href="#equation-e6b3e39c-6292-4cdf-92b2-957d5234247f" title="Permalink to this equation">#</a></span>\[\begin{equation}
	a_0 + \cfrac{1}{a_{1} + \cfrac{1}{a_{2} + \cfrac{1}{a_{3} + \cfrac{1}{a_{4} + \cfrac{1}{\ddots}}}}} ,
\end{equation}\]</div>
<p>which produces a sequence of numbers, <span class="math notranslate nohighlight">\(a_{0} + \left[ a_{1}, a_{2}, a_{3}, a_{4}, \cdots\right]\)</span>, called the partial quotients of the continued fraction. This can be demonstrated with an example: we can rewrite <span class="math notranslate nohighlight">\(\frac{9}{7}\)</span> in the form</p>
<div class="amsmath math notranslate nohighlight" id="equation-7ddd0dc1-5149-4ba5-8777-c7ce07872ef6">
<span class="eqno">(74)<a class="headerlink" href="#equation-7ddd0dc1-5149-4ba5-8777-c7ce07872ef6" title="Permalink to this equation">#</a></span>\[\begin{equation}
	\dfrac{9}{7} = 1 + \dfrac{2}{7} = 1 + \dfrac{1}{\frac{7}{2}} = 1 + \cfrac{1}{3 + \frac{1}{2}} = 1 + \dfrac{1}{3 + \cfrac{1}{1 + \frac{1}{1}}} .
	\label{eqn:cfrac}
\end{equation}\]</div>
<p>Here, the partial quotients of <span class="math notranslate nohighlight">\(\frac{9}{7}\)</span> are the elements of <span class="math notranslate nohighlight">\(1 + \left[3, 1, 1\right]\)</span>. The students then looked at the continued fractions of other numbers, such as <span class="math notranslate nohighlight">\(\sqrt{2}\)</span>, <span class="math notranslate nohighlight">\(e\)</span> and <span class="math notranslate nohighlight">\(\pi\)</span>, where the sequences of their partial quotients were used in chaos game representations.</p>
<p>One recurring theme of the experimental mathematics course was <span class="math notranslate nohighlight">\(\sqrt{2}\)</span>, so with this in mind, some students jumped on the opportunity to plot the chaos game representation of the quotients of the continued fraction of <span class="math notranslate nohighlight">\(\sqrt{2}\)</span>. As we had seen in an earlier unit, the sequence goes like</p>
<div class="amsmath math notranslate nohighlight">
\[\begin{equation*}
	1 + [2, 2, 2, 2, \ldots, 2] .
\end{equation*}\]</div>
<p>Because all elements of the sequence (apart from the first one) are 2’s, it is not surprising to see a (faint) diagonal line with most of the points in the upper right corner, shown in the left figure below. The students also tried the partial quotients of <span class="math notranslate nohighlight">\(e\)</span>, which is equal to</p>
<div class="amsmath math notranslate nohighlight">
\[\begin{equation*}
	2 + \left[1, 2, 1, 1, 4, 1, 1, 6, \cdots \right]
\end{equation*}\]</div>
<p><span id="id53">[<a class="reference internal" href="../references.html#id40" title="C. D. Olds. The simple continued fraction expansion of e. American Mathematical Monthly, pages 968–974, 1970.">Olds, 1970</a>]</span>, shown in the middle figure. What is seen here makes sense as the sequence mostly contains 1’s and these alternates with even values, so in this case, either 0 or 2. Therefore, it is clear that there are no points in the lower right corner since that represents the value 3. Unfortunately, these two plots do not look all that impressive; in fact, it is pretty underwhelming. Disappointed with this result, the students decided to experiment with other sequences in which all four coordinates occur.</p>
<p>The students then thought, “Why not take the partial quotients of the continued fraction of <span class="math notranslate nohighlight">\(\pi\)</span>? The results <em>must</em> be random.” As shown in the right figure, unexpectedly, there is indeed a pattern, which shows that the sequence of partial quotients of the continued fraction of <span class="math notranslate nohighlight">\(\pi\)</span> is not as random as we thought at the beginning of the course. This showed the students that the distribution is not uniform.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">confrac</span><span class="p">(</span><span class="n">expr</span><span class="p">,</span> <span class="n">n</span><span class="p">):</span>
    <span class="n">result</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">continued_fraction_iterator</span><span class="p">(</span><span class="n">expr</span><span class="p">)):</span>
        <span class="k">if</span> <span class="n">i</span> <span class="o">&gt;</span> <span class="p">(</span><span class="n">n</span><span class="o">-</span><span class="mi">1</span><span class="p">):</span>
            <span class="k">break</span>
        <span class="n">result</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">v</span><span class="p">)</span>
    <span class="k">return</span><span class="p">(</span><span class="n">result</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">fig</span><span class="p">,</span> <span class="p">((</span><span class="n">ax1</span><span class="p">,</span> <span class="n">ax2</span><span class="p">,</span> <span class="n">ax3</span><span class="p">))</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">12</span><span class="p">,</span> <span class="mi">5</span><span class="p">))</span>
<span class="n">fig</span><span class="o">.</span><span class="n">suptitle</span><span class="p">(</span><span class="s1">&#39;Examples of CGR for partial quotients of continued fractions&#39;</span><span class="p">)</span>

<span class="n">pq_sqrt2</span> <span class="o">=</span> <span class="n">confrac</span><span class="p">(</span><span class="n">sympy</span><span class="o">.</span><span class="n">sqrt</span><span class="p">(</span><span class="mi">2</span><span class="p">),</span> <span class="mi">200</span><span class="p">)</span>
<span class="p">(</span><span class="n">coord</span><span class="p">,</span> <span class="n">points</span><span class="p">)</span> <span class="o">=</span> <span class="n">numeric_cgr</span><span class="p">(</span><span class="n">pq_sqrt2</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="o">-</span><span class="mi">135</span><span class="p">)</span>
<span class="n">ax1</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">coord</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="p">:],</span> <span class="n">coord</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="p">:],</span> <span class="s1">&#39;k&#39;</span><span class="p">)</span>
<span class="n">ax1</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">points</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="p">:],</span> <span class="n">points</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="p">:],</span> <span class="s1">&#39;k.&#39;</span><span class="p">,</span> <span class="n">markersize</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="n">ax1</span><span class="o">.</span><span class="n">axis</span><span class="p">(</span><span class="s1">&#39;off&#39;</span><span class="p">)</span>
<span class="n">ax1</span><span class="o">.</span><span class="n">set_aspect</span><span class="p">(</span><span class="s1">&#39;equal&#39;</span><span class="p">,</span> <span class="s1">&#39;box&#39;</span><span class="p">)</span>
<span class="n">ax1</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;sqrt(2) mod 4&#39;</span><span class="p">)</span>

<span class="n">pq_exp</span> <span class="o">=</span> <span class="n">confrac</span><span class="p">(</span><span class="n">sympy</span><span class="o">.</span><span class="n">exp</span><span class="p">(</span><span class="mi">1</span><span class="p">),</span> <span class="mi">200</span><span class="p">)</span>
<span class="p">(</span><span class="n">coord</span><span class="p">,</span> <span class="n">points</span><span class="p">)</span> <span class="o">=</span> <span class="n">numeric_cgr</span><span class="p">(</span><span class="n">pq_exp</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">45</span><span class="p">)</span>
<span class="n">ax2</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">coord</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="p">:],</span> <span class="n">coord</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="p">:],</span> <span class="s1">&#39;k&#39;</span><span class="p">)</span>
<span class="n">ax2</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">points</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="p">:],</span> <span class="n">points</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="p">:],</span> <span class="s1">&#39;k.&#39;</span><span class="p">,</span> <span class="n">markersize</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="n">ax2</span><span class="o">.</span><span class="n">axis</span><span class="p">(</span><span class="s1">&#39;off&#39;</span><span class="p">)</span>
<span class="n">ax2</span><span class="o">.</span><span class="n">set_aspect</span><span class="p">(</span><span class="s1">&#39;equal&#39;</span><span class="p">,</span> <span class="s1">&#39;box&#39;</span><span class="p">)</span>
<span class="n">ax2</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;exp(1) mod 4&#39;</span><span class="p">)</span>

<span class="n">pq_pi</span> <span class="o">=</span> <span class="n">confrac</span><span class="p">(</span><span class="n">sympy</span><span class="o">.</span><span class="n">pi</span><span class="p">,</span> <span class="mi">450</span><span class="p">)</span>
<span class="p">(</span><span class="n">coord</span><span class="p">,</span> <span class="n">points</span><span class="p">)</span> <span class="o">=</span> <span class="n">numeric_cgr</span><span class="p">(</span><span class="n">pq_pi</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">45</span><span class="p">)</span>
<span class="n">ax3</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">coord</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="p">:],</span> <span class="n">coord</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="p">:],</span> <span class="s1">&#39;k&#39;</span><span class="p">)</span>
<span class="n">ax3</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">points</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="p">:],</span> <span class="n">points</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="p">:],</span> <span class="s1">&#39;k.&#39;</span><span class="p">,</span> <span class="n">markersize</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="n">ax3</span><span class="o">.</span><span class="n">axis</span><span class="p">(</span><span class="s1">&#39;off&#39;</span><span class="p">)</span>
<span class="n">ax3</span><span class="o">.</span><span class="n">set_aspect</span><span class="p">(</span><span class="s1">&#39;equal&#39;</span><span class="p">,</span> <span class="s1">&#39;box&#39;</span><span class="p">)</span>
<span class="n">ax3</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;pi mod 4&#39;</span><span class="p">)</span>

<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/chaos-game-representation_104_0.png" src="../_images/chaos-game-representation_104_0.png" />
</div>
</div>
<p>In fact, the distribution of partial quotients is very well known, owing to the results of Khinchin <span id="id54">[<a class="reference internal" href="../references.html#id34" title="A. Y. Khinchin and T. Teichmann. Continued fractions. Physics Today, 17:70, 1964.">Khinchin and Teichmann, 1964</a>]</span>. As a very startling aside, recently, Bill Gosper has remarked on an amazing identity</p>
<div class="amsmath math notranslate nohighlight" id="equation-f37eb2ea-da5c-4866-a3ad-86d10a483868">
<span class="eqno">(75)<a class="headerlink" href="#equation-f37eb2ea-da5c-4866-a3ad-86d10a483868" title="Permalink to this equation">#</a></span>\[\begin{equation}
	\prod k!^{\Delta_{3}\ln k} = K^{\ln 2} ,
\end{equation}\]</div>
<p>where <span class="math notranslate nohighlight">\(K\)</span> is Khinchin’s constant. We did not inform our students of Khinchin’s remarkable results because CGR was only done at the end of the course. Next time, perhaps!</p>
<div class="tip admonition" id="cgractivity-9">
<p class="admonition-title">CGR Activity 9</p>
<p>Create another CGR using the partial quotients of a continued fraction of any irrational number of your choosing. Feel free to use CGRs with different number of vertices or grouping the numbers in different ways (such as what we did for the digits of <span class="math notranslate nohighlight">\(\pi\)</span> in a previous activity). <a class="reference internal" href="../Solutions/Reports%20on%20Chaos%20Game%20Activities.html#cgractivityreport-9"><span class="std std-ref">[What happened when we did this]</span></a></p>
</div>
</section>
</section>
<section id="random-vs-pseudorandom">
<span id="sec-random"></span><h2>Random vs Pseudorandom<a class="headerlink" href="#random-vs-pseudorandom" title="Permalink to this headline">#</a></h2>
<p>The theory of probability arose much later in the development of science than the theory of dynamical systems and of exact trajectories, and therefore, it may be more difficult to grasp. There are many different works on the fundamentals of probability which discuss this difficulty; see for instance <span id="id55">[<a class="reference internal" href="../references.html#id25" title="R. W. Hamming. Art of Probability. Addison Wesley Publishing Company, 1991.">Hamming, 1991</a>]</span>, but also <span id="id56">[<a class="reference internal" href="../references.html#id29" title="M. Kac. Probability and related topics in physical sciences. Volume 1. AMS, 1959.">Kac, 1959</a>]</span>. We believe that combining deterministic dynamics with randomness is even more difficult; and that this, in fact, is the main business nowadays of the applied mathematician.  There have been, of course, millions of words written on the topic.  We believe that it is crucial that entering students see some of the discussions of the fundamentals; they need to have a chance to grasp the deeper, most practical aspects of the theory.  We believe that this module offers the instructor a chance to begin those discussions.</p>
<p>One important early part of the discussion is whether or not what the computer produces is “really” random, and whether entries in a sequence are “really” independent.  We alluded earlier to the fact that the poor quality of some early bad random number generators was detected by patterns arising in two-dimensional pictures.  How are the patterns arising in the pictures in this paper different?  Would they arise if “really” random numbers were used instead of the pseudorandom numbers generated by computer? [The answer is yes.]  And what is the difference, anyway?  There are knots here that the students (and professors) can tie themselves in: once a sequence of numbers is written down, however randomly it was generated, how can it be random, any more?  It’s now perfectly predictable!  The most comforting words that we know about this come from Kolmogorov himself (quoted in full at the end of this paper) which we paraphrase as “it’s only a model”. Indeed see <span id="id57">[<a class="reference internal" href="../references.html#id36" title="A. N. Kolmogorov. On logical foundations of probability theory. In Probability theory and mathematical statistics, pages 1–5. Springer, 1983.">Kolmogorov, 1983</a>]</span> for a brief discussion of a <em>mathematical</em> foundation for probability. This applies whether your sequence is “really” random, or only “pseudo” random.</p>
<p>These discussions are important, in a situation such as that of this paper, where we are trying to tease out deterministic aspects of <em>apparently</em> random sequences (or of sequences such as that of DNA which is surely influenced both by random events (mutation, horizontal gene transfer) and very non-random events (selection)).  We have used some “clearly” non-random (in some sense) sequences such as the digits of <span class="math notranslate nohighlight">\(\pi\)</span> to show that we can detect a signature of randomness there; we have used the same techniques to detect a signature of regularity. To be convinced, the students must be allowed to discuss these issues at some length. In particular, this may be the first time the students have encountered pseudorandom numbers.</p>
<p>There are several high-quality methods for uniform pseudorandom number generation: the most popular includes the <em>multiple-recursive method</em> and families formed by <em>shift-register methods</em> <span id="id58">[<a class="reference internal" href="../references.html#id18" title="M. R. Dennis, P. Glendinning, P. A Martin, F. Santosa, and J. Tanner. The Princeton companion to applied mathematics. Princeton University Press, 2015.">Dennis <em>et al.</em>, 2015</a>]</span>. The multiple-recursive method is an extension of the linear congruential method by replacing the first-order linear recursion by one of higher order. One family in the shift-register method generates uniform pseudorandom numbers by means of linear recurring sequences modulo <span class="math notranslate nohighlight">\(2\)</span>. Another uses vector recursions modulo <span class="math notranslate nohighlight">\(2\)</span> of higher order; this family includes the very popular Mersenne twister MT19937 <span id="id59">[<a class="reference internal" href="../references.html#id37" title="M. Matsumoto and T. Nishimura. Mersenne twister: a 623-dimensionally equidistributed uniform pseudo-random number generator. ACM Transactions on Modeling and Computer Simulation (TOMACS), 8(1):3–30, 1998.">Matsumoto and Nishimura, 1998</a>]</span>, which is the default for Python’s pseudorandom number generating library <a class="reference external" href="https://docs.python.org/3/library/random.html"><code class="docutils literal notranslate"><span class="pre">random</span></code></a>, which includes <code class="docutils literal notranslate"><span class="pre">random.random()</span></code> (which returns a random floating point number in the range <span class="math notranslate nohighlight">\([0.0, 1.0)\)</span>), <code class="docutils literal notranslate"><span class="pre">random.uniform(a,</span> <span class="pre">b)</span></code> (which returns a random floating point number between <span class="math notranslate nohighlight">\(a\)</span> and <span class="math notranslate nohighlight">\(b\)</span>), and <code class="docutils literal notranslate"><span class="pre">random.randint(a,</span> <span class="pre">b)</span></code> (which returns a random integer in the range <span class="math notranslate nohighlight">\([a, b]\)</span>).</p>
<p>As mentioned above, Python’s <code class="docutils literal notranslate"><span class="pre">random</span></code> library uses is the Mersenne twister MT19937 to generate uniform pseudorandom numbers. The MT19937 produces sequences of uniform pseudorandom numbers with period length <span class="math notranslate nohighlight">\(2^{19937} - 1\)</span> that possess 623-dimensional equidistribution up to 32 bits accuracy <span id="id60">[<a class="reference internal" href="../references.html#id18" title="M. R. Dennis, P. Glendinning, P. A Martin, F. Santosa, and J. Tanner. The Princeton companion to applied mathematics. Princeton University Press, 2015.">Dennis <em>et al.</em>, 2015</a>]</span>. This would seem to create a better “random” outcome in comparison to an actual sequence of die rolls according to some measures. However, <a class="reference external" href="https://www.random.org/">random.org</a> claims that their “dice roller” is, for many purposes, better than pseudorandom number algorithms typically used in computer programs. The randomness in their program comes from atmospheric noise.</p>
<p>Using a pseudorandom number generator differs to using dice or the atmospheric-sourced generators at <a class="reference external" href="https://www.random.org/">random.org</a>. One gets different results due to the different methods that are used. Now, physically-based systems that we believe are random are sometimes more problematic than we think. For instance, rolling a die is not quite as random as one would expect. Stein’s article <span id="id61">[<a class="reference internal" href="../references.html#id42" title="B. P. Stein. Dice rolls are not completely random. 2012. URL: https://www.insidescience.org/news/dice-rolls-are-not-completely-random.">Stein, 2012</a>]</span> in Inside Science states that dice rolls are not completely random: the initial position of the (fair) die affects the outcome of the die roll. And is anything in the atmosphere truly random?  Or is this just a statement of our ignorance?  Kolmogorov’s words—it’s only a model—are comforting here. But they help with the use of pseudorandom numbers, as well.</p>
<p>In practice, pseudorandom number generators are indispensable for working with probability, and an enormous amount of work has gone into making them of very high-quality indeed.  We know of no way to distinguish—in practice—results from these good generators from results given by physically-based generators that nearly everyone believes are “really” random.</p>
</section>
<section id="concluding-remarks">
<h2>Concluding Remarks<a class="headerlink" href="#concluding-remarks" title="Permalink to this headline">#</a></h2>
<p>We have only begun a discussion of what it means to be random, and what it means to have a pattern. Randomness in a model can be a way to hide our ignorance, or it can be a profound statement of our understanding. Modelling viral evolution using randomness <span id="id62">[<a class="reference internal" href="../references.html#id44" title="L. M. Wahl and T. Pattenden. Prophage provide a safe haven for adaptive exploration in temperate viruses. Genetics, 206(1):407–416, 2017.">Wahl and Pattenden, 2017</a>]</span>, where mutations occur in large populations, is clearly warranted; similarly for the vast lengths of DNA sequences. For prime numbers, in one sense clearly not: primes cannot be random, even though they behave in some ways as if they were.</p>
<blockquote class="epigraph">
<div><p>Such considerations may be repeated as often as we like, but it is clear that this procedure will never allow us to be free of the necessity, at the last stage, of referring to probabilities in the primitive imprecise sense of this term.</p>
<p>It would be quite wrong to think that difficulties of this kind are peculiar in some way to the theory of probability. In the mathematical investigation of actual events, we always make a model of them. The discrepancies between the actual course of events and the theoretical model can, in its turn, be made the subject of mathematical investigation. But for these discrepancies we must construct a model that we will use without formal mathematical analysis of the discrepancies which again would arise in it in actual experiment.</p>
<p class="attribution">—<a class="reference external" href="https://mathshistory.st-andrews.ac.uk/Biographies/Kolmogorov/">Andrey Kolmogorov</a>, <span id="id63">[<a class="reference internal" href="../references.html#id7" title="A. D. Aleksandrov, M. A. Lavrentév, and others. Mathematics: Its content, methods and meaning. Courier Corporation, 1999.">Aleksandrov <em>et al.</em>, 1999</a>]</span>
</p>
</div></blockquote>
<p>Indeed, much the same comments can be made about our use of the words “fractal” and “attractor” here.  All of our computations are finite, so <em>nothing</em> that we see is actually a “true fractal”, and similarly our computations show only  parts of the transients, not the attracting set (much less the attractor). This doesn’t matter: again, what matters is that fractals and attractors, in the infinite limit, provide good models for the pictures that our finite computations produce.  For more details both about the rigour and the utility of fractals, consult <span id="id64">[<a class="reference internal" href="../references.html#id12" title="B. B. Mandelbrot. Fractals and Chaos. Springer New York, New York, 2004. URL: https://doi.org/10.1007/978-1-4757-4017-2, doi:10.1007/978-1-4757-4017-2.">Mandelbrot, 2004</a>]</span>.  As a final remark, what convinced us of the value of CGR for DNA sequences was nothing mathematical, but rather its wonderfully apropos placement of curious creatures on the “Map of Life”: for instance, picking out an organism mapped to a point near the border between the group of “fish” and the group  of “amphibians” we find an air-breathing fish!</p>
<p>This raises a final point. This technique really works, for DNA sequences; and it can be genuinely informative for other kinds of sequences, as we have seen.  For instance, we were able to detect the influence of rounding errors (for the Fibonacci sequence computation).  One also needs an explanation of <em>why</em> this works; these can make good discussion points with students.  Consider the impact of short stretches of so-called “junk” or non-coding DNA, distributed throughout a long coding sequence.  Will the “junk” make a difference to the CGR representation? In some cases not, but if there is enough junk, then maybe it will.  But in many cases we could claim that one reason this works is that CGR really gives a “whole picture” of the sequence, and is relatively robust under errors such as shifts and interruptions with short insertions.  Then one should investigate that claim, to see if it was true.</p>
<hr class="footnotes docutils" />
<dl class="footnote brackets">
<dt class="label" id="id65"><span class="brackets"><a class="fn-backref" href="#id6">1</a></span></dt>
<dd><p>To be more specific, this is a <em>hyperbolic</em> iterated function system. However, the word “hyperbolic” is sometimes dropped in practice <span id="id66">[<a class="reference internal" href="../references.html#id14" title="M. F. Barnsley. Fractals everywhere. Academic press, 2014.">Barnsley, 2014</a>]</span>(p. 82).</p>
</dd>
<dt class="label" id="id67"><span class="brackets"><a class="fn-backref" href="#id7">2</a></span></dt>
<dd><p>It can be proved that in the limit of infinite computation for a random sequence of infinite length that the picture really is that of the Sierpinski triangle <span id="id68">[<a class="reference internal" href="../references.html#id13" title="M. F. Barnsley and A. Vince. The chaos game on a general iterated function system. Ergodic Theory and Dynamical Systems, 31(4):1073–1079, 2011.">Barnsley and Vince, 2011</a>]</span>.</p>
</dd>
<dt class="label" id="id69"><span class="brackets"><a class="fn-backref" href="#id10">3</a></span></dt>
<dd><p>The word “game” here does not refer to an adversarial game with choices and strategies as in <em>game theory</em>, but instead treats sequences as something like the outcomes of a game of chance like tossing coins. The use of “chaos” here happened because the biologist H.J. Jeffrey used the word, likely to indicate generally from which area of mathematics the ideas were taken from.</p>
</dd>
<dt class="label" id="id70"><span class="brackets"><a class="fn-backref" href="#id14">4</a></span></dt>
<dd><p>We tried this, though not in class. We got up to 80 points on one figure, using the die roller on <a class="reference external" href="https://www.random.org/">random.org</a>. It’s kind of fun, in a physical “drawing and measuring” way, but it’s hard not to make mistakes (it might be easier to be correct if we are working in a group). It could make a useful activity and we might try it in class in the future.</p>
</dd>
<dt class="label" id="id71"><span class="brackets"><a class="fn-backref" href="#id17">5</a></span></dt>
<dd><p>This is a loose way of thinking about it: we somehow imagine regions of the gameboard that equally attract elements of the sequence, and if these regions overlap, then we see spuriously dense regions on the gameboard.  Alternatively, we could think (more rigorously) about the <em>probability measure</em> of the attractor (now really spread over the whole gameboard).  One would want a “random” sequence to have a uniform measure across the gameboard.</p>
</dd>
<dt class="label" id="id72"><span class="brackets"><a class="fn-backref" href="#id50">6</a></span></dt>
<dd><p><span id="id73">[<a class="reference internal" href="../references.html#id48" title="Francisco J Aragón Artacho, David H Bailey, Jonathan M Borwein, and Peter B Borwein. Walking on real numbers. Math. Intelligencer, 35(1):42–60, March 2013.">Aragón Artacho <em>et al.</em>, 2013</a>]</span> performed a deterministic walk over the digits of <span class="math notranslate nohighlight">\(\pi\)</span> and compared it visually with pseudorandom walks of the same length to determine whether the digits of <span class="math notranslate nohighlight">\(\pi\)</span> are random. See their paper for details (and their pretty visualizations of their walk over the digits of <span class="math notranslate nohighlight">\(\pi\)</span>).</p>
</dd>
<dt class="label" id="id74"><span class="brackets"><a class="fn-backref" href="#id51">7</a></span></dt>
<dd><p>The round-to-even rule is a topic that many students stumble over. As is pointed out in <span id="id75">[<a class="reference internal" href="../references.html#id8" title="N. J. Higham. Accuracy and Stability of Numerical Algorithms. Society for Industrial and Applied Mathematics, January 2002. doi:10.1137/1.9780898718027.">Higham, 2002</a>]</span>(Section 1.17), rounding errors are not random; but the round-to-even rule was invented to reduce the <em>bias</em> of rounding error in long sequences of computation. This sounds paradoxical but although rounding errors are not random, or independent, there are situations where it is a useful model to assume so: see <span id="id76">[<a class="reference internal" href="../references.html#id8" title="N. J. Higham. Accuracy and Stability of Numerical Algorithms. Society for Industrial and Applied Mathematics, January 2002. doi:10.1137/1.9780898718027.">Higham, 2002</a>]</span>(Section 2.8) for further discussion of this point.</p>
</dd>
</dl>
</section>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            kernelName: "python3",
            path: "./Contents"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

              </div>
              
            </main>
            <footer class="footer-article noprint">
                
    <!-- Previous / next buttons -->
<div class='prev-next-area'>
    <a class='left-prev' id="prev-link" href="mandelbrot.html" title="previous page">
        <i class="fas fa-angle-left"></i>
        <div class="prev-next-info">
            <p class="prev-next-subtitle">previous</p>
            <p class="prev-next-title">Mandelbrot Polynomials and Matrices</p>
        </div>
    </a>
    <a class='right-next' id="next-link" href="../Looking-back.html" title="next page">
    <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Zanshin</p>
    </div>
    <i class="fas fa-angle-right"></i>
    </a>
</div>
            </footer>
        </div>
    </div>
    <div class="footer-content row">
        <footer class="col footer"><p>
  
    By Neil J. Calkin, Eunice Y.S. Chan, and Robert M. Corless<br/>
  
      &copy; Copyright 2022.<br/>
</p>
        </footer>
    </div>
    
</div>


      </div>
    </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../_static/scripts/pydata-sphinx-theme.js?digest=1999514e3f237ded88cf"></script>


  </body>
</html>